<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SpliceWiz: the cookbook • SpliceWiz</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SpliceWiz: the cookbook">
<meta property="og:description" content="SpliceWiz">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SpliceWiz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SW_Cookbook.html">SpliceWiz: the cookbook</a>
    </li>
    <li>
      <a href="../articles/SW_QuickStart.html">SpliceWiz: Quick Start</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/alexchwong/SpliceWiz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SpliceWiz: the cookbook</h1>
                        <h4 data-toc-skip class="author">Alex CH
Wong</h4>
            
            <h4 data-toc-skip class="date">11/15/2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/alexchwong/SpliceWiz/blob/HEAD/vignettes/SW_Cookbook.Rmd" class="external-link"><code>vignettes/SW_Cookbook.Rmd</code></a></small>
      <div class="hidden name"><code>SW_Cookbook.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      This vignette is a guide containing example code for performing
      real-life tasks. Importantly, it covers some functionality that
      were not covered in the Quick-Start vignette (because they are too
      computationally intensive to be reproducible in a vignette).
      Version 1.1.2
    </div>
    
<p><br></p>
<div class="section level2">
<h2 id="loading-splicewiz">Loading SpliceWiz<a class="anchor" aria-label="anchor" href="#loading-splicewiz"></a>
</h2>
<p>For instructions on installing and configuring SpliceWiz, please see
the Quick-Start vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/alexchwong/SpliceWiz" class="external-link">SpliceWiz</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: NxtIRFdata</span></span>
<span><span class="co">#&gt; SpliceWiz package loaded with 1 threads</span></span>
<span><span class="co">#&gt; Use setSWthreads() to set the number of SpliceWiz threads</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="reference-generation">Reference Generation<a class="anchor" aria-label="anchor" href="#reference-generation"></a>
</h2>
<p>First, define the path to the directory in which the reference should
be stored. This directory will be made by SpliceWiz, but its parent
directory must exist, otherwise an error will be returned.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref_path</span> <span class="op">&lt;-</span> <span class="st">"./Reference"</span></span></code></pre></div>
<p><br></p>
<div class="section level4">
<h4 id="create-a-splicewiz-reference-from-user-defined-fasta-and-gtf-files-locally">Create a SpliceWiz reference from user-defined FASTA and GTF files
locally<a class="anchor" aria-label="anchor" href="#create-a-splicewiz-reference-from-user-defined-fasta-and-gtf-files-locally"></a>
</h4>
<p>Note that setting <code>genome_path = "hg38"</code> will prompt
SpliceWiz to use the default files for nonPolyA and Mappability
exclusion references in the generation of its reference. Valid options
for <code>genome_path</code> are “hg38”, “hg19”, “mm10” and “mm9”.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Build-Reference-methods.html">buildRef</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    fasta <span class="op">=</span> <span class="st">"genome.fa"</span>, gtf <span class="op">=</span> <span class="st">"transcripts.gtf"</span>,</span>
<span>    genome_type <span class="op">=</span> <span class="st">"hg38"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="prepare-genome-resources-and-building-the-reference-as-separate-steps">Prepare genome resources and building the reference as separate
steps<a class="anchor" aria-label="anchor" href="#prepare-genome-resources-and-building-the-reference-as-separate-steps"></a>
</h4>
<p><code><a href="../reference/Build-Reference-methods.html">buildRef()</a></code> first fetches the genome and gene annotations
and makes a compressed local copy in the <code>resources</code>
subdirectory of the given reference path. This can sometimes be a long
process (especially when downloading the genome from the internet).
Also, one may choose to generate the mappability exclusion regions
manually (see <code>Reference Generation</code> -
<code>Mappability exclusion generation using Rsubread</code> section).
This needs to occur prior to generation of the SpliceWiz reference.</p>
<p>To separately prepare the annotation data and build the SpliceWiz
reference, use the <code><a href="../reference/Build-Reference-methods.html">getResources()</a></code> function to specify the
FASTA and GTF files. Once <code><a href="../reference/Build-Reference-methods.html">getResources()</a></code> has completed
successfully, call <code><a href="../reference/Build-Reference-methods.html">buildRef()</a></code> and leave the
<code>fasta</code> and <code>gtf</code> arguments blank, as in the
example below:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Build-Reference-methods.html">getResources</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    fasta <span class="op">=</span> <span class="st">"genome.fa"</span>,</span>
<span>    gtf <span class="op">=</span> <span class="st">"transcripts.gtf"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/Build-Reference-methods.html">buildRef</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    genome_type <span class="op">=</span> <span class="st">"hg38"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="overwriting-an-existing-reference-but-using-the-same-annotations">Overwriting an existing reference, but using the same
annotations<a class="anchor" aria-label="anchor" href="#overwriting-an-existing-reference-but-using-the-same-annotations"></a>
</h4>
<p>To re-build and overwrite an existing reference, using the same
resource annotations:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assuming hg38 genome:</span></span>
<span></span>
<span><span class="fu"><a href="../reference/Build-Reference-methods.html">buildRef</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    genome_type <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>    overwrite <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="create-a-splicewiz-reference-using-web-resources-from-ensembls-ftp">Create a SpliceWiz reference using web resources from Ensembl’s
FTP<a class="anchor" aria-label="anchor" href="#create-a-splicewiz-reference-using-web-resources-from-ensembls-ftp"></a>
</h4>
<p>The following will first download the genome and gene annotation
files from the online resource and store a local copy of it in a file
cache, facilitated by BiocFileCache. Then, it uses the downloaded
resource to create the SpliceWiz reference.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">FTP</span> <span class="op">&lt;-</span> <span class="st">"ftp://ftp.ensembl.org/pub/release-94/"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/Build-Reference-methods.html">buildRef</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    fasta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">FTP</span>, <span class="st">"fasta/homo_sapiens/dna/"</span>,</span>
<span>        <span class="st">"Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz"</span><span class="op">)</span>, </span>
<span>    gtf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">FTP</span>, <span class="st">"gtf/homo_sapiens/"</span>,</span>
<span>        <span class="st">"Homo_sapiens.GRCh38.94.chr.gtf.gz"</span><span class="op">)</span>,</span>
<span>    genome_type <span class="op">=</span> <span class="st">"hg38"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="create-a-splicewiz-reference-using-annotationhub-resources">Create a SpliceWiz reference using AnnotationHub resources<a class="anchor" aria-label="anchor" href="#create-a-splicewiz-reference-using-annotationhub-resources"></a>
</h4>
<p>AnnotationHub contains Ensembl references for many genomes. To browse
what is available:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: AnnotationHub</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: BiocFileCache</span></span>
<span><span class="co">#&gt; Loading required package: dbplyr</span></span>
<span></span>
<span><span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; snapshotDate(): 2022-10-26</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">ah</span>, <span class="st">"Ensembl"</span><span class="op">)</span></span>
<span><span class="co">#&gt; AnnotationHub with 33426 records</span></span>
<span><span class="co">#&gt; # snapshotDate(): 2022-10-26</span></span>
<span><span class="co">#&gt; # $dataprovider: Ensembl, FANTOM5,DLRP,IUPHAR,HPRD,STRING,SWISSPROT,TREMBL,E...</span></span>
<span><span class="co">#&gt; # $species: Homo sapiens, Mus musculus, Danio rerio, Rattus norvegicus, Pan ...</span></span>
<span><span class="co">#&gt; # $rdataclass: TwoBitFile, GRanges, EnsDb, SQLiteFile, data.frame, OrgDb, list</span></span>
<span><span class="co">#&gt; # additional mcols(): taxonomyid, genome, description,</span></span>
<span><span class="co">#&gt; #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,</span></span>
<span><span class="co">#&gt; #   rdatapath, sourceurl, sourcetype </span></span>
<span><span class="co">#&gt; # retrieve records with, e.g., 'object[["AH5046"]]' </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;              title                                                       </span></span>
<span><span class="co">#&gt;   AH5046   | Ensembl Genes                                               </span></span>
<span><span class="co">#&gt;   AH5160   | Ensembl Genes                                               </span></span>
<span><span class="co">#&gt;   AH5311   | Ensembl Genes                                               </span></span>
<span><span class="co">#&gt;   AH5434   | Ensembl Genes                                               </span></span>
<span><span class="co">#&gt;   AH5435   | Ensembl EST Genes                                           </span></span>
<span><span class="co">#&gt;   ...        ...                                                         </span></span>
<span><span class="co">#&gt;   AH107263 | LRBaseDb for Taeniopygia guttata (Zebra finch, v004)        </span></span>
<span><span class="co">#&gt;   AH107264 | LRBaseDb for Takifugu rubripes (Fugu, v004)                 </span></span>
<span><span class="co">#&gt;   AH107265 | LRBaseDb for Ursus maritimus (Polar bear, v004)             </span></span>
<span><span class="co">#&gt;   AH107266 | LRBaseDb for Vulpes vulpes (Red fox, v004)                  </span></span>
<span><span class="co">#&gt;   AH107267 | LRBaseDb for Xenopus tropicalis (Tropical clawed frog, v004)</span></span></code></pre></div>
<p>For a more specific query:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">ah</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Homo Sapiens"</span>, <span class="st">"release-94"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; AnnotationHub with 9 records</span></span>
<span><span class="co">#&gt; # snapshotDate(): 2022-10-26</span></span>
<span><span class="co">#&gt; # $dataprovider: Ensembl</span></span>
<span><span class="co">#&gt; # $species: Homo sapiens</span></span>
<span><span class="co">#&gt; # $rdataclass: TwoBitFile, GRanges</span></span>
<span><span class="co">#&gt; # additional mcols(): taxonomyid, genome, description,</span></span>
<span><span class="co">#&gt; #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,</span></span>
<span><span class="co">#&gt; #   rdatapath, sourceurl, sourcetype </span></span>
<span><span class="co">#&gt; # retrieve records with, e.g., 'object[["AH64628"]]' </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;             title                                           </span></span>
<span><span class="co">#&gt;   AH64628 | Homo_sapiens.GRCh38.94.abinitio.gtf             </span></span>
<span><span class="co">#&gt;   AH64629 | Homo_sapiens.GRCh38.94.chr.gtf                  </span></span>
<span><span class="co">#&gt;   AH64630 | Homo_sapiens.GRCh38.94.chr_patch_hapl_scaff.gtf </span></span>
<span><span class="co">#&gt;   AH64631 | Homo_sapiens.GRCh38.94.gtf                      </span></span>
<span><span class="co">#&gt;   AH65744 | Homo_sapiens.GRCh38.cdna.all.2bit               </span></span>
<span><span class="co">#&gt;   AH65745 | Homo_sapiens.GRCh38.dna.primary_assembly.2bit   </span></span>
<span><span class="co">#&gt;   AH65746 | Homo_sapiens.GRCh38.dna_rm.primary_assembly.2bit</span></span>
<span><span class="co">#&gt;   AH65747 | Homo_sapiens.GRCh38.dna_sm.primary_assembly.2bit</span></span>
<span><span class="co">#&gt;   AH65748 | Homo_sapiens.GRCh38.ncrna.2bit</span></span></code></pre></div>
<p>We wish to fetch “AH65745” and “AH64631” which contains the desired
FASTA and GTF files, respectively. To build a reference using these
resources:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Build-Reference-methods.html">buildRef</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    fasta <span class="op">=</span> <span class="st">"AH65745"</span>,</span>
<span>    gtf <span class="op">=</span> <span class="st">"AH64631"</span>,</span>
<span>    genome_type <span class="op">=</span> <span class="st">"hg38"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code>Build-Reference-methods</code> will recognise the inputs of
<code>fasta</code> and <code>gtf</code> as AnnotationHub resources as
they begin with “AH”.<br><br></p>
</div>
<div class="section level4">
<h4 id="create-a-splicewiz-reference-from-species-other-than-human-or-mouse">Create a SpliceWiz reference from species other than human or
mouse<a class="anchor" aria-label="anchor" href="#create-a-splicewiz-reference-from-species-other-than-human-or-mouse"></a>
</h4>
<p>For human and mouse genomes, we highly recommend specifying
<code>genome_type</code> as the default mappability file is used to
exclude intronic regions with repeat sequences from intron retention
analysis. For other species, one could generate a SpliceWiz reference
without this reference:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Build-Reference-methods.html">buildRef</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    fasta <span class="op">=</span> <span class="st">"genome.fa"</span>, gtf <span class="op">=</span> <span class="st">"transcripts.gtf"</span>,</span>
<span>    genome_type <span class="op">=</span> <span class="st">""</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="star-reference-generation">STAR reference generation<a class="anchor" aria-label="anchor" href="#star-reference-generation"></a>
</h2>
<div class="section level4">
<h4 id="checking-if-star-is-installed">Checking if STAR is installed<a class="anchor" aria-label="anchor" href="#checking-if-star-is-installed"></a>
</h4>
<p>To use <code>STAR</code> to align FASTQ files, one must be using a
system with <code>STAR</code> installed. This software is not available
in Windows. To check if <code>STAR</code> is available:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/STAR-methods.html">STAR_version</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Nov 15 16:17:38 STAR wrappers are only supported on Linux</span></span>
<span><span class="co">#&gt; Nov 15 16:17:38 STAR is not installed</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="building-a-star-reference">Building a STAR reference<a class="anchor" aria-label="anchor" href="#building-a-star-reference"></a>
</h4>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref_path</span> <span class="op">=</span> <span class="st">"./Reference"</span></span>
<span></span>
<span><span class="co"># Ensure genome resources are prepared from genome FASTA and GTF file:</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ref_path</span>, <span class="st">"resource"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/Build-Reference-methods.html">getResources</a></span><span class="op">(</span></span>
<span>        reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>        fasta <span class="op">=</span> <span class="st">"genome.fa"</span>,</span>
<span>        gtf <span class="op">=</span> <span class="st">"transcripts.gtf"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Generate a STAR genome reference:</span></span>
<span><span class="fu">STAR_BuildRef</span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="building-a-star-reference-alongside-the-splicewiz-reference">Building a STAR reference alongside the SpliceWiz reference<a class="anchor" aria-label="anchor" href="#building-a-star-reference-alongside-the-splicewiz-reference"></a>
</h4>
<p>If <code>STAR</code> is available on the same computer or server
where R/RStudio is being run, we can use the one-line function
<code>buildFullRef</code>. This function will: * Prepare the resources
from the given FASTA and GTF files * Generate a STAR genome * Use the
STAR genome and the FASTA file to <em>de-novo</em> calculate and define
low mappability regions * Build the SpliceWiz reference using the genome
resources and mappability file</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Build-Reference-methods.html">buildFullRef</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    fasta <span class="op">=</span> <span class="st">"genome.fa"</span>, gtf <span class="op">=</span> <span class="st">"transcripts.gtf"</span>,</span>
<span>    genome_type <span class="op">=</span> <span class="st">""</span>,</span>
<span>    use_STAR_mappability <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code>n_threads</code> specify how many threads should be used to
build the STAR reference and to calculate the low mappability
regions<br><br></p>
</div>
<div class="section level4">
<h4 id="mappability-exclusion-generation-using-rsubread">Mappability exclusion generation using Rsubread<a class="anchor" aria-label="anchor" href="#mappability-exclusion-generation-using-rsubread"></a>
</h4>
<p>Finally, if <code>STAR</code> is not available, <code>Rsubread</code>
is available on Bioconductor to perform mappability calculations. The
example code in the manual is displayed here for convenience, to
demonstrate how this would be done:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># (1a) Creates genome resource files </span></span>
<span></span>
<span><span class="va">ref_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"Reference"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/Build-Reference-methods.html">getResources</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    fasta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/NxtIRFdata/man/NxtIRFdata-package.html" class="external-link">chrZ_genome</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    gtf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/NxtIRFdata/man/NxtIRFdata-package.html" class="external-link">chrZ_gtf</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (1b) Systematically generate reads based on the SpliceWiz example genome:</span></span>
<span></span>
<span><span class="fu"><a href="../reference/Mappability-methods.html">generateSyntheticReads</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (2) Align the generated reads using Rsubread:</span></span>
<span></span>
<span><span class="co"># (2a) Build the Rsubread genome index:</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">ref_path</span><span class="op">)</span></span>
<span><span class="fu">Rsubread</span><span class="fu">::</span><span class="fu">buildindex</span><span class="op">(</span>basename <span class="op">=</span> <span class="st">"./reference_index"</span>, </span>
<span>    reference <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/NxtIRFdata/man/NxtIRFdata-package.html" class="external-link">chrZ_genome</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (2b) Align the synthetic reads using Rsubread::subjunc()</span></span>
<span></span>
<span><span class="fu">Rsubread</span><span class="fu">::</span><span class="fu">subjunc</span><span class="op">(</span></span>
<span>    index <span class="op">=</span> <span class="st">"./reference_index"</span>, </span>
<span>    readfile1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ref_path</span>, <span class="st">"Mappability"</span>, <span class="st">"Reads.fa"</span><span class="op">)</span>, </span>
<span>    output_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ref_path</span>, <span class="st">"Mappability"</span>, <span class="st">"AlignedReads.bam"</span><span class="op">)</span>, </span>
<span>    useAnnotation <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    annot.ext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/NxtIRFdata/man/NxtIRFdata-package.html" class="external-link">chrZ_gtf</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>    isGTF <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (3) Analyse the aligned reads in the BAM file for low-mappability regions:</span></span>
<span></span>
<span><span class="fu"><a href="../reference/Mappability-methods.html">calculateMappability</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    aligned_bam <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ref_path</span>, <span class="st">"Mappability"</span>, <span class="st">"AlignedReads.bam"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (4) Build the SpliceWiz reference using the calculated Mappability Exclusions</span></span>
<span></span>
<span><span class="fu"><a href="../reference/Build-Reference-methods.html">buildRef</a></span><span class="op">(</span><span class="va">ref_path</span><span class="op">)</span>      </span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="aligning-raw-rna-seq-data-using-splicewizs-star-wrappers">Aligning Raw RNA-seq data using SpliceWiz’s STAR wrappers<a class="anchor" aria-label="anchor" href="#aligning-raw-rna-seq-data-using-splicewizs-star-wrappers"></a>
</h2>
<p><br>
First, remember to check that STAR is available via command line:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/STAR-methods.html">STAR_version</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Nov 15 16:17:39 STAR wrappers are only supported on Linux</span></span>
<span><span class="co">#&gt; Nov 15 16:17:39 STAR is not installed</span></span></code></pre></div>
<div class="section level4">
<h4 id="aligning-a-single-sample-using-star">Aligning a single sample using STAR<a class="anchor" aria-label="anchor" href="#aligning-a-single-sample-using-star"></a>
</h4>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/STAR-methods.html">STAR_alignReads</a></span><span class="op">(</span></span>
<span>    STAR_ref_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ref_path</span>, <span class="st">"STAR"</span><span class="op">)</span>,</span>
<span>    BAM_output_path <span class="op">=</span> <span class="st">"./bams/sample1"</span>,</span>
<span>    fastq_1 <span class="op">=</span> <span class="st">"sample1_1.fastq"</span>, fastq_2 <span class="op">=</span> <span class="st">"sample1_2.fastq"</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="aligning-multiple-samples-using-star">Aligning multiple samples using STAR<a class="anchor" aria-label="anchor" href="#aligning-multiple-samples-using-star"></a>
</h4>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_A"</span>, <span class="st">"sample_B"</span><span class="op">)</span>,</span>
<span>    forward <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"raw_data"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_A"</span>, <span class="st">"sample_B"</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_A_1.fastq"</span>, <span class="st">"sample_B_1.fastq"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    reverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"raw_data"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_A"</span>, <span class="st">"sample_B"</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_A_2.fastq"</span>, <span class="st">"sample_B_2.fastq"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/STAR-methods.html">STAR_alignExperiment</a></span><span class="op">(</span></span>
<span>    Experiment <span class="op">=</span> <span class="va">Experiment</span>,</span>
<span>    STAR_ref_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"Reference_FTP"</span>, <span class="st">"STAR"</span><span class="op">)</span>,</span>
<span>    BAM_output_path <span class="op">=</span> <span class="st">"./bams"</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">8</span>,</span>
<span>    two_pass <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>To use two-pass mapping, set <code>two_pass = TRUE</code>. We
recommend disabling this feature, as one-pass mapping is adequate in
typical-use cases.<br><br></p>
</div>
<div class="section level4">
<h4 id="finding-fastq-files-recursively-from-a-given-directory">Finding FASTQ files recursively from a given directory<a class="anchor" aria-label="anchor" href="#finding-fastq-files-recursively-from-a-given-directory"></a>
</h4>
<p>SpliceWiz can identify sequencing FASTQ files recursively from a
given directory. It assumes that forward and reverse reads are suffixed
as <code>_1</code> and <code>_2</code>, respectively. Users can choose
to identify such files using a specified file extension. For example, to
recursively identify FASTQ files of the format
<code>{sample}_1.fq.gz</code> and <code>{sample}_2.fq.gz</code>, use the
following:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assuming sequencing files are named by their respective sample names</span></span>
<span><span class="va">fastq_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findSamples.html">findFASTQ</a></span><span class="op">(</span><span class="st">"./sequencing_files"</span>, paired <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    fastq_suffix <span class="op">=</span> <span class="st">".fq.gz"</span>, level <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>Then, these can be aligned as follows:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/STAR-methods.html">STAR_alignExperiment</a></span><span class="op">(</span></span>
<span>    Experiment <span class="op">=</span> <span class="va">fastq_files</span>,</span>
<span>    STAR_ref_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"Reference_FTP"</span>, <span class="st">"STAR"</span><span class="op">)</span>,</span>
<span>    BAM_output_path <span class="op">=</span> <span class="st">"./bams"</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">8</span>,</span>
<span>    two_pass <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-processbam-on-bam-files">Running processBAM on BAM files<a class="anchor" aria-label="anchor" href="#running-processbam-on-bam-files"></a>
</h2>
<p>To conveniently find all BAM files recursively in a given path:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bams</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findSamples.html">findBAMS</a></span><span class="op">(</span><span class="st">"./bams"</span>, level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>This convenience function returns the putative sample names, either
from BAM file names themselves (<code>level = 0</code>), or from the
names of their parent directories (<code>level = 1</code>).</p>
<p>To run <code><a href="../reference/processBAM.html">processBAM()</a></code> using 4 OpenMP threads:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># assume SpliceWiz reference has been generated in `ref_path`</span></span>
<span></span>
<span><span class="fu"><a href="../reference/processBAM.html">processBAM</a></span><span class="op">(</span></span>
<span>    bamfiles <span class="op">=</span> <span class="va">bams</span><span class="op">$</span><span class="va">path</span>,</span>
<span>    sample_names <span class="op">=</span> <span class="va">bams</span><span class="op">$</span><span class="va">sample</span>,</span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    output_path <span class="op">=</span> <span class="st">"./pb_output"</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">4</span>,</span>
<span>    useOpenMP <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level4">
<h4 id="creating-cov-files-from-bam-files-without-running-processbam">Creating COV files from BAM files without running processBAM<a class="anchor" aria-label="anchor" href="#creating-cov-files-from-bam-files-without-running-processbam"></a>
</h4>
<p>Sometimes one may wish to create a COV file from a BAM file without
running <code><a href="../reference/processBAM.html">processBAM()</a></code>. One reason might be because a
SpliceWiz reference is not available.</p>
<p>To convert a list of BAM files, run <code><a href="../reference/processBAM.html">BAM2COV()</a></code>. This is a
function structurally similar to <code><a href="../reference/processBAM.html">processBAM()</a></code> but without
the need to give the path to the SpliceWiz reference:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/processBAM.html">BAM2COV</a></span><span class="op">(</span></span>
<span>    bamfiles <span class="op">=</span> <span class="va">bams</span><span class="op">$</span><span class="va">path</span>,</span>
<span>    sample_names <span class="op">=</span> <span class="va">bams</span><span class="op">$</span><span class="va">sample</span>,</span>
<span>    output_path <span class="op">=</span> <span class="st">"./pb_output"</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">4</span>,</span>
<span>    useOpenMP <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="collating-the-experiment">Collating the experiment<a class="anchor" aria-label="anchor" href="#collating-the-experiment"></a>
</h2>
<p>Assuming the SpliceWiz reference is in <code>ref_path</code>, after
running <code><a href="../reference/processBAM.html">processBAM()</a></code> as shown in the previous section, use
the convenience function <code><a href="../reference/findSamples.html">findSpliceWizOutput()</a></code> to tabulate
a list of samples and their corresponding <code><a href="../reference/processBAM.html">processBAM()</a></code>
outputs:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findSamples.html">findSpliceWizOutput</a></span><span class="op">(</span><span class="st">"./pb_output"</span><span class="op">)</span></span></code></pre></div>
<p>This data.frame can be directly used to run
<code>collateData</code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/collateData.html">collateData</a></span><span class="op">(</span></span>
<span>    Experiment <span class="op">=</span> <span class="va">expr</span>,</span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    output_path <span class="op">=</span> <span class="st">"./NxtSE_output"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>NB: Novel splicing detection can be enabled by setting
<code>novelSplicing = TRUE</code>. See the SpliceWiz Quick-Start
vignette for more details.</li>
</ul>
<p>Then, the collated data can be imported as a <code>NxtSE</code>
object, which is an object that inherits
<code>SummarizedExperiment</code> and has specialized containers to hold
additional data required by SpliceWiz.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSE.html">makeSE</a></span><span class="op">(</span><span class="st">"./NxtSE_output"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="downstream-analysis-using-splicewiz">Downstream analysis using SpliceWiz<a class="anchor" aria-label="anchor" href="#downstream-analysis-using-splicewiz"></a>
</h2>
<p>Please refer to SpliceWiz: Quick-Start vignette for worked examples
using the example dataset.<br><br></p>
</div>
<div class="section level2">
<h2 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Big Sur ... 10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] AnnotationHub_3.6.0 BiocFileCache_2.6.0 dbplyr_2.2.1       </span></span>
<span><span class="co">#&gt; [4] BiocGenerics_0.44.0 SpliceWiz_1.1.2     NxtIRFdata_1.4.0   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] systemfonts_1.0.4             lazyeval_0.2.2               </span></span>
<span><span class="co">#&gt;   [3] shinydashboard_0.7.2          splines_4.2.2                </span></span>
<span><span class="co">#&gt;   [5] BiocParallel_1.32.1           GenomeInfoDb_1.34.3          </span></span>
<span><span class="co">#&gt;   [7] ggplot2_3.4.0                 digest_0.6.30                </span></span>
<span><span class="co">#&gt;   [9] ca_0.71.1                     foreach_1.5.2                </span></span>
<span><span class="co">#&gt;  [11] htmltools_0.5.3               viridis_0.6.2                </span></span>
<span><span class="co">#&gt;  [13] fansi_1.0.3                   magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;  [15] memoise_2.0.1                 BSgenome_1.66.1              </span></span>
<span><span class="co">#&gt;  [17] shinyFiles_0.9.3              Biostrings_2.66.0            </span></span>
<span><span class="co">#&gt;  [19] annotate_1.76.0               matrixStats_0.62.0           </span></span>
<span><span class="co">#&gt;  [21] R.utils_2.12.2                pkgdown_2.0.6                </span></span>
<span><span class="co">#&gt;  [23] prettyunits_1.1.1             colorspace_2.0-3             </span></span>
<span><span class="co">#&gt;  [25] blob_1.2.3                    rappdirs_0.3.3               </span></span>
<span><span class="co">#&gt;  [27] textshaping_0.3.6             xfun_0.34                    </span></span>
<span><span class="co">#&gt;  [29] dplyr_1.0.10                  crayon_1.5.2                 </span></span>
<span><span class="co">#&gt;  [31] RCurl_1.98-1.9                jsonlite_1.8.3               </span></span>
<span><span class="co">#&gt;  [33] genefilter_1.80.0             iterators_1.0.14             </span></span>
<span><span class="co">#&gt;  [35] survival_3.4-0                glue_1.6.2                   </span></span>
<span><span class="co">#&gt;  [37] registry_0.5-1                gtable_0.3.1                 </span></span>
<span><span class="co">#&gt;  [39] zlibbioc_1.44.0               XVector_0.38.0               </span></span>
<span><span class="co">#&gt;  [41] webshot_0.5.4                 DelayedArray_0.24.0          </span></span>
<span><span class="co">#&gt;  [43] Rhdf5lib_1.20.0               HDF5Array_1.26.0             </span></span>
<span><span class="co">#&gt;  [45] scales_1.2.1                  pheatmap_1.0.12              </span></span>
<span><span class="co">#&gt;  [47] DBI_1.1.3                     Rcpp_1.0.9                   </span></span>
<span><span class="co">#&gt;  [49] progress_1.2.2                viridisLite_0.4.1            </span></span>
<span><span class="co">#&gt;  [51] xtable_1.8-4                  bit_4.0.4                    </span></span>
<span><span class="co">#&gt;  [53] stats4_4.2.2                  DT_0.26                      </span></span>
<span><span class="co">#&gt;  [55] htmlwidgets_1.5.4             httr_1.4.4                   </span></span>
<span><span class="co">#&gt;  [57] fstcore_0.9.12                RColorBrewer_1.1-3           </span></span>
<span><span class="co">#&gt;  [59] ellipsis_0.3.2                pkgconfig_2.0.3              </span></span>
<span><span class="co">#&gt;  [61] XML_3.99-0.12                 R.methodsS3_1.8.2            </span></span>
<span><span class="co">#&gt;  [63] sass_0.4.2                    utf8_1.2.2                   </span></span>
<span><span class="co">#&gt;  [65] tidyselect_1.2.0              rlang_1.0.6                  </span></span>
<span><span class="co">#&gt;  [67] later_1.3.0                   AnnotationDbi_1.60.0         </span></span>
<span><span class="co">#&gt;  [69] munsell_0.5.0                 BiocVersion_3.16.0           </span></span>
<span><span class="co">#&gt;  [71] tools_4.2.2                   cachem_1.0.6                 </span></span>
<span><span class="co">#&gt;  [73] cli_3.4.1                     generics_0.1.3               </span></span>
<span><span class="co">#&gt;  [75] RSQLite_2.2.18                evaluate_0.18                </span></span>
<span><span class="co">#&gt;  [77] stringr_1.4.1                 fastmap_1.1.0                </span></span>
<span><span class="co">#&gt;  [79] heatmaply_1.4.0               yaml_2.3.6                   </span></span>
<span><span class="co">#&gt;  [81] ragg_1.2.4                    ompBAM_1.2.0                 </span></span>
<span><span class="co">#&gt;  [83] knitr_1.40                    bit64_4.0.5                  </span></span>
<span><span class="co">#&gt;  [85] fs_1.5.2                      purrr_0.3.5                  </span></span>
<span><span class="co">#&gt;  [87] KEGGREST_1.38.0               dendextend_1.16.0            </span></span>
<span><span class="co">#&gt;  [89] sparseMatrixStats_1.10.0      mime_0.12                    </span></span>
<span><span class="co">#&gt;  [91] rhandsontable_0.3.8           R.oo_1.25.0                  </span></span>
<span><span class="co">#&gt;  [93] compiler_4.2.2                plotly_4.10.1                </span></span>
<span><span class="co">#&gt;  [95] filelock_1.0.2                curl_4.3.3                   </span></span>
<span><span class="co">#&gt;  [97] png_0.1-7                     interactiveDisplayBase_1.36.0</span></span>
<span><span class="co">#&gt;  [99] tibble_3.1.8                  bslib_0.4.1                  </span></span>
<span><span class="co">#&gt; [101] stringi_1.7.8                 desc_1.4.2                   </span></span>
<span><span class="co">#&gt; [103] lattice_0.20-45               Matrix_1.5-1                 </span></span>
<span><span class="co">#&gt; [105] vctrs_0.5.0                   pillar_1.8.1                 </span></span>
<span><span class="co">#&gt; [107] lifecycle_1.0.3               rhdf5filters_1.10.0          </span></span>
<span><span class="co">#&gt; [109] BiocManager_1.30.19           jquerylib_0.1.4              </span></span>
<span><span class="co">#&gt; [111] data.table_1.14.4             bitops_1.0-7                 </span></span>
<span><span class="co">#&gt; [113] seriation_1.4.0               httpuv_1.6.6                 </span></span>
<span><span class="co">#&gt; [115] rtracklayer_1.58.0            GenomicRanges_1.50.1         </span></span>
<span><span class="co">#&gt; [117] R6_2.5.1                      BiocIO_1.8.0                 </span></span>
<span><span class="co">#&gt; [119] TSP_1.2-1                     promises_1.2.0.1             </span></span>
<span><span class="co">#&gt; [121] gridExtra_2.3                 IRanges_2.32.0               </span></span>
<span><span class="co">#&gt; [123] codetools_0.2-18              assertthat_0.2.1             </span></span>
<span><span class="co">#&gt; [125] rhdf5_2.42.0                  SummarizedExperiment_1.28.0  </span></span>
<span><span class="co">#&gt; [127] rprojroot_2.0.3               rjson_0.2.21                 </span></span>
<span><span class="co">#&gt; [129] withr_2.5.0                   shinyWidgets_0.7.4           </span></span>
<span><span class="co">#&gt; [131] GenomicAlignments_1.34.0      Rsamtools_2.14.0             </span></span>
<span><span class="co">#&gt; [133] S4Vectors_0.36.0              GenomeInfoDbData_1.2.9       </span></span>
<span><span class="co">#&gt; [135] hms_1.1.2                     parallel_4.2.2               </span></span>
<span><span class="co">#&gt; [137] fst_0.9.8                     grid_4.2.2                   </span></span>
<span><span class="co">#&gt; [139] tidyr_1.2.1                   rmarkdown_2.18               </span></span>
<span><span class="co">#&gt; [141] DelayedMatrixStats_1.20.0     MatrixGenerics_1.10.0        </span></span>
<span><span class="co">#&gt; [143] Biobase_2.58.0                shiny_1.7.3                  </span></span>
<span><span class="co">#&gt; [145] restfulr_0.0.15</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alex Chit Hei Wong.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
