<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SpliceWiz: Quick Start • SpliceWiz</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SpliceWiz: Quick Start">
<meta property="og:description" content="SpliceWiz">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SpliceWiz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SW_Cookbook.html">SpliceWiz: the cookbook</a>
    </li>
    <li>
      <a href="../articles/SW_QuickStart.html">SpliceWiz: Quick Start</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/alexchwong/SpliceWiz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SpliceWiz: Quick Start</h1>
                        <h4 data-toc-skip class="author">Alex CH
Wong</h4>
            
            <h4 data-toc-skip class="date">11/15/2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/alexchwong/SpliceWiz/blob/HEAD/vignettes/SW_QuickStart.Rmd" class="external-link"><code>vignettes/SW_QuickStart.Rmd</code></a></small>
      <div class="hidden name"><code>SW_QuickStart.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      This Quick-Start is a runnable example showing the functionalities
      of the SpliceWiz workflow. Version 1.1.2
    </div>
    
<p><br></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>SpliceWiz is a graphical interface for differential alternative
splicing and visualization in R. It differs from other alternative
splicing tools as it is designed for users with basic bioinformatic
skills to analyze datasets containing up to hundreds of samples!
SpliceWiz contains a number of innovations including:</p>
<ul>
<li>Super-fast handling of alignment BAM files using ompBAM, our
developer resource for multi-threaded BAM processing,</li>
<li>Alternative splicing event filters, designed to remove unreliable
measurements prior to differential analysis, which improves accuracy of
reported results.</li>
<li>Group-averaged coverage plots: publication-ready figures to clearly
visualize differential alternative splicing between biological /
experimental conditions</li>
<li>Seamless storage and recall of sequencing coverage, using the COV
format that stores strand-specific coverage typical of current RNA-seq
protocols</li>
<li>Interactive figures, including scatter and volcano plots, heatmaps,
and scrollable coverage plots, powered using the shinyDashboard
interface</li>
</ul>
<p>This vignette is a runnable working example of the SpliceWiz
workflow. The purpose is to quickly demonstrate the basic
functionalities of SpliceWiz.</p>
<p>We provide here a brief outline of the workflow for users to get
started as quickly as possible. However, we also provide more details
for those wishing to know more. Many sections will contain extra
information that can be displayed when clicked on, such as these:</p>
<details><summary>
Click on me for more details
</summary><p><br>
In most sections, we offer more details about each step of the workflow,
that can be revealed in text segments like this one. Be sure to click on
buttons like these, where available.</p>
</details><p><br></p>
</div>
<div class="section level2">
<h2 id="whats-new">What’s New?<a class="anchor" aria-label="anchor" href="#whats-new"></a>
</h2>
<details><summary>
What’s New: Novel splice detection (version 0.99.3+)
</summary><p><br>
In version 0.99.3, SpliceWiz offers detection of novel events in
addition to annotated events. How this works:</p>
<ul>
<li>SpliceWiz compiles counts of all junctions (split reads) compatible
with splice junction events (having compatible donor / acceptor splice
motifs)</li>
<li>Additionally, it also uses “tandem junction” to identify novel exons
(tandem junctions are reads that are split into 3 or more segments,
arising from splicing of 2+ consecutive introns from a short-read
sequence).</li>
</ul>
<p>To reduce false positives in novel splicing detection, SpliceWiz
provides several filters to reduce the number of novel junctions fed
into the analysis:</p>
<ul>
<li>Novel junctions that are lowly expressed (only in a small number of
samples) are removed. The minimum number of samples required to retain a
novel junction is set using <code>novelSplicing_minSamples</code>
parameter</li>
<li>Alternately, junctions are retained if its expression exceeds a
certain threshold (set using <code>novelSplicing_countThreshold</code>)
in a smaller number of samples (set using
<code>novelSplicing_minSamplesAboveThreshold</code>)</li>
<li>Further, novel junctions can be filtered by requiring at least one
end to be an annotated splice site (this is enabled using
<code>novelSplicing_requireOneAnnotatedSJ = TRUE</code>)</li>
</ul>
<p>Novel ASE detection is integrated into the SpliceWiz pipeline at the
collation step. After compilation and processing of novel junctions /
TJ’s, the novel transcripts are appended to the transcript annotation,
which is then used to re-construct the SpliceWiz reference. This
reference is contained in the “Reference” subfolder of the output folder
of <code><a href="../reference/collateData.html">collateData()</a></code> function.</p>
<p><strong>TL/DR - how to enable novel ASE mode</strong></p>
<ul>
<li>To enable analysis involving both annotated and novel ASEs, simply
set <code>novelSplicing = TRUE</code> when running
<code><a href="../reference/collateData.html">collateData()</a></code>. For example:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Usual pipeline:</span></span>
<span><span class="va">ref_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"Reference"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Build-Reference-methods.html">buildRef</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    fasta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/NxtIRFdata/man/NxtIRFdata-package.html" class="external-link">chrZ_genome</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    gtf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/NxtIRFdata/man/NxtIRFdata-package.html" class="external-link">chrZ_gtf</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">pb_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pb_output"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/processBAM.html">processBAM</a></span><span class="op">(</span></span>
<span>    bamfiles <span class="op">=</span> <span class="va">bams</span><span class="op">$</span><span class="va">path</span>,</span>
<span>    sample_names <span class="op">=</span> <span class="va">bams</span><span class="op">$</span><span class="va">sample</span>,</span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    output_path <span class="op">=</span> <span class="va">pb_path</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Modified pipeline - collateData with novel ASE discovery:</span></span>
<span></span>
<span><span class="va">nxtse_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"NxtSE_output"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/collateData.html">collateData</a></span><span class="op">(</span></span>
<span>    Experiment <span class="op">=</span> <span class="va">expr</span>,</span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    output_path <span class="op">=</span> <span class="va">nxtse_path</span>,</span>
<span>    </span>
<span>        <span class="co">## NEW ##</span></span>
<span>    novelSplicing <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        <span class="co"># switches on novel splice detection</span></span>
<span>    </span>
<span>    novelSplicing_requireOneAnnotatedSJ <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        <span class="co"># novel junctions must share one annotated splice site</span></span>
<span></span>
<span>    novelSplicing_minSamples <span class="op">=</span> <span class="fl">3</span>,</span>
<span>        <span class="co"># retain junctions observed in 3+ samples (of any non-zero expression)</span></span>
<span>    </span>
<span>    novelSplicing_minSamplesAboveThreshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        <span class="co"># only 1 sample required if its junction count exceeds a set threshold</span></span>
<span>    novelSplicing_countThreshold <span class="op">=</span> <span class="fl">10</span>  </span>
<span>        <span class="co"># threshold for previous parameter</span></span>
<span><span class="op">)</span></span></code></pre></div>
</details><p><br></p>
<details><summary>
What’s New: Visualising junction reads in coverage plots (version
0.99.5+)
</summary><p><br>
In version 0.99.5, SpliceWiz visualises split/junction reads in
individual samples and in sample groups</p>
<p>For individual sample coverage plots (i.e. when
<code>condition</code> is not set), junction counts for each sample are
plotted. Samples with low junction counts (less than 0.01x of the track
height) are omitted to reduce clutter.</p>
<p>For group-normalized coverage plots (where coverage of multiple
samples in a condition group are combined), junctions are instead
labeled by their “provisional PSIs”. These PSIs are calculated per
junction (instead of per ASE). This is done by determining the ratio of
junction counts as a proportion of all junction reads that share a
common exon cluster as the junction being assessed.</p>
<p><strong>TL/DR - how to enable junction plotting</strong></p>
<ul>
<li>To enable plotting of junctions, set
<code>plotJunctions = TRUE</code> from within
<code><a href="../reference/plotCoverage.html">plotCoverage()</a></code>
</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Retrieve example NxtSE object</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/example-SpliceWiz-data.html">SpliceWiz_example_NxtSE</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign annotation of the experimental conditions</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Return a list of ggplot and plotly objects, also plotting junction counts</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCoverage.html">plotCoverage</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se</span>,</span>
<span>    Event <span class="op">=</span> <span class="st">"SE:SRSF3-203-exon4;SRSF3-202-int3"</span>,</span>
<span>    tracks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, </span>
<span>    </span>
<span>        <span class="co">## NEW ##</span></span>
<span>    plotJunctions <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in geom_line(data = dfJn, aes_string(x = "x", y = "yarc", group = "junction", : Ignoring unknown aesthetics: <span style="color: #00BB00;">label</span></span></span>
<span><span class="co">#&gt; Ignoring unknown aesthetics: <span style="color: #00BB00;">label</span></span></span>
<span><span class="co">#&gt; Ignoring unknown aesthetics: <span style="color: #00BB00;">label</span></span></span>
<span><span class="co">#&gt; Ignoring unknown aesthetics: <span style="color: #00BB00;">label</span></span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as plotly object</span></span>
<span>    <span class="va">p</span><span class="op">$</span><span class="va">final_plot</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as ggplot</span></span>
<span>    <span class="fu"><a href="../reference/plotCoverage.html">as_ggplot_cov</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot by condition "treatment", including provisional PSIs</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCoverage.html">plotCoverage</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se</span>,</span>
<span>    Event <span class="op">=</span> <span class="st">"SE:SRSF3-203-exon4;SRSF3-202-int3"</span>,</span>
<span>    tracks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"treatment"</span>, </span>
<span></span>
<span>        <span class="co">## NEW ##</span></span>
<span>    plotJunctions <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in geom_line(data = dfJn, aes_string(x = "x", y = "yarc", group = "junction", : Ignoring unknown aesthetics: <span style="color: #00BB00;">label</span></span></span>
<span><span class="co">#&gt; Ignoring unknown aesthetics: <span style="color: #00BB00;">label</span></span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as plotly object</span></span>
<span>    <span class="va">p</span><span class="op">$</span><span class="va">final_plot</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as ggplot</span></span>
<span>    <span class="fu"><a href="../reference/plotCoverage.html">as_ggplot_cov</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</details><p><br></p>
</div>
<div class="section level2">
<h2 id="workflow-from-a-glance">Workflow from a glance<a class="anchor" aria-label="anchor" href="#workflow-from-a-glance"></a>
</h2>
<p>The basic steps of SpliceWiz are as follows:</p>
<ul>
<li>Building the SpliceWiz reference</li>
<li>Process BAM files using SpliceWiz</li>
<li>Collate results of individual samples into an experiment</li>
<li>Importing the collated experiment as an NxtSE object</li>
<li>Alternative splicing event filtering</li>
<li>Differential ASE analysis</li>
<li>Visualization<br><br>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="quick-start">Quick-Start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>To install SpliceWiz, start R (version “4.2.1”) and enter:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"SpliceWiz"</span><span class="op">)</span></span></code></pre></div>
<details><summary>
Setting up a Conda environment to use SpliceWiz
</summary><p><br>
For those wishing to set up a self-contained environment with SpliceWiz
installed (e.g. on a high performance cluster), we recommend using
miniconda. For installation instructions, see the <a href="https://docs.conda.io/en/latest/miniconda.html" class="external-link">documentation on
how to install miniconda</a></p>
<p>After installing miniconda, create a conda environment as
follows:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-n</span> envSpliceWiz python=3.9</span></code></pre></div>
<p>After following the prompts, activate the environment:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate envSpliceWiz</span></code></pre></div>
<p>Next, install R 4.2.1 as follows:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> conda-forge r-base=4.2.1</span></code></pre></div>
<p>Many of SpliceWiz’s dependencies are up-to-date from the conda-forge
channel, so they are best installed via conda:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> conda-forge r-devtools r-essentials r-xml r-biocmanager <span class="dt">\</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  r-fst r-plotly r-rsqlite r-rcurl</span></code></pre></div>
<p>After this is done, the remainder of the packages need to be
installed from the R terminal. This is because most Bioconductor
packages are from the bioconda channel and depend on Bioc 3.14 (which is
not the latest version). So, lets enter the R terminal from the command
line:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div>
<p>Set up Bioconductor devel (soon to be Bioc 3.16):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"devel"</span><span class="op">)</span></span></code></pre></div>
<p>Again, follow the prompts to update any necessary packages.</p>
<p>Once this is done, install SpliceWiz as per usual.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"SpliceWiz"</span><span class="op">)</span></span></code></pre></div>
<p>The last step will install any remaining dependencies, taking
approximately 20-30 minutes depending on your system.</p>
</details><p><br></p>
<details><summary>
Enabling OpenMP (multi-threading) for MacOS users (Optional)
</summary><p><br>
For <strong>MacOS</strong> users, make sure OpenMP libraries are
installed correctly. We recommend users follow this <a href="https://mac.r-project.org/openmp/" class="external-link">guide</a>, but the quickest way
to get started is to install <code>libomp</code> via brew:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">brew</span> install libomp</span></code></pre></div>
</details><p><br></p>
<details><summary>
Installing statistical package dependencies (Optional)
</summary><p><br>
SpliceWiz uses established statistical tools to perform alternative
splicing differential analysis:</p>
<ul>
<li>limma: models included and excluded counts as log-normal
distributions</li>
<li>DESeq2: models included and excluded counts as negative binomial
distributions</li>
<li>DoubleExpSeq: models included and excluded counts using beta
binomial distributions</li>
<li>satuRn: models included and excluded counts using quasi-binomial
distributions.</li>
</ul>
<p>To install all of these packages:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"DoubleExpSeq"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DESeq2"</span>, <span class="st">"limma"</span>, <span class="st">"satuRn"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>SpliceWiz’s satuRn wrapper implements counts filtering using edgeR’s
<code>filterByExpr</code> method, which requires edgeR:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"edgeR"</span><span class="op">)</span></span></code></pre></div>
</details><p><br></p>
</div>
<div class="section level3">
<h3 id="loading-splicewiz">Loading SpliceWiz<a class="anchor" aria-label="anchor" href="#loading-splicewiz"></a>
</h3>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/alexchwong/SpliceWiz" class="external-link">SpliceWiz</a></span><span class="op">)</span></span></code></pre></div>
<details><summary>
Details
</summary><p><br>
The SpliceWiz package loads the <code>NxtIRFdata</code> data package.
This data package contains the example “chrZ” genome / annotations and 6
example BAM files that are used in this working example. Also,
NxtIRFdata provides pre-generated mappability exclusion annotations for
building human and mouse SpliceWiz references</p>
</details><p><br></p>
</div>
<div class="section level3">
<h3 id="the-splicewiz-graphics-user-interface-gui">The SpliceWiz Graphics User Interface (GUI)<a class="anchor" aria-label="anchor" href="#the-splicewiz-graphics-user-interface-gui"></a>
</h3>
<p>SpliceWiz offers a graphical user interface (GUI) for interactive
users, e.g. in the RStudio environment. To start using SpliceWiz
GUI:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/Graphics-User-Interface.html">spliceWiz</a></span><span class="op">(</span>demo <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><br>
Note that the GUI demo mode is not supported on Bioconductor 3.13 or
below.<br></p>
<div class="section level4">
<h4 id="navigating-the-gui">Navigating the GUI<a class="anchor" aria-label="anchor" href="#navigating-the-gui"></a>
</h4>
<p>The SpliceWiz GUI uses the <code>shinyDashboard</code> interface. Use
the side menu on the left hand side of the interface to navigate across
the various sub-panels in the SpliceWiz GUI.</p>
<div class="figure" style="text-align: center">
<img src="img/MenuBar.jpg" alt="Menu Side Bar" width="231pt"><p class="caption">
Menu Side Bar
</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="building-the-splicewiz-reference">Building the SpliceWiz reference<a class="anchor" aria-label="anchor" href="#building-the-splicewiz-reference"></a>
</h3>
<details><summary>
Why do we need the SpliceWiz reference?
</summary><p><br>
SpliceWiz first needs to generate a set of reference files. The
SpliceWiz reference is used to quantitate alternative splicing in BAM
files, as well as in downstream collation, differential analysis and
visualisation.</p>
<p>SpliceWiz generates a reference from a user-provided genome FASTA and
genome annotation GTF file, and is optimised for Ensembl references but
can accept other reference GTF files. Alternatively, SpliceWiz accepts
AnnotationHub resources, using the record names of AnnotationHub records
as input.</p>
</details><p><br></p>
<p>Using the example FASTA and GTF files, use the
<code><a href="../reference/Build-Reference-methods.html">buildRef()</a></code> function to build the SpliceWiz reference:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"Reference"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Build-Reference-methods.html">buildRef</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    fasta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/NxtIRFdata/man/NxtIRFdata-package.html" class="external-link">chrZ_genome</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    gtf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/NxtIRFdata/man/NxtIRFdata-package.html" class="external-link">chrZ_gtf</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>(NEW in version &gt;= 0.99.3) The SpliceWiz reference can be viewed
as data frames using various getter functions. For example, to view the
annotated alternative splicing events (ASE):</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/View-Reference-methods.html">viewASE</a></span><span class="op">(</span><span class="va">ref_path</span><span class="op">)</span></span></code></pre></div>
<p>See <code>?View-Reference-methods</code> for a comprehensive list of
getter functions</p>
<details><summary>
Using the GUI
</summary><p><br>
After starting the SpliceWiz GUI in demo mode, click the
<code>Reference</code> tab from the menu side bar. The following
interface will be shown:</p>
<div class="figure" style="text-align: center">
<img src="img/buildRef.jpg" alt="Reference GUI" width="800pt"><p class="caption">
Reference GUI
</p>
</div>
<ol style="list-style-type: decimal">
<li><p>The first step to building a SpliceWiz reference is to select a
directory in which to create the reference.</p></li>
<li><p>SpliceWiz provides an interface to retrieve the genome sequence
(FASTA) and transcriptome annotation (GTF) files from the Ensembl FTP
server, by first selecting the “Release” and then “Species” from the
drop-down boxes.</p></li>
<li><p>Alternatively, users can provide their own FASTA and GTF
files.</p></li>
<li><p>Human (hg38, hg19) and mouse genomes (mm10, mm9) have the option
of further refining IR analysis using built-in mappability exclusion
annotations, allowing SpliceWiz to ignore intronic regions of low
mappability.</p></li>
</ol>
<p>For now, to continue with the demo and create the reference using the
GUI, click on the <code>Load Demo FASTA/GTF</code> (5), and then click
<code>Build Reference</code> (6)</p>
</details><p><br></p>
<details><summary>
Where did the FASTA and GTF files come from?
</summary><p><br>
The helper functions <code><a href="https://rdrr.io/pkg/NxtIRFdata/man/NxtIRFdata-package.html" class="external-link">chrZ_genome()</a></code> and
<code><a href="https://rdrr.io/pkg/NxtIRFdata/man/NxtIRFdata-package.html" class="external-link">chrZ_gtf()</a></code> returns the paths to the example genome (FASTA)
and transcriptome (GTF) file included with the <code>NxtIRFdata</code>
package that contains the working example used by SpliceWiz:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Provides the path to the example genome:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/NxtIRFdata/man/NxtIRFdata-package.html" class="external-link">chrZ_genome</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/Library/Frameworks/R.framework/Versions/4.2/Resources/library/NxtIRFdata/extdata/genome.fa"</span></span>
<span></span>
<span><span class="co"># Provides the path to the example gene annotation:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/NxtIRFdata/man/NxtIRFdata-package.html" class="external-link">chrZ_gtf</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/Library/Frameworks/R.framework/Versions/4.2/Resources/library/NxtIRFdata/extdata/transcripts.gtf"</span></span></code></pre></div>
</details><p><br></p>
<details><summary>
What is the chrZ genome?
</summary><p><br>
For the purpose of generating a running example to demonstrate
SpliceWiz, we created an artificial genome / gene annotation. This was
created using 7 human genes (SRSF1, SRSF2, SRSF3, TRA2A, TRA2B, TP53 and
NSUN5). The SRSF and TRA family of genes all contain poison exons
flanked by retained introns. Additionally, NSUN5 contains an annotated
IR event in its terminal intron. Sequences from these 7 genes were
aligned into one sequence to create an artificial chromosome Z (chrZ).
The gene annotations were modified to only contain the 7 genes with the
modified genomic coordinates.</p>
</details><p><br></p>
<details><summary>
What is Mappability and why should I care about it?
</summary><p><br>
For the most part, the SpliceWiz reference can be built with just the
FASTA and GTF files. This is sufficient for assessment for most forms of
alternative splicing events.</p>
<p>For intron retention, accurate assessment of intron depth is
important. However, introns contain many repetitive regions that are
difficult to map. We refer to these regions as “mappability
exclusions”.</p>
<p>We adopt IRFinder’s algorithm to identify these mappability
exclusions. This is determined empirically by generating synthetic reads
systematically from the genome, then aligning these reads back to the
same genome. Regions that contain less than the expected coverage depth
of reads define “mappability exclusions”.</p>
<p>See the vignette: SpliceWiz cookbook for details on how to generate
“mappability exclusions” for any genome.<br></p>
</details><p><br></p>
<details><summary>
How do I use pre-built mappability exclusions to generate human and
mouse references?
</summary><p><br>
For human and mouse genomes, SpliceWiz provides pre-built mappability
exclusion references that can be used to build the SpliceWiz reference.
SpliceWiz provides these annotations via the <code>NxtIRFdata</code>
package.</p>
<p>Simply specify the genome in the parameter <code>genome_type</code>
in the <code><a href="../reference/Build-Reference-methods.html">buildRef()</a></code> function (which accepts
<code>hg38</code>, <code>hg19</code>, <code>mm10</code> and
<code>mm9</code>).</p>
<p>Additionally, a reference for non-polyadenylated transcripts is used.
This has a minor role in QC of samples (to assess the adequacy of polyA
capture).</p>
<p>For example, assuming your genome file <code>"genome.fa"</code> and a
transcript annotation <code>"transcripts.gtf"</code> are in the working
directory, a SpliceWiz reference can be built using the built-in
<code>hg38</code> low mappability regions and non-polyadenylated
transcripts as follows:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref_path_hg38</span> <span class="op">&lt;-</span> <span class="st">"./Reference"</span></span>
<span><span class="fu"><a href="../reference/Build-Reference-methods.html">buildRef</a></span><span class="op">(</span></span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path_hg38</span>,</span>
<span>    fasta <span class="op">=</span> <span class="st">"genome.fa"</span>,</span>
<span>    gtf <span class="op">=</span> <span class="st">"transcripts.gtf"</span>,</span>
<span>    genome_type <span class="op">=</span> <span class="st">"hg38"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</details><p><br></p>
</div>
<div class="section level3">
<h3 id="process-bam-files-using-splicewiz">Process BAM files using SpliceWiz<a class="anchor" aria-label="anchor" href="#process-bam-files-using-splicewiz"></a>
</h3>
<p>The function <code><a href="../reference/example-SpliceWiz-data.html">SpliceWiz_example_bams()</a></code> retrieves 6
example BAM files from ExperimentHub and places a copy of these in the
temporary directory.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bams</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/example-SpliceWiz-data.html">SpliceWiz_example_bams</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<details><summary>
What are these example BAM files and how were they generated?
</summary><p><br>
In this vignette, we provide 6 example BAM files. These were generated
based on aligned RNA-seq BAMs of 6 samples from the Leucegene AML
dataset (GSE67039). Sequences aligned to hg38 were filtered to only
include genes aligned to that used to create the chrZ chromosome. These
sequences were then re-aligned to the chrZ reference using STAR.<br></p>
</details><p><br></p>
<details><summary>
How can I easily locate multiple BAM files?
</summary><p><br>
Often, alignment pipelines process multiple samples. SpliceWiz provides
convenience functions to recursively locate all the BAM files in a given
folder, and tries to ascertain sample names. Often sample names can be
gleaned when: * The BAM files are named by their sample names, e.g.
“sample1.bam”, “sample2.bam”. In this case, <code>level = 0</code> * The
BAM files have generic names but are contained inside parent directories
labeled by their sample names, e.g. “sample1/Unsorted.bam”,
“sample2/Unsorted.bam”. In this case, <code>level = 1</code></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># as BAM file names denote their sample names</span></span>
<span><span class="va">bams</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findSamples.html">findBAMS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, level <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># In the case where BAM files are labelled using sample names as parent </span></span>
<span><span class="co"># directory names (which oftens happens with the STAR aligner), use level = 1</span></span></code></pre></div>
</details><p><br></p>
<p>Process these BAM files using SpliceWiz:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pb_output"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/processBAM.html">processBAM</a></span><span class="op">(</span></span>
<span>    bamfiles <span class="op">=</span> <span class="va">bams</span><span class="op">$</span><span class="va">path</span>,</span>
<span>    sample_names <span class="op">=</span> <span class="va">bams</span><span class="op">$</span><span class="va">sample</span>,</span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    output_path <span class="op">=</span> <span class="va">pb_path</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<details><summary>
Using the GUI
</summary><p><br>
After building the demo reference as shown in the previous section,
start SpliceWiz GUI in demo mode. Then, click the
<code>Experiment</code> tab from the menu side bar. The following
interface will be shown:</p>
<div class="figure" style="text-align: center">
<img src="img/Expr_empty.jpg" alt="Experiment GUI" width="8000pt"><p class="caption">
Experiment GUI
</p>
</div>
<p>Click on (2) <code>Define Project Folders</code> to bring up the
following drop-down box:</p>
<div class="figure" style="text-align: center">
<img src="img/Expr_drop_2.jpg" alt="Define Project Folders" width="480pt"><p class="caption">
Define Project Folders
</p>
</div>
<p>We need to define the folders that contain our reference, BAM files,
as well as processBAM output files and the final compiled experiment
(that generates the NxtSE object).</p>
<ul>
<li>Click on (1) <code>Choose Folder (Reference)</code> and select the
<code>Reference</code> directory (where the SpliceWiz reference was
generated by the previous step. Then,</li>
<li>Click on (2) <code>Choose Folder (BAM files)</code> and select the
<code>bams</code> directory (where the demo BAM files have been
generated).</li>
<li>Click on (3) <code>Choose Folder (processBAM output)</code> and
select the <code>pb_output</code> directory (which should currently be
empty).</li>
<li>Finally, click on (4) <code>Choose Folder (NxtSE files)</code> and
select the <code>NxtSE</code> directory(which should currently be
empty).</li>
</ul>
<p>After our folders have been defined, on the right hand side, an
interactive table should be displayed that looks like the following:</p>
<div class="figure" style="text-align: center">
<img src="img/Expr_pb.jpg" alt="Running processBAM" width="640pt"><p class="caption">
Running processBAM
</p>
</div>
<p>To run processBAM on the example BAM files, first drag to select the
6 cells containing the BAM file paths (as shown). Next, click the
<code>Process BAM files</code> to open the dropdown menu, and then click
<code>Run processBAM()</code>. A prompt should then be displayed asking
whether you wish to proceed. Click <code>OK</code> to start running
processBAM.</p>
</details><p><br></p>
<details><summary>
What is the <code>processBAM()</code> function
</summary><p><br>
SpliceWiz’s <code><a href="../reference/processBAM.html">processBAM()</a></code> function can process one or more
BAM files. This function is ultra-fast, relying on an internal native
C++ function that uses OpenMP multi-threading (via the
<code>ompBAM</code> C++ API).</p>
<p>Input BAM files can be either read-name sorted or coordinate sorted
(although SpliceWiz prefers the former). Indexing of coordinate-sorted
BAMs are not necessary.</p>
<p><code><a href="../reference/processBAM.html">processBAM()</a></code> loads the SpliceWiz reference. Then, it
reads each BAM file in their entirety, and quantifies the following:</p>
<ul>
<li>Basic QC parameters including number of reads, directionality,
etc</li>
<li>Counts of gapped (junction) reads / fragments</li>
<li>Intron coverage depths and other parameters (identical output to
IRFinder)</li>
<li>COV files (which are like BigWig files but record strand-specific
coverage)</li>
<li>Miscellaneous quants including coverage of chromosomes, intergenic
regions, rRNAs, and non-polyadenylated regions</li>
</ul>
<p>For each BAM file, <code><a href="../reference/processBAM.html">processBAM()</a></code> generates two output
files. The first is a gzipped text file containing all the quantitation
data. The second is a <code>COV</code> file which contains the
per-nucleotide RNA-seq coverage of the sample.<br></p>
</details><p><br></p>
<details><summary>
More details on the <code>processBAM()</code> function
</summary><p><br>
At minimum, <code><a href="../reference/processBAM.html">processBAM()</a></code> requires four parameters:</p>
<ul>
<li>
<code>bamfiles</code> : The paths of the BAM files</li>
<li>
<code>sample_names</code> : The sample names corresponding to the
given BAM files</li>
<li>
<code>reference_path</code> : The directory containing the SpliceWiz
reference</li>
<li>
<code>output_path</code> : The directory where the output of
<code><a href="../reference/processBAM.html">processBAM()</a></code> should go</li>
</ul>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pb_output"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/processBAM.html">processBAM</a></span><span class="op">(</span></span>
<span>    bamfiles <span class="op">=</span> <span class="va">bams</span><span class="op">$</span><span class="va">path</span>,</span>
<span>    sample_names <span class="op">=</span> <span class="va">bams</span><span class="op">$</span><span class="va">sample</span>,</span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    output_path <span class="op">=</span> <span class="va">pb_path</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/processBAM.html">processBAM()</a></code> also takes several optional, but useful,
parameters:</p>
<ul>
<li>
<code>n_threads</code> : The number of threads for
multi-threading</li>
<li>
<code>overwrite</code> : Whether existing files in the output
directory should be overwritten</li>
<li>
<code>run_featureCounts</code> : (Requires the Rsubread package)
runs featureCounts to obtain gene counts (which outputs results as an
RDS file)</li>
</ul>
<p>For example, to run <code><a href="../reference/processBAM.html">processBAM()</a></code> using 2 threads,
disallow overwrite of existing <code><a href="../reference/processBAM.html">processBAM()</a></code> outputs, and
run featureCounts afterwards, one would run the following:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># NOT RUN</span></span>
<span></span>
<span><span class="co"># Re-run IRFinder without overwrite, and run featureCounts</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">Rsubread</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/processBAM.html">processBAM</a></span><span class="op">(</span></span>
<span>    bamfiles <span class="op">=</span> <span class="va">bams</span><span class="op">$</span><span class="va">path</span>,</span>
<span>    sample_names <span class="op">=</span> <span class="va">bams</span><span class="op">$</span><span class="va">sample</span>,</span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    output_path <span class="op">=</span> <span class="va">pb_path</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    run_featureCounts <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load gene counts</span></span>
<span><span class="va">gene_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">pb_path</span>, <span class="st">"main.FC.Rds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Access gene counts:</span></span>
<span><span class="va">gene_counts</span><span class="op">$</span><span class="va">counts</span></span></code></pre></div>
</details><p><br></p>
</div>
<div class="section level3">
<h3 id="collate-the-experiment">Collate the experiment<a class="anchor" aria-label="anchor" href="#collate-the-experiment"></a>
</h3>
<p>The helper function <code><a href="../reference/findSamples.html">findSpliceWizOutput()</a></code> organises the
output files of SpliceWiz’s <code><a href="../reference/processBAM.html">processBAM()</a></code> function. It
identifies matching <code>"txt.gz"</code> and <code>"cov"</code> files
for each sample, and organises these file paths conveniently into a
3-column data frame:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findSamples.html">findSpliceWizOutput</a></span><span class="op">(</span><span class="va">pb_path</span><span class="op">)</span></span></code></pre></div>
<p>Using this data frame, collate the experiment using
<code><a href="../reference/collateData.html">collateData()</a></code>. We name the output directory as
<code>NxtSE_output</code> as this folder will contain the data needed to
import the NxtSE object:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nxtse_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"NxtSE_output"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/collateData.html">collateData</a></span><span class="op">(</span></span>
<span>    Experiment <span class="op">=</span> <span class="va">expr</span>,</span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    output_path <span class="op">=</span> <span class="va">nxtse_path</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<details><summary>
What is the <code>collateData()</code> function
</summary><p><br><code><a href="../reference/collateData.html">collateData()</a></code> combines the <code><a href="../reference/processBAM.html">processBAM()</a></code> output
files of multiple samples and builds a single database.
<code><a href="../reference/collateData.html">collateData()</a></code> creates a number of files in the chosen
output directory. These outputs can then be imported into the R session
as a <code>NxtSE</code> data object for downstream analysis.</p>
<p>At minimum, <code><a href="../reference/collateData.html">collateData()</a></code> takes the following
parameters:</p>
<ul>
<li>
<code>Experiment</code> : The 2- or 3- column data frame. The first
column should contain (unique) sample names. The second and (optional)
third columns contain the <code>"txt.gz"</code> and <code>"cov"</code>
file paths</li>
<li>
<code>reference_path</code> : The directory containing the SpliceWiz
reference</li>
<li>
<code>output_path</code> : The directory where the output of
<code><a href="../reference/processBAM.html">processBAM()</a></code> should go</li>
</ul>
<p><code><a href="../reference/collateData.html">collateData()</a></code> can take some optional parameters:</p>
<ul>
<li>
<code>IRMode</code> : Whether to use SpliceWiz’s
<code>SpliceOver</code> method, or IRFinder’s <code>SpliceMax</code>
method, to determine total spliced transcript abundance. Briefly,
<code>SpliceMax</code> considers junction reads that have either
flanking splice site coordinate. <code>SpliceOver</code> considers
additional junction reads that splices across exon clusters in common.
Exon clusters are groups of mutually-overlapping exons.
<code>SpliceOver</code> is the default option.</li>
<li>
<code>overwrite</code> : Whether files in the output directory
should be overwritten</li>
<li>
<code>n_threads</code> : Use multi-threaded operations where
possible</li>
<li>
<code>lowMemoryMode</code> : Minimise memory usage where possible.
Note that most of the collateData pipeline will be single-threaded if
this is set to <code>TRUE</code>.</li>
</ul>
<p><code><a href="../reference/collateData.html">collateData()</a></code> is a memory-intensive operation when run
using multiple threads. We estimate it can use up to 6-7 Gb per thread.
<code>lowMemoryMode</code> will minimise RAM usage to ~ 8 Gb, but will
be slower and run on a single thread.<br></p>
</details><p><br></p>
<details><summary>
Enabling novel splicing detection
</summary><p><br>
Novel splicing detection can be switched on by setting
<code>novelSplicing = TRUE</code> from within the
<code><a href="../reference/collateData.html">collateData()</a></code> function:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Modified pipeline - collateData with novel ASE discovery:</span></span>
<span></span>
<span><span class="va">nxtse_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"NxtSE_output_novel"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/collateData.html">collateData</a></span><span class="op">(</span></span>
<span>    Experiment <span class="op">=</span> <span class="va">expr</span>,</span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    output_path <span class="op">=</span> <span class="va">nxtse_path</span>,</span>
<span>    novelSplicing <span class="op">=</span> <span class="cn">TRUE</span>     <span class="co">## NEW ##</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/collateData.html">collateData()</a></code> uses split reads that are not annotated
introns to help construct hypothetical minimal transcripts. These are
then injected into the original transcriptome annotation (GTF) file,
whereby the SpliceWiz reference is rebuilt. The new SpliceWiz reference
(which contains these novel transcripts) is then used to collate the
samples.</p>
<p>To reduce false positives in novel splicing detection, SpliceWiz
provides several filters to reduce the number of novel junctions fed
into the analysis:</p>
<ul>
<li>Novel junctions that are lowly expressed (only in a small number of
samples) are removed. The minimum number of samples required to retain a
novel junction is set using <code>novelSplicing_minSamples</code>
parameter</li>
<li>Alternately, junctions are retained if its expression exceeds a
certain threshold (set using <code>novelSplicing_countThreshold</code>)
in a smaller number of samples (set using
<code>novelSplicing_minSamplesAboveThreshold</code>)</li>
<li>Further, novel junctions can be filtered by requiring at least one
end to be an annotated splice site (this is enabled using
<code>novelSplicing_requireOneAnnotatedSJ = TRUE</code>).</li>
</ul>
<p>For example, if one wished to retain novel reads seen in 3 or more
samples, or novel spliced reads with 10 or more counts in at least 1
sample, and requiring at least one end of a novel junction being an
annotated splice site:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nxtse_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"NxtSE_output_novel"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/collateData.html">collateData</a></span><span class="op">(</span></span>
<span>    Experiment <span class="op">=</span> <span class="va">expr</span>,</span>
<span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span>
<span>    output_path <span class="op">=</span> <span class="va">nxtse_path</span>,</span>
<span>    </span>
<span>        <span class="co">## NEW ##</span></span>
<span>    novelSplicing <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        <span class="co"># switches on novel splice detection</span></span>
<span>    </span>
<span>    novelSplicing_requireOneAnnotatedSJ <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        <span class="co"># novel junctions must share one annotated splice site</span></span>
<span></span>
<span>    novelSplicing_minSamples <span class="op">=</span> <span class="fl">3</span>,</span>
<span>        <span class="co"># retain junctions observed in 3+ samples (of any non-zero expression)</span></span>
<span>    </span>
<span>    novelSplicing_minSamplesAboveThreshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        <span class="co"># only 1 sample required if its junction count exceeds a set threshold</span></span>
<span>    novelSplicing_countThreshold <span class="op">=</span> <span class="fl">10</span>  </span>
<span>        <span class="co"># threshold for previous parameter</span></span>
<span><span class="op">)</span></span></code></pre></div>
</details><p><br></p>
<details><summary>
Using the GUI
</summary><p><br>
After running the Reference and processBAM steps as indicated in the
previous sections (of the GUI instructions), click
<code>Construct Experiment</code>. Your interface should now look like
this:</p>
<div class="figure" style="text-align: center">
<img src="img/Expr_cd.jpg" alt="collateData GUI" width="800pt"><p class="caption">
collateData GUI
</p>
</div>
<p>This drop-down dialog box contains several parameters related to
novel splicing detection:</p>
<ol style="list-style-type: decimal">
<li>Toggle on/off novel splice detection</li>
<li>Restrict novel junction reads to having one annotated splice
site</li>
<li>Filter novel junction counts based on any expression (min number of
samples set here)</li>
<li>Filter novel junction counts based on expression threshold (min
number of samples set here)</li>
<li>Threshold split read count (expression threshold based novel
junction filter)</li>
</ol>
<p>To collate the experiment, click on (6)
<code>Run collateData()</code>. An alert pop-up should be displayed when
this process is complete.</p>
<p>To clear all the setting on the Experiment page, click (7).</p>
</details><p><br></p>
</div>
<div class="section level3">
<h3 id="importing-the-experiment">Importing the experiment<a class="anchor" aria-label="anchor" href="#importing-the-experiment"></a>
</h3>
<p>Before differential analysis can be performed, the collated
experiment must be imported into the R session as an <code>NxtSE</code>
data object.</p>
<p>After running <code><a href="../reference/collateData.html">collateData()</a></code>, import the experiment using
the <code><a href="../reference/makeSE.html">makeSE()</a></code> function:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSE.html">makeSE</a></span><span class="op">(</span><span class="va">nxtse_path</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<details><summary>
Using the GUI
</summary><p><br>
After running the steps in the previous GUI sections, navigate to
<code>Analysis</code> and then click the <code>Load Experiment</code> on
the menu bar. The display should look like this:</p>
<div class="figure" style="text-align: center">
<img src="img/Expr_load_empty.jpg" alt="Loading the NxtSE GUI" width="480pt"><p class="caption">
Loading the NxtSE GUI
</p>
</div>
<p>First, click the <code>Open Folder containing NxtSE</code> to open
the dropdown, then click <code>Choose Folder (NxtSE)</code>. Select the
<code>NxtSE</code> directory. The interface should now look like
this:</p>
<div class="figure" style="text-align: center">
<img src="img/Expr_load_folder.jpg" alt="Loading the NxtSE GUI" width="800pt"><p class="caption">
Loading the NxtSE GUI
</p>
</div>
<p>To load the NxtSE object, click the <code>Load NxtSE object</code> to
open the dropdown, then click <code>Load NxtSE object</code> to load the
NxtSE object into the current session.</p>
</details><p><br></p>
<details><summary>
What is the <code>makeSE()</code> function
</summary><p><br>
The <code><a href="../reference/makeSE.html">makeSE()</a></code> function imports the compiled data generated
by the <code><a href="../reference/collateData.html">collateData()</a></code> function. Data is imported as an
<code>NxtSE</code> object. Downstream analysis, including differential
analysis and visualization, is performed using the <code>NxtSE</code>
object.<br></p>
</details><p><br></p>
<details><summary>
More details about the <code>makeSE()</code> function
</summary><p><br>
By default, <code><a href="../reference/makeSE.html">makeSE()</a></code> uses delayed operations to avoid
consuming memory until the data is actually needed. This is advantageous
in analysis of hundreds of samples on a computer with limited resources.
However, it will be slower. To load all the data into memory, we need to
“realize” the NxtSE object, as follows:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NxtSE-class.html">realize_NxtSE</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, <code><a href="../reference/makeSE.html">makeSE()</a></code> can realize the NxtSE object at
construction:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSE.html">makeSE</a></span><span class="op">(</span><span class="va">nxtse_path</span>, realize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>By default, <code><a href="../reference/makeSE.html">makeSE()</a></code> constructs the NxtSE object using
all the samples in the collated data. It is possible (and particularly
useful in large data sets) to read only a subset of samples. In this
case, construct a data frame object with the first column containing the
desired sample names and parse this into the <code>colData</code>
parameter as shown:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subset_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="va">subset_samples</span><span class="op">)</span></span>
<span><span class="va">se_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSE.html">makeSE</a></span><span class="op">(</span><span class="va">nxtse_path</span>, colData <span class="op">=</span> <span class="va">df</span>, RemoveOverlapping <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>In complex transcriptomes including those of human and mouse,
alternative splicing implies that introns are often overlapping. Thus,
algorithms run the risk of over-calling intron retention where
overlapping introns are assessed. SpliceWiz removes overlapping introns
by considering only introns belonging to the major splice isoforms. It
estimates a list of introns of major isoforms by assessing the
compatible splice junctions of each isoform, and removes overlapping
introns belonging to minor isoforms. To disable this functionality, set
<code>RemoveOverlapping = FALSE</code>.<br></p>
</details><p><br></p>
</div>
<div class="section level3">
<h3 id="differential-analysis">Differential analysis<a class="anchor" aria-label="anchor" href="#differential-analysis"></a>
</h3>
<div class="section level4">
<h4 id="assigning-annotations-to-samples">Assigning annotations to samples<a class="anchor" aria-label="anchor" href="#assigning-annotations-to-samples"></a>
</h4>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"K"</span>, <span class="st">"L"</span>, <span class="st">"M"</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<details><summary>
Using the GUI
</summary><p><br>
After following the previous GUI sections, with the NxtSE loaded, click
on <code>Review Annotations</code>. Then, click
<code>Edit Interactively</code>. Your interface should now look like
this:</p>
<div class="figure" style="text-align: center">
<img src="img/Expr_anno_empty.jpg" alt="NxtSE Annotations GUI" width="800pt"><p class="caption">
NxtSE Annotations GUI
</p>
</div>
<p>Now, lets add an extra column. In the <code>Annotation Columns</code>
box which has just appeared, type in a name <code>condition</code>, then
click <code>Add</code>. Then, click <code>batch</code> and again click
<code>Add</code>. Then, click on each empty cell and fill in the fields
as shown:</p>
<div class="figure" style="text-align: center">
<img src="img/Expr_anno_filled.jpg" alt="NxtSE Annotations GUI" width="800pt"><p class="caption">
NxtSE Annotations GUI
</p>
</div>
<p>Finally, the annotated NxtSE needs to be reloaded into memory. Click
<code>Load NxtSE object</code> and then <code>Load NxtSE object</code>
to reload the NxtSE object with the new annotations.</p>
<p>NB: Annotations can be assigned prior to loading the NxtSE object,
but can only be done after the NxtSE folder is selected.</p>
<p>Alternatively, users can provide a file that contains the annotations
in tabular format (e.g. a comma-separated values (CSV) or tab-separated
values (TSV) file). Click on <code>Review Annotations</code> and then
<code>Import Data Frame from File</code>. Then, select the
<code>demo_annotations.csv</code> file that has been pre-generated by
SpliceWiz’s demo mode.</p>
</details><p><br></p>
<details><summary>
Saving and reloading the NxtSE as an RDS file (GUI)
</summary><p><br>
Once an NxtSE object has been loaded into memory, you can save it as an
RDS object so it can be reloaded in a later session. To do this, click
the <code>Save NxtSE to/from RDS file</code>, then click
<code>Save NxtSE as RDS</code>. Choose a file name and location and
press <code>OK</code>. This RDS file can be loaded as an NxtSE object in
a later GUI session by clicking the <code>Load NxtSE from RDS</code>
button.</p>
</details><p><br></p>
<details><summary>
What is the <code>NxtSE</code> object
</summary><p><br><code>NxtSE</code> is a data object which contains all the required data
for downstream analysis after all the BAM alignment files have been
process and the experiment is collated.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span></span>
<span><span class="co">#&gt; class: NxtSE </span></span>
<span><span class="co">#&gt; dim: 168 6 </span></span>
<span><span class="co">#&gt; metadata(11): Up_Inc Down_Inc ... ref BuildVersion</span></span>
<span><span class="co">#&gt; assays(5): Included Excluded Depth Coverage minDepth</span></span>
<span><span class="co">#&gt; rownames(168): TRA2B/ENST00000453386_Intron8/clean</span></span>
<span><span class="co">#&gt;   TRA2B/ENST00000453386_Intron7/clean ...</span></span>
<span><span class="co">#&gt;   RI:SRSF2-203-exon2;SRSF2-202-intron2</span></span>
<span><span class="co">#&gt;   RI:SRSF2-203-exon2;SRSF2-206-intron1</span></span>
<span><span class="co">#&gt; rowData names(18): EventName EventType ... is_annotated_IR</span></span>
<span><span class="co">#&gt;   NMD_direction</span></span>
<span><span class="co">#&gt; colnames(6): 02H003 02H025 ... 02H043 02H046</span></span>
<span><span class="co">#&gt; colData names(2): condition batch</span></span></code></pre></div>
<p>The <code>NxtSE</code> object inherits the
<code>SummarizedExperiment</code> object. This means that the functions
for SummarizedExperiment can be used on the NxtSE object. These include
row and column annotations using the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> and
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> accessors.</p>
<p>Rows in the <code>NxtSE</code> object contain information about each
alternate splicing event. For example:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 18 columns</span></span>
<span><span class="co">#&gt;                                                  EventName   EventType</span></span>
<span><span class="co">#&gt;                                                &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron8/clean TRA2B/ENST0000045338..          IR</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron7/clean TRA2B/ENST0000045338..          IR</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron6/clean TRA2B/ENST0000045338..          IR</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron5/clean TRA2B/ENST0000045338..          IR</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron4/clean TRA2B/ENST0000045338..          IR</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron3/clean TRA2B/ENST0000045338..          IR</span></span>
<span><span class="co">#&gt;                                          EventRegion              intron_id</span></span>
<span><span class="co">#&gt;                                          &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron8/clean chrZ:1921-2559/- ENST00000453386_Intr..</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron7/clean chrZ:2634-3631/- ENST00000453386_Intr..</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron6/clean chrZ:3692-5298/- ENST00000453386_Intr..</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron5/clean chrZ:5383-6205/- ENST00000453386_Intr..</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron4/clean chrZ:6322-7990/- ENST00000453386_Intr..</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron3/clean chrZ:8180-9658/- ENST00000453386_Intr..</span></span>
<span><span class="co">#&gt;                                     Inc_Is_Protein_Coding Exc_Is_Protein_Coding</span></span>
<span><span class="co">#&gt;                                                 &lt;logical&gt;             &lt;logical&gt;</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron8/clean                  TRUE                  TRUE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron7/clean                  TRUE                  TRUE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron6/clean                  TRUE                  TRUE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron5/clean                  TRUE                  TRUE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron4/clean                  TRUE                  TRUE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron3/clean                  TRUE                  TRUE</span></span>
<span><span class="co">#&gt;                                     Exc_Is_NMD Inc_Is_NMD     Inc_TSL</span></span>
<span><span class="co">#&gt;                                      &lt;logical&gt;  &lt;logical&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron8/clean      FALSE       TRUE           1</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron7/clean      FALSE       TRUE           1</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron6/clean      FALSE       TRUE           1</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron5/clean      FALSE       TRUE           1</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron4/clean      FALSE       TRUE           1</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron3/clean      FALSE       TRUE           1</span></span>
<span><span class="co">#&gt;                                         Exc_TSL          Event1a     Event2a</span></span>
<span><span class="co">#&gt;                                     &lt;character&gt;      &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron8/clean           1 chrZ:1921-2559/-          NA</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron7/clean           1 chrZ:2634-3631/-          NA</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron6/clean           1 chrZ:3692-5298/-          NA</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron5/clean           1 chrZ:5383-6205/-          NA</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron4/clean           1 chrZ:6322-7990/-          NA</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron3/clean           1 chrZ:8180-9658/-          NA</span></span>
<span><span class="co">#&gt;                                         Event1b     Event2b</span></span>
<span><span class="co">#&gt;                                     &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron8/clean          NA          NA</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron7/clean          NA          NA</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron6/clean          NA          NA</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron5/clean          NA          NA</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron4/clean          NA          NA</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron3/clean          NA          NA</span></span>
<span><span class="co">#&gt;                                     is_always_first_intron</span></span>
<span><span class="co">#&gt;                                                  &lt;logical&gt;</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron8/clean                  FALSE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron7/clean                  FALSE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron6/clean                  FALSE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron5/clean                  FALSE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron4/clean                  FALSE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron3/clean                  FALSE</span></span>
<span><span class="co">#&gt;                                     is_always_last_intron is_annotated_IR</span></span>
<span><span class="co">#&gt;                                                 &lt;logical&gt;       &lt;logical&gt;</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron8/clean                    NA           FALSE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron7/clean                 FALSE           FALSE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron6/clean                 FALSE           FALSE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron5/clean                 FALSE           FALSE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron4/clean                 FALSE            TRUE</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron3/clean                 FALSE           FALSE</span></span>
<span><span class="co">#&gt;                                     NMD_direction</span></span>
<span><span class="co">#&gt;                                         &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron8/clean             1</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron7/clean             1</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron6/clean             1</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron5/clean             1</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron4/clean             1</span></span>
<span><span class="co">#&gt; TRA2B/ENST00000453386_Intron3/clean             1</span></span></code></pre></div>
<p>Columns contain information about each sample. By default, no
annotations are assigned to each sample. These can be assigned as shown
above.</p>
<p>Also, <code>NxtSE</code> objects can be subsetted by rows (ASEs) or
columns (samples). This is useful if one wishes to perform analysis on a
subset of the dataset, or only on a subset of ASEs (say for example,
only skipped exon events). Subsetting is performed just like for
<code>SummarizedExperiment</code> objects:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Subset by columns: select the first 2 samples</span></span>
<span><span class="va">se_sample_subset</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Subset by rows: select the first 10 ASE events</span></span>
<span><span class="va">se_ASE_subset</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span></span></code></pre></div>
<p><br></p>
</details><p><br></p>
</div>
<div class="section level4">
<h4 id="filtering-high-confidence-events">Filtering high-confidence events<a class="anchor" aria-label="anchor" href="#filtering-high-confidence-events"></a>
</h4>
<p>SpliceWiz offers default filters to identify and remove low
confidence alternative splice events (ASEs). Run the default filter
using the following:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se.filtered</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu"><a href="../reference/Run_SpliceWiz_Filters.html">applyFilters</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; Running Depth filter</span></span>
<span><span class="co">#&gt; Running Participation filter</span></span>
<span><span class="co">#&gt; Running Participation filter</span></span>
<span><span class="co">#&gt; Running Consistency filter</span></span>
<span><span class="co">#&gt; Running Terminus filter</span></span>
<span><span class="co">#&gt; Running ExclusiveMXE filter</span></span></code></pre></div>
<p><br></p>
<details><summary>
Using the GUI
</summary><p><br>
After following the GUI tutorials in the prior sections, click on
<code>Analysis</code> and then <code>Filters</code> from the menu bar.
It should look like this:</p>
<div class="figure" style="text-align: center">
<img src="img/Filters_empty.jpg" alt="Filters - GUI" width="800pt"><p class="caption">
Filters - GUI
</p>
</div>
<p>To load SpliceWiz’s default filters, click the top right button
<code>Load Default Filters</code>. Then to apply these filters to the
NxtSE, click <code>Apply Filters</code>. After the filters have been
run, your session should now look like this:</p>
<div class="figure" style="text-align: center">
<img src="img/Filters_applied.jpg" alt="SpliceWiz default filters- GUI" width="800pt"><p class="caption">
SpliceWiz default filters- GUI
</p>
</div>
</details><p><br></p>
<details><summary>
Why do we need to filter alternative splicing events?
</summary><p><br>
Often, the gene annotations contain isoforms for all discovered splicing
events. Most annotated transcripts are not expressed, and their
inclusion in differential analysis complicates results including
adjusting for multiple testing. It is prudent to filter these out using
various approaches, akin to removing genes with low gene counts in
differential gene analysis. We suggest using the default filters which
work well for small experiments with sequencing depths at 100-million
paired-end reads.</p>
<p>To learn more about filters, consult the documentation via
<code>?ASEFilters</code></p>
</details><p><br></p>
</div>
<div class="section level4">
<h4 id="performing-differential-analysis">Performing differential analysis<a class="anchor" aria-label="anchor" href="#performing-differential-analysis"></a>
</h4>
<p>Using the limma wrapper <code><a href="../reference/ASE-methods.html">ASE_limma()</a></code>, perform
differential ASE analysis between conditions “A” and “B”:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Requires limma to be installed:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">"limma"</a></span><span class="op">)</span></span>
<span><span class="va">res_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ASE-methods.html">ASE_limma</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se.filtered</span>,</span>
<span>    test_factor <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>    test_nom <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>    test_denom <span class="op">=</span> <span class="st">"A"</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<details><summary>
Using the GUI
</summary><p><br>
After running the previous sections (of the GUI instructions), click
<code>Analysis</code> and then
<code>Differential Expression Analysis</code> on the menu side bar. It
should look something like this (but with empty fields):</p>
<div class="figure" style="text-align: center">
<img src="img/DE_empty.jpg" alt="Differential Expression GUI" width="800pt"><p class="caption">
Differential Expression GUI
</p>
</div>
<p>To perform limma-based differential analysis, first ensure
<code>Method</code> is set to <code>limma</code>. Using the
<code>Variable</code> drop-down box, select <code>condition</code>.
Then, select the <code>Nominator</code> and <code>Denominator</code>
fields to <code>A</code> and <code>B</code>, respectively. Leave the
batch factor fields as <code>(none)</code>. Then, click
<code>Perform DE</code>.</p>
<p>Once differential expression analysis has finished, your session
should look like below. The output is a DT-based data table equivalent
to the <code><a href="../reference/ASE-methods.html">ASE_limma()</a></code> function.</p>
<div class="figure" style="text-align: center">
<img src="img/DE_limma.jpg" alt="Differential Expression (limma) GUI" width="800pt"><p class="caption">
Differential Expression (limma) GUI
</p>
</div>
<p>NB: The interface allows users to choose to sort the results either
by nominal or (multiple-testing) adjusted P values</p>
<p>NB2: There are 3 different ways Intron Retention events can be
quantified and analysed - see “What are the different ways intron
retention is measured?” below for further details.</p>
</details><p><br></p>
<details><summary>
What are the options for differential ASE analysis?
</summary><p><br>
SpliceWiz provides wrappers to three established algorithms:</p>
<ul>
<li>
<code>ASE_limma</code> uses <code>limma</code> to model isoform
counts as log-normal distributions. Limma is probably the fastest method
and is ideal for large datasets. Time series analysis is available for
this mode.</li>
<li>
<code>ASE_DESeq</code> uses <code>DESeq2</code> to model isoform
counts as negative binomial distribution. This method is the most
computationally expensive, but gives robust results. Time series
analysis is also available for this mode</li>
<li>
<code>ASE_DoubleExpSeq</code> uses the lesser-known CRAN package
<code>DoubleExpSeq</code>. This package uses the beta-binomial
distribution to model isoform counts. The method is at least as fast as
<code>limma</code>, but for now it is restricted to analysis between two
groups (i.e. batch correction is not implemented)</li>
<li>(New in 0.99.4+) <code>ASE_satuRn</code> uses the new
<code>satuRn</code> package which models counts using the quasi-binomial
distribution.</li>
</ul>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Requires limma to be installed:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">"limma"</a></span><span class="op">)</span></span>
<span><span class="va">res_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ASE-methods.html">ASE_limma</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se.filtered</span>,</span>
<span>    test_factor <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>    test_nom <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>    test_denom <span class="op">=</span> <span class="st">"A"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Requires DESeq2 to be installed:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/mikelove/DESeq2" class="external-link">"DESeq2"</a></span><span class="op">)</span></span>
<span><span class="va">res_deseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ASE-methods.html">ASE_DESeq</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se.filtered</span>,</span>
<span>    test_factor <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>    test_nom <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>    test_denom <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Requires DoubleExpSeq to be installed:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"DoubleExpSeq"</span><span class="op">)</span></span>
<span><span class="va">res_DES</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ASE-methods.html">ASE_DoubleExpSeq</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se.filtered</span>,</span>
<span>    test_factor <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>    test_nom <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>    test_denom <span class="op">=</span> <span class="st">"A"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Requires satuRn to be installed:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/statOmics/satuRn" class="external-link">"satuRn"</a></span><span class="op">)</span></span>
<span><span class="va">res_saturn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ASE-methods.html">ASE_satuRn</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se.filtered</span>,</span>
<span>    test_factor <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>    test_nom <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>    test_denom <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</details><p><br></p>
<details><summary>
What are the different ways intron retention is measured?
</summary><p><br>
Intron retention can be measured via two approaches.</p>
<p>The first (and preferred) approach is using IR-ratio. We presume that
every intron is potentially retained (thus ignoring annotation). Given
this results in many overlapping introns, SpliceWiz adjusts for this via
the following:</p>
<ul>
<li>Where there are mutually-overlapping introns, the less abundant
intron is removed from the analysis. Abundance is estimated across the
entire dataset, and less-abundant overlapping introns are removed at the
<code><a href="../reference/makeSE.html">makeSE()</a></code> step.</li>
<li>IR-ratio measures included (intronic) abundance using an identical
approach to IRFinder, i.e., it calculates the trimmed mean of sequencing
depth across the intron, excluding annotated outliers (other exons,
intronic elements). Given we cannot assume whether the exact intron
corresponds to the major isoform, we estimate splicing abundance by
summing junction reads that share either exon cluster as the intron of
interest (<code>SpliceOver</code> method in SpliceWiz). Alternately,
users can choose to use IRFinder’s <code>SpliceMax</code> method,
summing junction reads that share either splice junction with the intron
of interest. This choice is also made by the user at the
<code><a href="../reference/makeSE.html">makeSE()</a></code> step.</li>
</ul>
<p>At the differential analysis step, users can choose the
following:</p>
<ul>
<li>
<code>IRmode = "all"</code> - all introns are potentially retained,
use IR-ratio to quantify IR (<code>EventType = "IR"</code>)</li>
<li>
<code>IRmode = "annotated"</code> - only annotated retained intron
events are considered, but use IR-ratio to quantify IR
(<code>EventType = "IR"</code>)</li>
<li>
<code>IRmode = "annotated_binary"</code> - only annotated retained
intron events are considered, use PSI to quantify IR - which considers
the IR-transcript and the transcript isoform with the exactly-spliced
intron as binary alternatives. Splicing of overlapping introns are not
considered in PSI quantitation.</li>
</ul>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_limma_allIntrons</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ASE-methods.html">ASE_limma</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se.filtered</span>,</span>
<span>    test_factor <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>    test_nom <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>    test_denom <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>    IRmode <span class="op">=</span> <span class="st">"all"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">res_limma_annotatedIR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ASE-methods.html">ASE_limma</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se.filtered</span>,</span>
<span>    test_factor <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>    test_nom <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>    test_denom <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>    IRmode <span class="op">=</span> <span class="st">"annotated"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">res_limma_annotated_binaryIR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ASE-methods.html">ASE_limma</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se.filtered</span>,</span>
<span>    test_factor <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>    test_nom <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>    test_denom <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>    IRmode <span class="op">=</span> <span class="st">"annotated_binary"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br>
&lt;&gt;<br></p>
<details><summary>
Can I account for batch factors?
</summary><p><br><code>ASE_limma</code>, <code>ASE_DESeq</code> and
<code>ASE_satuRn</code> can accept up to 2 categories of batches from
which to normalize. For example, to normalize the analysis by the
<code>batch</code> category, one would run:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">"limma"</a></span><span class="op">)</span></span>
<span><span class="va">res_limma_batchnorm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ASE-methods.html">ASE_limma</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se.filtered</span>,</span>
<span>    test_factor <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>    test_nom <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>    test_denom <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>    batch1 <span class="op">=</span> <span class="st">"batch"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</details><p><br></p>
<details><summary>
Can I do time series analysis?
</summary><p><br>
Time series analysis can be performed using limma and DESeq2.</p>
<p>For limma, time series analysis is done using the
<code><a href="../reference/ASE-methods.html">ASE_limma_timeseries()</a></code> function. <code>test_factor</code>,
despite its name, should be a column in <code>colData(se)</code>
containing numerical values that represent time series data.</p>
<p>Note that limma’s time series wrapper function requires the
<code>splines</code> package.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se.filtered</span><span class="op">)</span><span class="op">$</span><span class="va">timevar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">"limma"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"splines"</span><span class="op">)</span></span>
<span><span class="va">res_limma_cont</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ASE-methods.html">ASE_limma_timeseries</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se.filtered</span>,</span>
<span>    test_factor <span class="op">=</span> <span class="st">"timevar"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For DESeq2, time series analysis is performed using the
<code><a href="../reference/ASE-methods.html">ASE_DESeq()</a></code> funcction. The key difference is that, for time
series analysis, simply do not specify the <code>test_nom</code> and
<code>test_denom</code> parameters. As long as the
<code>test_factor</code> contains numeric values, <code>ASE_DESeq</code>
will treat it as a continuous variable. See the following example:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se.filtered</span><span class="op">)</span><span class="op">$</span><span class="va">timevar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/mikelove/DESeq2" class="external-link">"DESeq2"</a></span><span class="op">)</span></span>
<span><span class="va">res_deseq_cont</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ASE-methods.html">ASE_DESeq</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se.filtered</span>,</span>
<span>    test_factor <span class="op">=</span> <span class="st">"timevar"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br>
&lt;&gt;<br></p>
</details></details>
</div>
</div>
<div class="section level3">
<h3 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h3>
<div class="section level4">
<h4 id="volcano-plots">Volcano plots<a class="anchor" aria-label="anchor" href="#volcano-plots"></a>
</h4>
<p>Volcano plots show changes in PSI levels (log fold change, x axis)
against statistical significance (-log10 p values, y axis):</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">res_limma</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">adj.P.Val</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Differential analysis - B vs A"</span>,</span>
<span>         x <span class="op">=</span> <span class="st">"Log2-fold change"</span>, y <span class="op">=</span> <span class="st">"BH-adjusted P values (-log10)"</span><span class="op">)</span></span></code></pre></div>
<details><summary>
Can I visualize significant events for each modality of alternative
splicing events?
</summary><p><br>
Yes. We can use <code>ggplot2</code>’s <code>facet_wrap</code> function
to separately plot volcanos for each modality of ASE. The type of ASE is
contained in the <code>EventType</code> column of the differential
results data frame.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">res_limma</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">adj.P.Val</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">EventType</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Differential analysis - B vs A"</span>,</span>
<span>         x <span class="op">=</span> <span class="st">"Log2-fold change"</span>, y <span class="op">=</span> <span class="st">"BH-adjusted P values (-log10)"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</details><p><br></p>
<details><summary>
Using the GUI
</summary><p><br>
After following the previous sections including differential analysis,
navigate to <code>Display</code> and then <code>Volcano Plot</code>. The
default volcano plot (top 1000 ASEs) will be automatically generated as
a plotly object:</p>
<div class="figure" style="text-align: center">
<img src="img/Vis_volcano.jpg" alt="Volcano Plot - GUI" width="800pt"><p class="caption">
Volcano Plot - GUI
</p>
</div>
<p>You can customize this volcano plot using the following controls:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Adjust the number of top events. Differential ASEs are ranked either
by adjusted or nominal p-values depending on the
<code>Sort by Adjusted P Values</code> in the differential analysis
interface.</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>Select one or more <code>Splice Type</code>s. The volcano plot will
be filtered by the selected ASE modalities</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>You can facet the volcano plot to show separate volcano plots for
each modality</li>
</ol></li>
<li><ol start="4" style="list-style-type: decimal">
<li>Instead of nominal p values, Benjamini Hochberg adjusted p values
are shown if this is <code>ON</code>.</li>
</ol></li>
<li><ol start="5" style="list-style-type: decimal">
<li>If <code>NMD Mode</code> is <code>ON</code>, the horizontal axis
represents whether splicing is shifted towards (positive values) or away
from (negative values) a NMD substrate transcript</li>
</ol></li>
<li><ol start="6" style="list-style-type: decimal">
<li>Clear settings</li>
</ol></li>
</ul></details><p><br></p>
</div>
<div class="section level4">
<h4 id="scatter-plots">Scatter plots<a class="anchor" aria-label="anchor" href="#scatter-plots"></a>
</h4>
<p>Scatter plots are useful for showing splicing levels
(percent-spliced-in, PSI) between two conditions. The results from
differential analysis contains these values and can be plotted:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">res_limma</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">100</span> <span class="op">*</span> <span class="va">AvgPSI_B</span>, y <span class="op">=</span> <span class="fl">100</span> <span class="op">*</span> <span class="va">AvgPSI_A</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">100</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">100</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"PSI values across conditions"</span>,</span>
<span>         x <span class="op">=</span> <span class="st">"PSI of condition B"</span>, y <span class="op">=</span> <span class="st">"PSI of condition A"</span><span class="op">)</span></span></code></pre></div>
<details><summary>
Using the GUI
</summary><p><br>
After following the previous sections including differential analysis,
navigate to <code>Display</code> and then <code>Scatter Plot</code>.</p>
<p>You can customize the scatter plot using the following controls:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Adjust the number of top events. Differential ASEs are ranked either
by adjusted or nominal p-values depending on the
<code>Sort by Adjusted P Values</code> in the differential analysis
interface.</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>Select one or more <code>Splice Type</code>s. The scatter plot will
be filtered by the selected ASE modalities</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>
<code>Variable</code> refers to the annotation category</li>
</ol></li>
<li><ol start="4" style="list-style-type: decimal">
<li>and (5) The <code>X-axis</code> and <code>Y-axis</code> dropdowns
refers to the experimental conditions for which to contrast the average
PSI values.</li>
</ol></li>
<li><ol start="6" style="list-style-type: decimal">
<li>If <code>NMD Mode</code> is <code>ON</code>, the PSI values are
altered such that they represent the inclusion values of the NMD
substrate (instead of the “included” isoform)</li>
</ol></li>
<li><ol start="7" style="list-style-type: decimal">
<li>Clear settings</li>
</ol></li>
</ul>
<p>For now, change the <code>Variable</code> to <code>condition</code>,
<code>X-axis condition</code> to <code>A</code> and
<code>Y-axis condition</code> to <code>B</code>. A scatter plot should
be automatically generated as follows:</p>
<div class="figure" style="text-align: center">
<img src="img/Vis_scatter.jpg" alt="Scatter plot - GUI" width="800pt"><p class="caption">
Scatter plot - GUI
</p>
</div>
</details><p><br></p>
<details><summary>
Selecting ASEs of interest using the interactive plots (GUI)
</summary><p><br>
SpliceWiz GUI generates plotly interactive figures. For volcano and
scatter plots, points of interest can be selected using the lasso or box
select tools. For example, we can select the top hits from the faceted
volcano plot as shown:</p>
<div class="figure" style="text-align: center">
<img src="img/Vis_volcano_select.jpg" alt="Volcano plot - selecting ASEs of interest via the GUI" width="800pt"><p class="caption">
Volcano plot - selecting ASEs of interest via the GUI
</p>
</div>
<p>These ASEs of interest will then be highlighted in other plots, for
example scatter plot:</p>
<div class="figure" style="text-align: center">
<img src="img/Vis_scatter_highlighted.jpg" alt="Scatter plot - highlighted ASE events" width="800pt"><p class="caption">
Scatter plot - highlighted ASE events
</p>
</div>
</details><p><br></p>
<details><summary>
How do I generate average PSI values for many conditions?
</summary><p><br>
SpliceWiz provides the <code><a href="../reference/make_plot_data.html">makeMeanPSI()</a></code> function that can
generate mean PSI values for each condition of a condition category. For
example, the below code will calculate the mean PSIs of each “batch” of
this example experiment:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meanPSIs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_plot_data.html">makeMeanPSI</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se</span>,</span>
<span>    condition <span class="op">=</span> <span class="st">"batch"</span>,</span>
<span>    conditionList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"K"</span>, <span class="st">"L"</span>, <span class="st">"M"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</details><p><br></p>
</div>
<div class="section level4">
<h4 id="heatmaps">Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h4>
<p>Heatmaps are useful for visualizing differential expression of
individual samples, as well as potential patterns of expression.</p>
<p>First, obtain a matrix of PSI values:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a matrix of values of the top 10 differentially expressed events:</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_plot_data.html">makeMatrix</a></span><span class="op">(</span></span>
<span>    <span class="va">se.filtered</span>,</span>
<span>    event_list <span class="op">=</span> <span class="va">res_limma</span><span class="op">$</span><span class="va">EventName</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"PSI"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<details><summary>
How does <code>makeMatrix()</code> work?
</summary><p><br><code><a href="../reference/make_plot_data.html">makeMatrix()</a></code> provides a matrix of PSI values from the given
<code>NxtSE</code> object. The parameters <code>event_list</code> and
<code>sample_list</code> allows subsetting for ASEs and/or samples,
respectively.</p>
<p>The parameter <code>method</code> accepts 3 options:</p>
<ul>
<li>
<code>"PSI"</code> : outputs raw PSI values</li>
<li>
<code>"logit"</code> : outputs logit PSI values</li>
<li>
<code>"Z-score"</code> : outputs Z-score transformed PSI values</li>
</ul>
<p>Also, <code><a href="../reference/make_plot_data.html">makeMatrix()</a></code> facilitates exclusion of low
confidence PSI values. These can occur when counts of both isoforms are
too low. Setting the <code>depth_threshold</code> (default
<code>10</code>) will set samples with total isoform count below this
value to be converted to <code>NA</code>.</p>
<p>Splicing events (ASEs) with too many <code>NA</code> values are
filtered out. Setting the parameter <code>na.percent.max</code> (default
<code>0.1</code>) means any ASE with the proportion of <code>NA</code>
above this threshold will be removed from the final matrix.<br></p>
</details><p><br></p>
<p>Plot this matrix of values in a heatmap:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span></span>
<span><span class="va">anno_col_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se.filtered</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">anno_col_df</span> <span class="op">&lt;-</span> <span class="va">anno_col_df</span><span class="op">[</span>, <span class="fl">1</span>, drop<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">mat</span>, annotation_col <span class="op">=</span> <span class="va">anno_col_df</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
<details><summary>
Using the GUI
</summary><p><br>
Navigate to <code>Display</code>, then <code>Heatmap</code> in the menu
side bar. A heatmap will be automatically generated:</p>
<div class="figure" style="text-align: center">
<img src="img/Vis_heatmap.jpg" alt="Heatmap - GUI" width="800pt"><p class="caption">
Heatmap - GUI
</p>
</div>
<p>The heatmap can be customized as follows:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Select the top hits from either <code>Top All Results</code> -
unfiltered results, <code>Top Filtered Results</code> - results filtered
using the Differential Expression interactive DT data table, or
<code>Top Selected Results</code> - results selected using the
interactive box or lasso select tools in the interactive volcano and/or
scatter plots.</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>Highlight annotation categories. Samples will be highlighted based
on the selected annotation categories</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>Number of top events to display</li>
</ol></li>
<li><ol start="4" style="list-style-type: decimal">
<li>Mode: Either PSI, Logit-PSI, or Z-score transformed values</li>
</ol></li>
<li><ol start="5" style="list-style-type: decimal">
<li>Color palette</li>
</ol></li>
</ul></details><p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="splicewiz-coverage-plots">SpliceWiz Coverage Plots<a class="anchor" aria-label="anchor" href="#splicewiz-coverage-plots"></a>
</h3>
<p>Coverage plots visualize RNA-seq coverage in individual samples.
SpliceWiz uses its coverage normalization algorithm to visualize group
differences in PSIs.<br></p>
<details><summary>
What are SpliceWiz coverage plots and how are they generated?
</summary><p><br>
SpliceWiz produces RNA-seq coverage plots of analysed samples. Coverage
data is compiled simultaneous to the IR and junction quantitation
performed by <code><a href="../reference/processBAM.html">processBAM()</a></code>. This data is saved in “COV”
files, which is a BGZF compressed and indexed file. COV files show
compression and performance gains over BigWig files.</p>
<p>Additionally, SpliceWiz visualizes plots group-averaged coverages,
based on user-defined experimental conditions. This is a powerful tool
to illustrate group-specific differential splicing or IR. SpliceWiz does
this by normalising the coverage depths of each sample based on
transcript depth at the splice junction / intron of interest. By doing
so, the coverage depths of constitutively expressed flanking exons are
normalised to unity. As a result, the intron depths reflect the fraction
of transcripts with retained introns and can be compared across
samples.<br></p>
</details><p><br></p>
<div class="section level4">
<h4 id="coverage-plots-of-individual-samples">Coverage plots of individual samples<a class="anchor" aria-label="anchor" href="#coverage-plots-of-individual-samples"></a>
</h4>
<p>First, lets obtain a list of differential events with delta PSI &gt;
5%:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_limma.filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">res_limma</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">deltaPSI</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p>Plot up to 4 individual samples using
<code><a href="../reference/plotCoverage.html">plotCoverage()</a></code>:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCoverage.html">plotCoverage</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se</span>,</span>
<span>    Event <span class="op">=</span> <span class="va">res_limma.filtered</span><span class="op">$</span><span class="va">EventName</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>    tracks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">5</span><span class="op">)</span><span class="op">]</span>,</span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as plotly object</span></span>
<span>    <span class="va">p</span><span class="op">$</span><span class="va">final_plot</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as ggplot</span></span>
<span>    <span class="fu"><a href="../reference/plotCoverage.html">as_ggplot_cov</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><br></p>
<details><summary>
Plotting junction counts (new in 0.99.4+)
</summary><p><br></p>
<p>Sashimi plots, first popularised by MISO, are coverage plots with
arcs that denote split reads. They are each labelled by a number
denoting the number of split reads spanning the corresponding
junctions.</p>
<p>In SpliceWiz (&gt;=0.99.4) junctions can be plotted by setting
<code>plotJunctions = TRUE</code> from within
<code><a href="../reference/plotCoverage.html">plotCoverage()</a></code></p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCoverage.html">plotCoverage</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se</span>,</span>
<span>    Event <span class="op">=</span> <span class="va">res_limma.filtered</span><span class="op">$</span><span class="va">EventName</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>    tracks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">5</span><span class="op">)</span><span class="op">]</span>,</span>
<span>    plotJunctions <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as plotly object</span></span>
<span>    <span class="va">p</span><span class="op">$</span><span class="va">final_plot</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as ggplot</span></span>
<span>    <span class="fu"><a href="../reference/plotCoverage.html">as_ggplot_cov</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><br></p>
</details><p><br></p>
<details><summary>
The genome track is too cluttered! How can I customize it?
</summary><p><br>
We can set the <code>selected_transcripts</code> parameter of the
<code><a href="../reference/plotCoverage.html">plotCoverage()</a></code> function to only display the specified
transcripts. Selected transcripts can be supplied by name or ID
(<code>transcript_name</code> and <code>transcript_id</code> field in
the GTF file, respectively)</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCoverage.html">plotCoverage</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se</span>,</span>
<span>    Event <span class="op">=</span> <span class="va">res_limma.filtered</span><span class="op">$</span><span class="va">EventName</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>    tracks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">5</span><span class="op">)</span><span class="op">]</span>,</span>
<span>    selected_transcripts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NSUN5-201"</span>, <span class="st">"NSUN5-203"</span>, <span class="st">"NSUN5-206"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as plotly object</span></span>
<span>    <span class="va">p</span><span class="op">$</span><span class="va">final_plot</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as ggplot</span></span>
<span>    <span class="fu"><a href="../reference/plotCoverage.html">as_ggplot_cov</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><br></p>
</details><p><br></p>
</div>
<div class="section level4">
<h4 id="group-averaged-coverage-plots">Group-averaged coverage plots<a class="anchor" aria-label="anchor" href="#group-averaged-coverage-plots"></a>
</h4>
<p>We plot average coverages of groups “A” and “B” of the “condition”
category, as follows:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCoverage.html">plotCoverage</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se</span>,</span>
<span>    Event <span class="op">=</span> <span class="va">res_limma.filtered</span><span class="op">$</span><span class="va">EventName</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>    tracks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>    condition <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>    plotJunctions <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as plotly object</span></span>
<span>    <span class="va">p</span><span class="op">$</span><span class="va">final_plot</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as ggplot</span></span>
<span>    <span class="fu"><a href="../reference/plotCoverage.html">as_ggplot_cov</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Note that junctions are labelled not by their counts (which are
different between replicates of a group, and are dependent on gene
expression and sequencing depth). Instead, SpliceWiz labels junction
arcs by the mean +/- sd of “provisional PSI”. These are PSI estimates
calculated for each sample, based on the ratio of the junction read
count as a proportion of total spliced read counts that share either
exon cluster with that of the assessed junction.</p>
<p><br></p>
<details><summary>
How do I stack conditions together to better visualize their coverage
differences?
</summary><p><br></p>
<p>Differences in group-normalized coverage are best visualized when the
traces are stacked together in the same track. To do this, set
<code>stack_tracks = TRUE</code> from within
<code><a href="../reference/plotCoverage.html">plotCoverage()</a></code>. Note this is only enabled when plotting
group-normalized coverage plots (i.e., when <code>condition</code> is
set).</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCoverage.html">plotCoverage</a></span><span class="op">(</span></span>
<span>    se <span class="op">=</span> <span class="va">se</span>,</span>
<span>    Event <span class="op">=</span> <span class="va">res_limma.filtered</span><span class="op">$</span><span class="va">EventName</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>    tracks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>    condition <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>    stack_tracks <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as plotly object</span></span>
<span>    <span class="va">p</span><span class="op">$</span><span class="va">final_plot</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="co"># Display as ggplot</span></span>
<span>    <span class="fu"><a href="../reference/plotCoverage.html">as_ggplot_cov</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><br></p>
</details><p><br></p>
<details><summary>
What are the shaded ribbons in this plot?
</summary><p><br></p>
<p>The shaded ribbons represent the uncertainty of the coverage. This is
denoted by the <code>ribbon_mode</code> parameter of the plotCoverage()
command. The default is standard deviation <code>sd</code>.
Alternatively, users can choose the 95% confidence interval
<code>ci</code>, or the standard error of the mean <code>sem</code>.</p>
<p>This feature is important for two reasons:</p>
<ul>
<li>The uncertainties around the regions expected to be invariable (e.g.
constitutively expressed flanking exons) are expected to be low if
SpliceWiz’s estimate of exonic abundance is accurate,</li>
<li>The confidences of the differentially expressed
alternatively-spliced exons (or retained introns) represents the
certainty of the differential expression of these regions.<br>
</li>
</ul></details><p><br></p>
</div>
<div class="section level4">
<h4 id="using-the-gui">Using the GUI<a class="anchor" aria-label="anchor" href="#using-the-gui"></a>
</h4>
<p>Click on <code>Display</code> and then <code>Coverage plot</code>
from the menu side bar. It should look like this:</p>
<div class="figure" style="text-align: center">
<img src="img/Cov.jpg" alt="Coverage Plots - GUI" width="800pt"><p class="caption">
Coverage Plots - GUI
</p>
</div>
<p>Coverage plots can be customized as follows:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Genes: display whole genes</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>Events: display individual ASE events</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>Coordinates: display by genomic coordinates</li>
</ol></li>
<li><ol start="4" style="list-style-type: decimal">
<li>Zoom in or out (each step is a factor of 3)</li>
</ol></li>
<li><ol start="5" style="list-style-type: decimal">
<li>Strand: select unstranded (*), positive (+) or negative (-). NB most
RNA-seq protocols are reverse-stranded; this means the strand is
opposite to that of the expressed RNA</li>
</ol></li>
<li><ol start="6" style="list-style-type: decimal">
<li>Select events ranked by p values from All results, Filtered results
(using the interactive DT data table), or Selected results using lasso /
box select tools on the volcano or scatter plots</li>
</ol></li>
<li><ol start="7" style="list-style-type: decimal">
<li>Graph mode. You can either pan (across the genomic axis), zoom tool
(drag to zoom in) or movable labels (labels can be moved by
dragging)</li>
</ol></li>
<li><ol start="8" style="list-style-type: decimal">
<li>Number of top events to display in the Events dropdown menu</li>
</ol></li>
<li><ol start="9" style="list-style-type: decimal">
<li>Refresh plot. The plot will be automatically refreshed after a few
seconds when selections are changed. However, for some controls (such as
genes, events), the plot needs to be manually refreshed</li>
</ol></li>
<li><ol start="10" style="list-style-type: decimal">
<li>View either individual samples or group-averaged by condition</li>
</ol></li>
<li><ol start="11" style="list-style-type: decimal">
<li>Normalize RNA-seq coverage values by the junction coverage reads of
the selected ASE event. This ensures that exon depths are normalized to
unity. Only applicable to <code>View</code> by
<code>condition</code>
</li>
</ol></li>
<li><ol start="12" style="list-style-type: decimal">
<li>Condition: select which category of condition to contrast</li>
</ol></li>
<li><ol start="13" style="list-style-type: decimal">
<li>Tracks: select either individual samples, or experimental
conditions</li>
</ol></li>
<li><ol start="14" style="list-style-type: decimal">
<li>Plot Junctions: Plot split read counts when viewing by individual
samples, or provisional PSI values (calculated based on split reads
across exon clusters) when viewing by conditions. Not plotted when stack
traces is ON.</li>
</ol></li>
<li><ol start="15" style="list-style-type: decimal">
<li>Stack traces: if on, all traces are output onto the same track (view
by condition only)</li>
</ol></li>
<li><ol start="16" style="list-style-type: decimal">
<li>Pairwise t-test (only applicable to view by condition, 2 tracks
highlighted only). A pairwise t-test of normalized coverages will be
performed, and per-window t-test results (-log10 transformed) will be
displayed. Higher values indicate higher significance</li>
</ol></li>
<li><ol start="17" style="list-style-type: decimal">
<li>Display Key Isoforms: Only plots transcripts that contain at least
one of the two alternate isoforms being plotted.</li>
</ol></li>
<li><ol start="18" style="list-style-type: decimal">
<li>Condensed tracks: whether the individual transcripts
<code>OFF</code> or collapsed combined transcripts <code>ON</code> is
displayed<br>
</li>
</ol></li>
</ul>
</div>
</div>
</div>
<div class="section level2">
<h2 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Big Sur ... 10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] splines   stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [8] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] satuRn_1.6.0                DoubleExpSeq_1.1           </span></span>
<span><span class="co">#&gt;  [3] DESeq2_1.38.0               SummarizedExperiment_1.28.0</span></span>
<span><span class="co">#&gt;  [5] Biobase_2.58.0              MatrixGenerics_1.10.0      </span></span>
<span><span class="co">#&gt;  [7] matrixStats_0.62.0          GenomicRanges_1.50.1       </span></span>
<span><span class="co">#&gt;  [9] GenomeInfoDb_1.34.3         IRanges_2.32.0             </span></span>
<span><span class="co">#&gt; [11] S4Vectors_0.36.0            BiocGenerics_0.44.0        </span></span>
<span><span class="co">#&gt; [13] limma_3.54.0                fstcore_0.9.12             </span></span>
<span><span class="co">#&gt; [15] SpliceWiz_1.1.2             NxtIRFdata_1.4.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] AnnotationHub_3.6.0           BiocFileCache_2.6.0          </span></span>
<span><span class="co">#&gt;   [3] systemfonts_1.0.4             lazyeval_0.2.2               </span></span>
<span><span class="co">#&gt;   [5] shinydashboard_0.7.2          crosstalk_1.2.0              </span></span>
<span><span class="co">#&gt;   [7] BiocParallel_1.32.1           ggplot2_3.4.0                </span></span>
<span><span class="co">#&gt;   [9] digest_0.6.30                 ca_0.71.1                    </span></span>
<span><span class="co">#&gt;  [11] foreach_1.5.2                 htmltools_0.5.3              </span></span>
<span><span class="co">#&gt;  [13] viridis_0.6.2                 fansi_1.0.3                  </span></span>
<span><span class="co">#&gt;  [15] magrittr_2.0.3                memoise_2.0.1                </span></span>
<span><span class="co">#&gt;  [17] BSgenome_1.66.1               shinyFiles_0.9.3             </span></span>
<span><span class="co">#&gt;  [19] Biostrings_2.66.0             annotate_1.76.0              </span></span>
<span><span class="co">#&gt;  [21] R.utils_2.12.2                pkgdown_2.0.6                </span></span>
<span><span class="co">#&gt;  [23] prettyunits_1.1.1             colorspace_2.0-3             </span></span>
<span><span class="co">#&gt;  [25] blob_1.2.3                    rappdirs_0.3.3               </span></span>
<span><span class="co">#&gt;  [27] textshaping_0.3.6             xfun_0.34                    </span></span>
<span><span class="co">#&gt;  [29] dplyr_1.0.10                  crayon_1.5.2                 </span></span>
<span><span class="co">#&gt;  [31] RCurl_1.98-1.9                jsonlite_1.8.3               </span></span>
<span><span class="co">#&gt;  [33] genefilter_1.80.0             iterators_1.0.14             </span></span>
<span><span class="co">#&gt;  [35] survival_3.4-0                glue_1.6.2                   </span></span>
<span><span class="co">#&gt;  [37] registry_0.5-1                gtable_0.3.1                 </span></span>
<span><span class="co">#&gt;  [39] zlibbioc_1.44.0               XVector_0.38.0               </span></span>
<span><span class="co">#&gt;  [41] webshot_0.5.4                 DelayedArray_0.24.0          </span></span>
<span><span class="co">#&gt;  [43] Rhdf5lib_1.20.0               HDF5Array_1.26.0             </span></span>
<span><span class="co">#&gt;  [45] scales_1.2.1                  pheatmap_1.0.12              </span></span>
<span><span class="co">#&gt;  [47] DBI_1.1.3                     Rcpp_1.0.9                   </span></span>
<span><span class="co">#&gt;  [49] progress_1.2.2                viridisLite_0.4.1            </span></span>
<span><span class="co">#&gt;  [51] xtable_1.8-4                  bit_4.0.4                    </span></span>
<span><span class="co">#&gt;  [53] DT_0.26                       htmlwidgets_1.5.4            </span></span>
<span><span class="co">#&gt;  [55] httr_1.4.4                    RColorBrewer_1.1-3           </span></span>
<span><span class="co">#&gt;  [57] ellipsis_0.3.2                farver_2.1.1                 </span></span>
<span><span class="co">#&gt;  [59] pkgconfig_2.0.3               XML_3.99-0.12                </span></span>
<span><span class="co">#&gt;  [61] R.methodsS3_1.8.2             sass_0.4.2                   </span></span>
<span><span class="co">#&gt;  [63] dbplyr_2.2.1                  locfit_1.5-9.6               </span></span>
<span><span class="co">#&gt;  [65] utf8_1.2.2                    labeling_0.4.2               </span></span>
<span><span class="co">#&gt;  [67] tidyselect_1.2.0              rlang_1.0.6                  </span></span>
<span><span class="co">#&gt;  [69] later_1.3.0                   AnnotationDbi_1.60.0         </span></span>
<span><span class="co">#&gt;  [71] munsell_0.5.0                 BiocVersion_3.16.0           </span></span>
<span><span class="co">#&gt;  [73] tools_4.2.2                   cachem_1.0.6                 </span></span>
<span><span class="co">#&gt;  [75] cli_3.4.1                     generics_0.1.3               </span></span>
<span><span class="co">#&gt;  [77] RSQLite_2.2.18                evaluate_0.18                </span></span>
<span><span class="co">#&gt;  [79] stringr_1.4.1                 fastmap_1.1.0                </span></span>
<span><span class="co">#&gt;  [81] heatmaply_1.4.0               yaml_2.3.6                   </span></span>
<span><span class="co">#&gt;  [83] ragg_1.2.4                    ompBAM_1.2.0                 </span></span>
<span><span class="co">#&gt;  [85] knitr_1.40                    bit64_4.0.5                  </span></span>
<span><span class="co">#&gt;  [87] fs_1.5.2                      purrr_0.3.5                  </span></span>
<span><span class="co">#&gt;  [89] KEGGREST_1.38.0               dendextend_1.16.0            </span></span>
<span><span class="co">#&gt;  [91] pbapply_1.5-0                 sparseMatrixStats_1.10.0     </span></span>
<span><span class="co">#&gt;  [93] mime_0.12                     rhandsontable_0.3.8          </span></span>
<span><span class="co">#&gt;  [95] R.oo_1.25.0                   compiler_4.2.2               </span></span>
<span><span class="co">#&gt;  [97] plotly_4.10.1                 filelock_1.0.2               </span></span>
<span><span class="co">#&gt;  [99] curl_4.3.3                    png_0.1-7                    </span></span>
<span><span class="co">#&gt; [101] interactiveDisplayBase_1.36.0 geneplotter_1.76.0           </span></span>
<span><span class="co">#&gt; [103] tibble_3.1.8                  bslib_0.4.1                  </span></span>
<span><span class="co">#&gt; [105] stringi_1.7.8                 highr_0.9                    </span></span>
<span><span class="co">#&gt; [107] desc_1.4.2                    lattice_0.20-45              </span></span>
<span><span class="co">#&gt; [109] Matrix_1.5-1                  vctrs_0.5.0                  </span></span>
<span><span class="co">#&gt; [111] pillar_1.8.1                  lifecycle_1.0.3              </span></span>
<span><span class="co">#&gt; [113] rhdf5filters_1.10.0           BiocManager_1.30.19          </span></span>
<span><span class="co">#&gt; [115] jquerylib_0.1.4               data.table_1.14.4            </span></span>
<span><span class="co">#&gt; [117] bitops_1.0-7                  seriation_1.4.0              </span></span>
<span><span class="co">#&gt; [119] httpuv_1.6.6                  rtracklayer_1.58.0           </span></span>
<span><span class="co">#&gt; [121] R6_2.5.1                      BiocIO_1.8.0                 </span></span>
<span><span class="co">#&gt; [123] TSP_1.2-1                     promises_1.2.0.1             </span></span>
<span><span class="co">#&gt; [125] gridExtra_2.3                 codetools_0.2-18             </span></span>
<span><span class="co">#&gt; [127] boot_1.3-28                   assertthat_0.2.1             </span></span>
<span><span class="co">#&gt; [129] rhdf5_2.42.0                  rprojroot_2.0.3              </span></span>
<span><span class="co">#&gt; [131] rjson_0.2.21                  withr_2.5.0                  </span></span>
<span><span class="co">#&gt; [133] shinyWidgets_0.7.4            GenomicAlignments_1.34.0     </span></span>
<span><span class="co">#&gt; [135] Rsamtools_2.14.0              locfdr_1.1-8                 </span></span>
<span><span class="co">#&gt; [137] GenomeInfoDbData_1.2.9        hms_1.1.2                    </span></span>
<span><span class="co">#&gt; [139] parallel_4.2.2                fst_0.9.8                    </span></span>
<span><span class="co">#&gt; [141] grid_4.2.2                    tidyr_1.2.1                  </span></span>
<span><span class="co">#&gt; [143] rmarkdown_2.18                DelayedMatrixStats_1.20.0    </span></span>
<span><span class="co">#&gt; [145] numDeriv_2016.8-1.1           shiny_1.7.3                  </span></span>
<span><span class="co">#&gt; [147] restfulr_0.0.15</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alex Chit Hei Wong.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
