<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Differential Alternative Splicing Event analysis — ASE-methods • SpliceWiz</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Differential Alternative Splicing Event analysis — ASE-methods"><meta property="og:description" content="Use Limma, DESeq2, DoubleExpSeq and satuRn wrapper functions to test for
differential Alternative Splice Events (ASEs)"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SpliceWiz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/SW_Cookbook.html">SpliceWiz: the cookbook</a>
    </li>
    <li>
      <a href="../articles/SW_QuickStart.html">SpliceWiz: Quick Start</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/alexchwong/SpliceWiz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Differential Alternative Splicing Event analysis</h1>
    <small class="dont-index">Source: <a href="https://github.com/alexchwong/SpliceWiz/blob/HEAD/R/ASE-methods.R" class="external-link"><code>R/ASE-methods.R</code></a></small>
    <div class="hidden name"><code>ASE-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Use Limma, DESeq2, DoubleExpSeq and satuRn wrapper functions to test for
differential Alternative Splice Events (ASEs)</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ASE_limma</span><span class="op">(</span></span>
<span>  <span class="va">se</span>,</span>
<span>  <span class="va">test_factor</span>,</span>
<span>  <span class="va">test_nom</span>,</span>
<span>  <span class="va">test_denom</span>,</span>
<span>  batch1 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  batch2 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  IRmode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"all"</span>, <span class="st">"annotated"</span>, <span class="st">"annotated_binary"</span><span class="op">)</span>,</span>
<span>  filter_antiover <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter_antinear <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ASE_limma_timeseries</span><span class="op">(</span></span>
<span>  <span class="va">se</span>,</span>
<span>  <span class="va">test_factor</span>,</span>
<span>  batch1 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  batch2 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  degrees_of_freedom <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  IRmode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"all"</span>, <span class="st">"annotated"</span>, <span class="st">"annotated_binary"</span><span class="op">)</span>,</span>
<span>  filter_antiover <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter_antinear <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ASE_DESeq</span><span class="op">(</span></span>
<span>  <span class="va">se</span>,</span>
<span>  <span class="va">test_factor</span>,</span>
<span>  <span class="va">test_nom</span>,</span>
<span>  <span class="va">test_denom</span>,</span>
<span>  batch1 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  batch2 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  n_threads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  IRmode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"all"</span>, <span class="st">"annotated"</span>, <span class="st">"annotated_binary"</span><span class="op">)</span>,</span>
<span>  filter_antiover <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter_antinear <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ASE_DoubleExpSeq</span><span class="op">(</span></span>
<span>  <span class="va">se</span>,</span>
<span>  <span class="va">test_factor</span>,</span>
<span>  <span class="va">test_nom</span>,</span>
<span>  <span class="va">test_denom</span>,</span>
<span>  IRmode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"all"</span>, <span class="st">"annotated"</span>, <span class="st">"annotated_binary"</span><span class="op">)</span>,</span>
<span>  filter_antiover <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter_antinear <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ASE_satuRn</span><span class="op">(</span></span>
<span>  <span class="va">se</span>,</span>
<span>  <span class="va">test_factor</span>,</span>
<span>  <span class="va">test_nom</span>,</span>
<span>  <span class="va">test_denom</span>,</span>
<span>  batch1 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  batch2 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  n_threads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  IRmode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"all"</span>, <span class="st">"annotated"</span>, <span class="st">"annotated_binary"</span><span class="op">)</span>,</span>
<span>  filter_antiover <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter_antinear <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  filterByMinCPM <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>se</dt>
<dd><p>The <a href="NxtSE-class.html">NxtSE</a> object created by <code><a href="makeSE.html">makeSE()</a></code>. To reduce
runtime and avoid excessive multiple testing, consider filtering
the object using <a href="Run_SpliceWiz_Filters.html">applyFilters</a></p></dd>


<dt>test_factor</dt>
<dd><p>The column name in the sample annotation <code>colData(se)</code>
that contains the desired variables to be contrasted. For
<code>ASE_limma_timeseries()</code> and <code>ASE_DESeq()</code> (when <code>test_nom</code> and
<code>test_denom</code> parameters are left blank), <code>test_factor</code> must contain
numerical values representing the time variable.</p></dd>


<dt>test_nom</dt>
<dd><p>The nominator condition to test for differential ASE. Usually
the "treatment" condition</p></dd>


<dt>test_denom</dt>
<dd><p>The denominator condition to test against for differential
ASE. Usually the "control" condition</p></dd>


<dt>batch1, batch2</dt>
<dd><p>(Optional, limma and DESeq2 only) One or two condition
types containing batch information to account for.</p></dd>


<dt>IRmode</dt>
<dd><p>(default <code>all</code>) Choose the approach to quantify IR events.
Default <code>all</code> considers all introns as potentially retained, and calculates
IR-ratio based on total splicing across the intron using the "SpliceOver"
or "SpliceMax" approach (see <a href="collateData.html">collateData</a>). Other options include
<code>annotated</code> which calculates IR-ratios for annotated introns only, and
<code>annotated_binary</code> which calculates PSI considering the "included"
isoform as the IR-transcript, and the "excluded" transcript is
quantified from splice counts only across the exact intron
(but not that of overlapping introns). IR-ratio are denoted as "IR" events,
whereas PSIs calculated using IR and intron-spliced binary alternatives are
denoted as "RI" events.</p></dd>


<dt>filter_antiover, filter_antinear</dt>
<dd><p>Whether to remove novel IR events that
overlap over or near anti-sense genes. Default will exclude antiover but
not antinear introns. These are ignored if strand-specific RNA-seq
protocols are used.</p></dd>


<dt>degrees_of_freedom</dt>
<dd><p>(default <code>1</code>) The complexity of time series trends
modeled by <code>ASE_limma_timeseries</code>. E.g., <code>1</code> will only model linear trends,
whereas <code>2</code> extends the capacity for quadratic trends, <code>3</code> for cubic
trends, etc.</p></dd>


<dt>n_threads</dt>
<dd><p>(DESeq2 only) How many threads to use for DESeq2
based analysis.</p></dd>


<dt>filterByMinCPM</dt>
<dd><p>(default <code>0</code>) In <code>ASE_satuRn()</code>, Included/Excluded
counts will be filtered using this value as the threshold prior to satuRn
analysis. Filtering is performed using <code>edgeR::filterByExpr()</code> parsing
this parameter into its <code>min.count</code> parameter.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>For all methods, a data.table containing the following:</p><ul><li><p>EventName: The name of the ASE event. This identifies each ASE
in downstream functions including <a href="make_plot_data.html">makeMeanPSI</a>, <a href="make_plot_data.html">makeMatrix</a>,
and <a href="plotCoverage.html">plotCoverage</a></p></li>
<li><p>EventType: The type of event. See details section above.</p></li>
<li><p>EventRegion: The genomic coordinates the event occupies. This spans the
most upstream and most downstream splice junction involved in the ASE,
and is use to guide the <a href="plotCoverage.html">plotCoverage</a> function.</p></li>
<li><p>flags: Indicates which isoforms are NMD substrates and/or which are
formed by novel splicing only.</p></li>
<li><p>AvgPSI_nom, Avg_PSI_denom: the average percent spliced in / percent
IR levels for the two conditions being contrasted. <code>nom</code> and <code>denom</code> in
column names are replaced with the condition names. Note this is a
geometric mean, based on the arithmetic mean of logit PSI values.</p></li>
<li><p>deltaPSI: The difference in PSI between the mean values of the two
conditions.</p></li>
</ul><p><strong>limma specific output</strong></p>


<ul><li><p>logFC, AveExpr, t, P.Value, adj.P.Val, B: limma topTable columns of
differential ASE. See limma::topTable for details.</p></li>
<li><p>inc/exc_(logFC, AveExpr, t, P.Value, adj.P.Val, B): limma results
for differential testing for raw included / excluded counts only</p></li>
</ul><p><strong>DESeq2 specific output</strong></p>


<ul><li><p>baseMean, log2FoldChange, lfcSE, stat, pvalue, padj:
DESeq2 results columns for differential ASE; see <a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">DESeq2::results</a> for
details.</p></li>
<li><p>inc/exc_(baseMean, log2FoldChange, lfcSE, stat, pvalue, padj):
DESeq2 results for differential testing for
raw included / excluded counts only</p></li>
</ul><p><strong>satuRn specific output</strong></p>


<ul><li><p>estimates, se, df, t, pval, regular_FDR:
estimated log-odds ratio, standard error, degrees of freedom, (Wald) t
statistic, nominal p-value and associated false discovery rate</p></li>
<li><p>empirical_pval, empirical_FDR: nominal p value and associated FDR
computed by estimating the null distribution of the test statistic
empirically (by satuRn).</p></li>
</ul><p><strong>DoubleExp specific output</strong></p>


<ul><li><p>MLE_nom, MLE_denom: Maximum likelihood expectation of PSI values for the
<code>denom</code> in column names are replaced with the condition names</p></li>
<li><p>MLE_LFC: Log2-fold change of the MLE</p></li>
<li><p>P.Value, adj.P.Val: Nominal and BH-adjusted P values</p></li>
<li><p>n_eff: Number of effective samples (i.e. non-zero or non-unity PSI)</p></li>
<li><p>mDepth: Mean Depth of splice coverage in each of the two groups.</p></li>
<li><p>Dispersion_Reduced, Dispersion_Full: Dispersion values for reduced and
full models. See <a href="https://rdrr.io/pkg/DoubleExpSeq/man/DBGLM1.html" class="external-link">DoubleExpSeq::DBGLM1</a> for details.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>Using <strong>limma</strong>, SpliceWiz models included and excluded counts as log-normal
distributed, whereas
using <strong>DESeq2</strong>, SpliceWiz models included and excluded counts as negative
binomial distributed with dispersion shrinkage according to their mean count
expressions.
For <strong>limma</strong> and <strong>DESeq2</strong>, differential ASE are considered as the
"interaction" between included and excluded splice counts for each sample.
See <a href="https://rpubs.com/mikelove/ase" class="external-link">this vignette</a> for an explanation of
how this is done.</p>
<p>SpliceWiz's <strong>limma</strong> wrapper implements an additional filter where ASEs with
an average cpm values of either Included or Excluded counts are less than 1.
<strong>DESeq2</strong> has its own method for handling outliers, which seems to work well
for handling situations where PSI ~= 0 or PSI ~= 1.</p>
<p>Time series are supported by SpliceWiz to a limited extent.
Time series analysis can be performed via limma or DESeq2.
For limma time-series analysis, use <code>ASE_limma_timeseries()</code>, specifying
the <code>test_factor</code> as the column of numeric values containing
time series data. For DESeq, time series differential analysis can be
activated using the <code>ASE_DESeq()</code> function, again specifying <code>test_factor</code>
as the column containing time series data (and leaving <code>test_nom</code>
and <code>test_denom</code> parameters blank). See examples below.</p>
<p>Using <strong>DoubleExpSeq</strong>, included and excluded counts are modeled using
the generalized beta prime distribution, using empirical Bayes shrinkage
to estimate dispersion.</p>
<p>Using <strong>satuRn</strong>, included and excluded counts are modeled using
the quasi-binomial distribution in a generalised linear model.</p>
<p><strong>EventType</strong> are as follow:</p><ul><li><p><code>IR</code> = intron retention (IR-ratio) - all introns are considered</p></li>
<li><p><code>MXE</code> = mutually exclusive exons</p></li>
<li><p><code>SE</code> = skipped exons</p></li>
<li><p><code>AFE</code> = alternate first exon</p></li>
<li><p><code>ALE</code> = alternate last exon</p></li>
<li><p><code>A5SS</code> = alternate 5'-splice site</p></li>
<li><p><code>A3SS</code> = alternate 3'-splice site</p></li>
<li><p><code>RI</code> = (known / annotated) intron retention (PSI).</p></li>
</ul><p>NB: SpliceWiz measures intron retention events using two different
approaches, the choice of which is left to the user - see ASE-methods:</p><ul><li><p><strong>IR</strong> (intron retention) events: considers all introns to be potentially
retained. Given in most scenarios there may be uncertainty as to which of the
many mutually-overlapping introns are spliced to produce the major isoform,
SpliceWiz adopts the IRFinder approach by using the IR-ratio. The "included"
isoform is the relative abundance of the IR-transcript, as approximated by
the trimmed-mean depth of coverage across the intron (excluding outliers
including exons of other transcripts, intronic elements such as snoRNAs,
etc). The "excluded isoform" includes <strong>all</strong> spliced transcripts that
contain an overlapping intron, as estimated via SpliceWiz's <code>SpliceOver</code> and
IRFinder's <code>SpliceMax</code> methods - see <a href="collateData.html">collateData</a>.</p></li>
<li><p><strong>RI</strong> (annotated retained introns) considers only annotated retained
introns, i.e., those annotated within the given reference. These are
quantified using PSI, considering the included (IR-transcript) and excluded
(splicing of the exact intron) as binary alternatives.</p></li>
</ul><p>SpliceWiz considers "included" counts as those that represent abundance of
the "included" isoform, whereas "excluded" counts represent the abundance of
the "excluded" isoform.
To allow comparison between modalities, SpliceWiz applies a convention
whereby the "included" transcript is one where its splice junctions
are by definition shorter than those of "excluded" transcripts.
Specifically, this means the included / excluded isoforms are as follows:</p><table class="table table"><tr><td>EventType</td><td>Included</td><td>Excluded</td></tr><tr><td>IR or RI</td><td>Intron Retention</td><td>Spliced Intron</td></tr><tr><td>MXE</td><td>Upstream exon inclusion</td><td>Downstream exon inclusion</td></tr><tr><td>SE</td><td>Exon inclusion</td><td>Exon skipping</td></tr><tr><td>AFE</td><td>Downstream exon usage</td><td>Upstream exon usage</td></tr><tr><td>ALE</td><td>Upstream exon usage</td><td>Downstream exon usage</td></tr><tr><td>A5SS</td><td>Downstream 5'-SS</td><td>Upstream 5'-SS</td></tr><tr><td>A3SS</td><td>Upstream 3'-SS</td><td>Downstream 3'-SS</td></tr></table></div>
    <div id="functions">
    <h2>Functions</h2>
    
<ul><li><p><code>ASE_limma</code>: Use limma to perform differential ASE analysis of
a filtered NxtSE object</p></li>
<li><p><code>ASE_limma_timeseries</code>: Use limma to perform differential ASE analysis of
a filtered NxtSE object (time series)</p></li>
<li><p><code>ASE_DESeq</code>: Use DESeq2 to perform differential ASE analysis of
a filtered NxtSE object</p></li>
<li><p><code>ASE_DoubleExpSeq</code>: Use DoubleExpSeq to perform differential ASE analysis
of a filtered NxtSE object (uses double exponential beta-binomial model)
to estimate group dispersions, followed by LRT</p></li>
<li><p><code>ASE_satuRn</code>: Use satuRn to perform differential ASE analysis of
a filtered NxtSE object</p></li>
</ul></div>
    <div id="references">
    <h2>References</h2>
    <p>Ritchie ME, Phipson B, Wu D, Hu Y, Law CW, Shi W, Smyth GK (2015).
'limma powers differential expression analyses for RNA-sequencing and
microarray studies.' Nucleic Acids Research, 43(7), e47.
<a href="https://doi.org/10.1093/nar/gkv007" class="external-link">https://doi.org/10.1093/nar/gkv007</a></p>
<p>Love MI, Huber W, Anders S (2014). 'Moderated estimation of fold change and
dispersion for RNA-seq data with DESeq2.' Genome Biology, 15, 550.
<a href="https://doi.org/10.1186/s13059-014-0550-8" class="external-link">https://doi.org/10.1186/s13059-014-0550-8</a></p>
<p>Ruddy S, Johnson M, Purdom E (2016). 'Shrinkage of dispersion parameters in
the binomial family, with application to differential exon skipping.'
Ann. Appl. Stat. 10(2): 690-725.
<a href="https://doi.org/10.1214/15-AOAS871" class="external-link">https://doi.org/10.1214/15-AOAS871</a></p>
<p>Gilis J, Vitting-Seerup K, Van den Berge K, Clement L (2021). 'Scalable
analysis of differential transcript usage for bulk and single-cell
RNA-sequencing applications.' F1000Research 2021, 10:374.
<a href="https://doi.org/10.12688/f1000research.51749.1" class="external-link">https://doi.org/10.12688/f1000research.51749.1</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Load the NxtSE object and set up the annotations</span></span></span>
<span class="r-in"><span><span class="co"># - see ?makeSE on example code of generating this NxtSE object</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="example-SpliceWiz-data.html">SpliceWiz_example_NxtSE</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">colData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">replicate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P"</span>,<span class="st">"Q"</span>,<span class="st">"R"</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Limma analysis (counts modeled using log-normal distribution)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">"limma"</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: limma</span>
<span class="r-in"><span><span class="va">res_limma</span> <span class="op">&lt;-</span> <span class="fu">ASE_limma</span><span class="op">(</span><span class="va">se</span>, <span class="st">"treatment"</span>, <span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:33 Performing limma contrast for included / excluded counts separately</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:34 Performing limma contrast for included / excluded counts together</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># DoubleExpSeq analysis (counts modeled using beta binomial distribution)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"DoubleExpSeq"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: DoubleExpSeq</span>
<span class="r-in"><span><span class="va">res_DES</span> <span class="op">&lt;-</span> <span class="fu">ASE_DoubleExpSeq</span><span class="op">(</span><span class="va">se</span>, <span class="st">"treatment"</span>, <span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:36 Running DoubleExpSeq::DBGLM1() on given data</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># satuRn analysis (quasi binomial), </span></span></span>
<span class="r-in"><span><span class="co">#   Filtering counts using 1 count per million as threshold prior to analysis</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/statOmics/satuRn" class="external-link">"satuRn"</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: satuRn</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"edgeR"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: edgeR</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>there is no package called ‘edgeR’</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">res_sat</span> <span class="op">&lt;-</span> <span class="fu">ASE_satuRn</span><span class="op">(</span><span class="va">se</span>, <span class="st">"treatment"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, filterByMinCPM <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:38 Using MulticoreParam 1 threads</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> Nov 15 16:15:38 edgeR version 3.28.1 is not installed; and is required for this function</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># DESeq2 analysis (counts modeled using negative binomial distribution)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://github.com/mikelove/DESeq2" class="external-link">"DESeq2"</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: DESeq2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: S4Vectors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: stats4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘BiocGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:limma’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     plotMA</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     tapply, union, unique, unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘S4Vectors’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     I, expand.grid, unname</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SummarizedExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MatrixGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biobase</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘Biobase’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:MatrixGenerics’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-in"><span><span class="va">res_DESeq</span> <span class="op">&lt;-</span> <span class="fu">ASE_DESeq</span><span class="op">(</span><span class="va">se</span>, <span class="st">"treatment"</span>, <span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:39 Using MulticoreParam 1 threads</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:39 Performing DESeq2 contrast for included / excluded counts separately</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> converting counts to integer mode</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ASE_DESeq: Profiling expression of Included and Excluded counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates: 1 workers</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates, fitting model and testing: 1 workers</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:42 Performing DESeq2 contrast for included / excluded counts separately</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ASE_DESeq: Profiling differential ASE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using pre-existing size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates: 1 workers</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates, fitting model and testing: 1 workers</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Time series examples</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">timepoint</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"2"</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">res_limma_timeseries</span> <span class="op">&lt;-</span> <span class="fu">ASE_limma_timeseries</span><span class="op">(</span><span class="va">se</span>, <span class="st">"timepoint"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:45 Performing limma contrast for included / excluded counts separately</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:46 Performing limma contrast for included / excluded counts together</span>
<span class="r-in"><span><span class="va">res_DESeq_timeseries</span> <span class="op">&lt;-</span> <span class="fu">ASE_DESeq</span><span class="op">(</span><span class="va">se</span>, <span class="st">"timepoint"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:49 Using MulticoreParam 1 threads</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:49 Performing DESeq2 contrast for included / excluded counts separately</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> converting counts to integer mode</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   the design formula contains one or more numeric variables with integer values,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   specifying a model with increasing fold change for higher values.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   did you mean for this to be a factor? if so, first convert</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   this variable to a factor using the factor() function</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ASE_DESeq: Profiling expression of Included and Excluded counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates: 1 workers</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates, fitting model and testing: 1 workers</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:51 Performing DESeq2 contrast for included / excluded counts separately</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   the design formula contains one or more numeric variables with integer values,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   specifying a model with increasing fold change for higher values.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   did you mean for this to be a factor? if so, first convert</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   this variable to a factor using the factor() function</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ASE_DESeq: Profiling differential ASE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> using pre-existing size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates: 1 workers</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates, fitting model and testing: 1 workers</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alex Chit Hei Wong.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

