<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Builds reference files used by SpliceWiz — Build-Reference-methods • SpliceWiz</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Builds reference files used by SpliceWiz — Build-Reference-methods"><meta property="og:description" content="These function builds the reference required by the SpliceWiz engine, as well
as alternative splicing annotation data for SpliceWiz. See examples
below for guides to making the SpliceWiz reference."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SpliceWiz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/SW_Cookbook.html">SpliceWiz: the cookbook</a>
    </li>
    <li>
      <a href="../articles/SW_QuickStart.html">SpliceWiz: Quick Start</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/alexchwong/SpliceWiz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Builds reference files used by SpliceWiz</h1>
    <small class="dont-index">Source: <a href="https://github.com/alexchwong/SpliceWiz/blob/HEAD/R/BuildRef.R" class="external-link"><code>R/BuildRef.R</code></a></small>
    <div class="hidden name"><code>Build-Reference-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>These function builds the reference required by the SpliceWiz engine, as well
as alternative splicing annotation data for SpliceWiz. See examples
below for guides to making the SpliceWiz reference.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getResources</span><span class="op">(</span></span>
<span>  reference_path <span class="op">=</span> <span class="st">"./Reference"</span>,</span>
<span>  fasta <span class="op">=</span> <span class="st">""</span>,</span>
<span>  gtf <span class="op">=</span> <span class="st">""</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  force_download <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">buildRef</span><span class="op">(</span></span>
<span>  reference_path <span class="op">=</span> <span class="st">"./Reference"</span>,</span>
<span>  fasta <span class="op">=</span> <span class="st">""</span>,</span>
<span>  gtf <span class="op">=</span> <span class="st">""</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  force_download <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  chromosome_aliases <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genome_type <span class="op">=</span> <span class="st">""</span>,</span>
<span>  nonPolyARef <span class="op">=</span> <span class="st">""</span>,</span>
<span>  MappabilityRef <span class="op">=</span> <span class="st">""</span>,</span>
<span>  BlacklistRef <span class="op">=</span> <span class="st">""</span>,</span>
<span>  useExtendedTranscripts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  lowMemoryMode <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">buildFullRef</span><span class="op">(</span></span>
<span>  <span class="va">reference_path</span>,</span>
<span>  <span class="va">fasta</span>,</span>
<span>  <span class="va">gtf</span>,</span>
<span>  chromosome_aliases <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  force_download <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  genome_type <span class="op">=</span> <span class="va">genome_type</span>,</span>
<span>  use_STAR_mappability <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nonPolyARef <span class="op">=</span> <span class="fu">getNonPolyARef</span><span class="op">(</span><span class="va">genome_type</span><span class="op">)</span>,</span>
<span>  BlacklistRef <span class="op">=</span> <span class="st">""</span>,</span>
<span>  useExtendedTranscripts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  n_threads <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">getNonPolyARef</span><span class="op">(</span><span class="va">genome_type</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>reference_path</dt>
<dd><p>(REQUIRED) The directory path to store the generated
reference files</p></dd>


<dt>fasta</dt>
<dd><p>The file path or web link to the user-supplied genome
FASTA file. Alternatively, the name of the AnnotationHub record containing
the genome resource. May be omitted if <code>getResources()</code> has already
been run using the same <code>reference_path</code>.</p></dd>


<dt>gtf</dt>
<dd><p>The file path or web link  to the user-supplied transcript
GTF file (or gzipped GTF file). Alternatively, the name of the
AnnotationHub record containing the transcript GTF file. May be omitted if
<code>getResources()</code> has already been run using the same
<code>reference_path</code>.</p></dd>


<dt>overwrite</dt>
<dd><p>(default <code>FALSE</code>) For <code>getResources()</code>: if the
genome FASTA and gene annotation GTF files already exist in the <code>resource</code>
subdirectory, it will not be overwritten. For <code>buildRef()</code> and
<code>buildFullRef()</code>: the SpliceWiz reference will not be overwritten
if one already exist. A reference is considered to exist if
the file <code>SpliceWiz.ref.gz</code> is present inside <code>reference_path</code>.</p></dd>


<dt>force_download</dt>
<dd><p>(default <code>FALSE</code>) When online resources are retrieved,
a local copy is stored in the <code>SpliceWiz</code> BiocFileCache. Subsequent calls
to the web resource will fetch the local copy. Set <code>force_download</code> to
<code>TRUE</code> will force the resource to be downloaded from the web. Set this to
<code>TRUE</code> only if the web resource has been updated since the last retrieval.</p></dd>


<dt>verbose</dt>
<dd><p>(default <code>TRUE</code>) If <code>FALSE</code>, will silence progress messages</p></dd>


<dt>chromosome_aliases</dt>
<dd><p>(Highly optional) A 2-column data frame containing
chromosome name conversions. If this is set, allows <a href="processBAM.html">processBAM</a> to parse
BAM alignments to a genome whose chromosomes are named
differently to the reference genome. The most common scenario is where
Ensembl genome typically use chromosomes "1", "2", ..., "X", "Y", whereas
UCSC/Gencode genome use "chr1", "chr2", ..., "chrX", "chrY". See example
below. Refer to <a href="https://github.com/dpryan79/ChromosomeMappings" class="external-link">https://github.com/dpryan79/ChromosomeMappings</a> for a
list of chromosome alias resources.</p></dd>


<dt>genome_type</dt>
<dd><p>Allows <code>buildRef()</code> to select default
<code>nonPolyARef</code> and <code>MappabilityRef</code> for selected genomes. Allowed options
are: <code>hg38</code>, <code>hg19</code>, <code>mm10</code>, and <code>mm9</code>.</p></dd>


<dt>nonPolyARef</dt>
<dd><p>(Optional) A BED file of regions defining known
non-polyadenylated transcripts. This file is used for QC analysis
to measure Poly-A enrichment quality of samples. An RDS file (openable
using <code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS()</a></code>) of a GRanges object is acceptable.
If omitted, and <code>genome_type</code> is defined, the default for the specified
genome will be used.</p></dd>


<dt>MappabilityRef</dt>
<dd><p>(Optional) A BED file of low mappability regions due to
repeat elements in the genome. If omitted, the file generated by
<code><a href="Mappability-methods.html">calculateMappability()</a></code> will be used where available, and if
this is not, the default file for the specified <code>genome_type</code> will be used.
If <code>genome_type</code> is not specified, <code>MappabilityRef</code> is not used.
An RDS file (openable using <code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS()</a></code>) of a GRanges object is acceptable.
See details.</p></dd>


<dt>BlacklistRef</dt>
<dd><p>A BED file of regions to be otherwise excluded from IR
analysis. If omitted, a blacklist is not used (this is the default).
An RDS file (openable using <code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS()</a></code>) of a GRanges object is acceptable.</p></dd>


<dt>useExtendedTranscripts</dt>
<dd><p>(default <code>TRUE</code>) Should non-protein-coding
transcripts such as anti-sense and lincRNA transcripts be included in
searching for IR / AS events? Setting <code>FALSE</code> (vanilla IRFinder) will
exclude transcripts other than <code>protein_coding</code> and
<code>processed_transcript</code> transcripts from IR analysis.</p></dd>


<dt>lowMemoryMode</dt>
<dd><p>(default <code>TRUE</code>) By default, SpliceWiz converts FASTA
files to TwoBit, then uses the TwoBit file to fetch genome sequences. In
most cases, this method uses less memory and is faster, but can be very
slow on some systems. Set this option to <code>FALSE</code> (which will convert the
TwoBit file back to FASTA) if you experience
very slow genome fetching (e.g. when annotating splice motifs).</p></dd>


<dt>use_STAR_mappability</dt>
<dd><p>(default FALSE) In <code>buildFullRef()</code>,
whether to run <a href="STAR-methods.html">STAR_mappability</a> to calculate low-mappability regions.
We recommend setting this to <code>FALSE</code> for the common genomes
(human and mouse), and to <code>TRUE</code> for genomes not supported by
<code>genome_type</code>. When set to false, the MappabilityExclusion default file
corresponding to <code>genome_type</code> will automatically be used.</p></dd>


<dt>n_threads</dt>
<dd><p>The number of threads used to generate the STAR reference
and mappability calculations. Multi-threading is not used for SpliceWiz
reference generation (but multiple cores are utilised in data-table
and fst file processing automatically, where available). See <a href="STAR-methods.html">STAR-methods</a></p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>For <code>getResources</code>: creates the following local resources:</p><ul><li><p><code>reference_path/resource/genome.2bit</code>: Local copy of the genome sequences
as a TwoBitFile.</p></li>
<li><p><code>reference_path/resource/transcripts.gtf.gz</code>: Local copy of the gene
annotation as a gzip-compressed file.</p></li>
</ul><p>For <code>buildRef()</code> and <code>buildFullRef()</code>: creates a SpliceWiz reference
which is written to the given directory specified by <code>reference_path</code>.
Files created includes:</p><ul><li><p><code>reference_path/settings.Rds</code>: An RDS file containing parameters used
to generate the SpliceWiz reference</p></li>
<li><p><code>reference_path/SpliceWiz.ref.gz</code>: A gzipped text file containing collated
SpliceWiz reference files. This file is used by <a href="processBAM.html">processBAM</a></p></li>
<li><p><code>reference_path/fst/</code>: Contains fst files for subsequent easy access to
SpliceWiz generated references</p></li>
<li><p><code>reference_path/cov_data.Rds</code>: An RDS file containing data required to
visualise genome / transcript tracks.</p></li>
</ul><p><code>buildFullRef()</code> also creates a <code>STAR</code> reference located in the <code>STAR</code></p>


<p>subdirectory inside the designated <code>reference_path</code></p>



<p>For <code>getNonPolyARef()</code>: Returns the file path to the BED file for
the nonPolyA loci for the specified genome.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><code>getResources()</code> processes the files, downloads resources from
web links or from <code>AnnotationHub()</code>, and saves a local copy in the "resource"
subdirectory within the given <code>reference_path</code>. Resources are retrieved via
either:</p><ol><li><p>User-supplied FASTA and GTF file. This can be a file path, or a web link
(e.g. 'http://', 'https://' or 'ftp://'). Use <code>fasta</code> and <code>gtf</code>
to specify the files or web paths to use.</p></li>
<li><p>AnnotationHub genome and gene annotation (Ensembl): supply the names of
the genome sequence and gene annotations to <code>fasta</code> and <code>gtf</code>.</p></li>
</ol><p><code>buildRef()</code> will first run <code>getResources()</code> if resources are
not yet saved locally (i.e. <code>getResources()</code> is not already run).
Then, it creates the SpliceWiz references. Typical run-times are
5 to 10 minutes for human and mouse genomes (after resources are downloaded).</p>
<p>NB: the parameters <code>fasta</code> and <code>gtf</code> can be omitted in <code>buildRef()</code> if
<code>getResources()</code> is already run.</p>
<p><code>buildFullRef()</code> builds the STAR aligner reference alongside the SpliceWiz
reference. The STAR reference will be located in the <code>STAR</code> subdirectory
of the specified reference path. If <code>use_STAR_mappability</code> is set to <code>TRUE</code>
this function will empirically compute regions of low mappability. This
function requires <code>STAR</code> to be installed on the system (which only runs on
linux-based systems).</p>
<p><code>getNonPolyARef()</code> returns the path of the non-polyA reference file for the
human and mouse genomes.</p>
<p>Typical usage involves running <code>buildRef()</code> for human and mouse genomes
and specifying the <code>genome_type</code> to use the default <code>MappabilityRef</code> and
<code>nonPolyARef</code> files for the specified genome. For non-human non-mouse
genomes, use one of the following alternatives:</p><ul><li><p>Create the SpliceWiz reference without using Mappability Exclusion regions.
To do this, simply run <code>buildRef()</code> and omit <code>MappabilityRef</code>. This is
acceptable assuming the introns assessed are short and do not contain
intronic repeats</p></li>
<li><p>Calculating Mappability Exclusion regions using the STAR aligner,
and building the SpliceWiz reference. This can be done using the
<code>buildFullRef()</code> function, on systems where <code>STAR</code> is installed</p></li>
<li><p>Instead of using the STAR aligner, any genome splice-aware aligner could be
used. See <a href="Mappability-methods.html">Mappability-methods</a> for
an example workflow using the Rsubread aligner. After producing the
<code>MappabilityExclusion.bed.gz</code> file (in the <code>Mappability</code> subfolder), run
<code>buildRef()</code> using this file (or simply leave it blank).</p></li>
</ul><p>BED files are tab-separated text files containing 3 unnamed columns
specifying chromosome, start and end coordinates. To view an example BED
file, open the file specified in the path returned by
<code>getNonPolyARef("hg38")</code></p>
<p>See examples below for common use cases.</p>
    </div>
    <div id="functions">
    <h2>Functions</h2>
    
<ul><li><p><code>getResources</code>: Processes / downloads a copy of the
genome and gene annotations and stores this in the "resource" subdirectory
of the given reference path</p></li>
<li><p><code>buildRef</code>: First calls <code>getResources()</code>
(if required). Afterwards creates the SpliceWiz reference in the
given reference path</p></li>
<li><p><code>buildFullRef</code>: One-step function that fetches resources,
creates a STAR reference (including mappability calculations), then
creates the SpliceWiz reference</p></li>
<li><p><code>getNonPolyARef</code>: Returns the path to the BED file
containing coordinates of known non-polyadenylated transcripts for genomes
<code>hg38</code>, <code>hg19</code>, <code>mm10</code> and <code>mm9</code>,</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="Mappability-methods.html">Mappability-methods</a> for methods to calculate low mappability regions<br><br><a href="STAR-methods.html">STAR-methods</a> for a list of STAR wrapper functions<br><br><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a><br><br><a href="https://github.com/alexchwong/SpliceWizResources" class="external-link">https://github.com/alexchwong/SpliceWizResources</a> for RDS files of
Mappability Exclusion GRanges objects (for hg38, hg19, mm10 and mm9)
that can be use as input files for <code>MappabilityRef</code> in <code>buildRef()</code>.
These resources are intended for SpliceWiz users on older Bioconductor
versions (3.13 or earlier)</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Quick runnable example: generate a reference using SpliceWiz's example genome</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">example_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"Reference"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">getResources</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="va">example_ref</span>,</span></span>
<span class="r-in"><span>    fasta <span class="op">=</span> <span class="fu">chrZ_genome</span><span class="op">(</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    gtf <span class="op">=</span> <span class="fu">chrZ_gtf</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:58 Converting FASTA to local TwoBitFile...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:59 Connecting to genome TwoBitFile...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:59 Making local copy of GTF file...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-in"><span><span class="fu">buildRef</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="va">example_ref</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:59 Reference generated without non-polyA reference</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:59 Reference generated without Mappability reference</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:59 Reference generated without Blacklist exclusion</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:59 Connecting to genome TwoBitFile...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:59 Reading source GTF file...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:15:59 Processing gtf file...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...genes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...transcripts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...CDS</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...exons</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:01 Processing introns...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...basic annotations</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...splice motifs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...other info</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...defining flanking exon clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:03 Generating processBAM reference</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...prepping data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...determining measurable introns (directional)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...determining measurable introns (non-directional)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...writing ref-cover.bed</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...writing ref-ROI.bed</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...writing ref-read-continues.ref</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...writing ref-sj.ref</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...writing ref-tj.ref</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> processBAM reference generated</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:08 Predicting NMD transcripts from genome sequence</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...exonic transcripts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...retained introns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |                                                                      |   0%  |                                                                              |======================================================================| 100%</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:09 Annotating Splice Events</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Annotating Mutually-Exclusive-Exon Splice Events...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Annotating Skipped-Exon Splice Events...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Annotating Alternate 5' / 3' Splice Site Splice Events...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Annotating Alternate First / Last Exon Splice Events...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Annotating known retained introns...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:10 Splice Annotations Filtered</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:10 Translating Alternate Splice Peptides...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:10 Splice Annotations finished</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reference build finished</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># NB: the above is equivalent to:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">example_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"Reference"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">buildRef</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="va">example_ref</span>,</span></span>
<span class="r-in"><span>    fasta <span class="op">=</span> <span class="fu">chrZ_genome</span><span class="op">(</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    gtf <span class="op">=</span> <span class="fu">chrZ_gtf</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:10 SpliceWiz reference already exists in given directory</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get the path to the Non-PolyA BED file for hg19</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">getNonPolyARef</span><span class="op">(</span><span class="st">"hg19"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "/private/var/folders/xy/6ybn2drx76lblc_fqhfwt8qw0000gr/T/RtmpNHD0Jh/temp_libpathb2037364a2f/SpliceWiz/extra-input-files/Human_hg19_nonPolyA_ROI.bed"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">### Long examples ###</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Generate a SpliceWiz reference from user supplied FASTA and GTF files for a</span></span></span>
<span class="r-in"><span><span class="co"># hg38-based genome:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">buildRef</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="st">"./Reference_user"</span>,</span></span>
<span class="r-in"><span>    fasta <span class="op">=</span> <span class="st">"genome.fa"</span>, gtf <span class="op">=</span> <span class="st">"transcripts.gtf"</span>,</span></span>
<span class="r-in"><span>    genome_type <span class="op">=</span> <span class="st">"hg38"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># NB: Setting `genome_type = hg38`, will automatically use default</span></span></span>
<span class="r-in"><span><span class="co"># nonPolyARef and MappabilityRef for `hg38`</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Reference generation from Ensembl's FTP links:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">FTP</span> <span class="op">&lt;-</span> <span class="st">"ftp://ftp.ensembl.org/pub/release-94/"</span></span></span>
<span class="r-in"><span><span class="fu">buildRef</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="st">"./Reference_FTP"</span>,</span></span>
<span class="r-in"><span>    fasta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">FTP</span>, <span class="st">"fasta/homo_sapiens/dna/"</span>,</span></span>
<span class="r-in"><span>        <span class="st">"Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    gtf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">FTP</span>, <span class="st">"gtf/homo_sapiens/"</span>,</span></span>
<span class="r-in"><span>        <span class="st">"Homo_sapiens.GRCh38.94.chr.gtf.gz"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    genome_type <span class="op">=</span> <span class="st">"hg38"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get AnnotationHub record names for Ensembl release-94:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># First, search for the relevant AnnotationHub record names:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu">AnnotationHub</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">AnnotationHub</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">ah</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Homo Sapiens"</span>, <span class="st">"release-94"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">buildRef</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="st">"./Reference_AH"</span>,</span></span>
<span class="r-in"><span>    fasta <span class="op">=</span> <span class="st">"AH65745"</span>,</span></span>
<span class="r-in"><span>    gtf <span class="op">=</span> <span class="st">"AH64631"</span>,</span></span>
<span class="r-in"><span>    genome_type <span class="op">=</span> <span class="st">"hg38"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Build a SpliceWiz reference, setting chromosome aliases to allow</span></span></span>
<span class="r-in"><span><span class="co"># this reference to process BAM files aligned to UCSC-style genomes:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">chrom.df</span> <span class="op">&lt;-</span> <span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">genomeStyles</a></span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">Homo_sapiens</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">buildRef</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="st">"./Reference_UCSC"</span>,</span></span>
<span class="r-in"><span>    fasta <span class="op">=</span> <span class="st">"AH65745"</span>,</span></span>
<span class="r-in"><span>    gtf <span class="op">=</span> <span class="st">"AH64631"</span>,</span></span>
<span class="r-in"><span>    genome_type <span class="op">=</span> <span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>    chromosome_aliases <span class="op">=</span> <span class="va">chrom.df</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Ensembl"</span>, <span class="st">"UCSC"</span><span class="op">)</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># One-step generation of SpliceWiz and STAR references, using 4 threads.</span></span></span>
<span class="r-in"><span><span class="co"># NB1: requires a linux-based system with STAR installed.</span></span></span>
<span class="r-in"><span><span class="co"># NB2: A STAR reference genome will be generated in the `STAR` subfolder</span></span></span>
<span class="r-in"><span><span class="co">#      inside the given `reference_path`.</span></span></span>
<span class="r-in"><span><span class="co"># NB3: A custom Mappability Exclusion file will be calculated using STAR</span></span></span>
<span class="r-in"><span><span class="co">#      and will be used to generate the SpliceWiz reference.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">buildFullRef</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="st">"./Reference_with_STAR"</span>,</span></span>
<span class="r-in"><span>    fasta <span class="op">=</span> <span class="st">"genome.fa"</span>, gtf <span class="op">=</span> <span class="st">"transcripts.gtf"</span>,</span></span>
<span class="r-in"><span>    genome_type <span class="op">=</span> <span class="st">""</span>,</span></span>
<span class="r-in"><span>    use_STAR_mappability <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    n_threads <span class="op">=</span> <span class="fl">4</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># NB: the above is equivalent to running the following in sequence:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">getResources</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="st">"./Reference_with_STAR"</span>,</span></span>
<span class="r-in"><span>    fasta <span class="op">=</span> <span class="st">"genome.fa"</span>, gtf <span class="op">=</span> <span class="st">"transcripts.gtf"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="STAR-methods.html">STAR_buildRef</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="va">reference_path</span>,</span></span>
<span class="r-in"><span>    also_generate_mappability <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    n_threads <span class="op">=</span> <span class="fl">4</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">buildRef</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="st">"./Reference_with_STAR"</span>,</span></span>
<span class="r-in"><span>    genome_type <span class="op">=</span> <span class="st">""</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alex Chit Hei Wong.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

