<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Collates a dataset from (processBAM) output files of individual samples — collateData • SpliceWiz</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Collates a dataset from (processBAM) output files of individual samples — collateData"><meta property="og:description" content="collateData() creates a dataset from a collection of processBAM
output files belonging to an experiment."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SpliceWiz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/SW_Cookbook.html">SpliceWiz: the cookbook</a>
    </li>
    <li>
      <a href="../articles/SW_QuickStart.html">SpliceWiz: Quick Start</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/alexchwong/SpliceWiz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Collates a dataset from (processBAM) output files of individual samples</h1>
    <small class="dont-index">Source: <a href="https://github.com/alexchwong/SpliceWiz/blob/HEAD/R/CollateData.R" class="external-link"><code>R/CollateData.R</code></a></small>
    <div class="hidden name"><code>collateData.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>collateData()</code> creates a dataset from a collection of <a href="processBAM.html">processBAM</a>
output files belonging to an experiment.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">collateData</span><span class="op">(</span></span>
<span>  <span class="va">Experiment</span>,</span>
<span>  <span class="va">reference_path</span>,</span>
<span>  <span class="va">output_path</span>,</span>
<span>  IRMode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SpliceOver"</span>, <span class="st">"SpliceMax"</span><span class="op">)</span>,</span>
<span>  novelSplicing <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  forceStrandAgnostic <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  novelSplicing_minSamples <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  novelSplicing_countThreshold <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  novelSplicing_minSamplesAboveThreshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  novelSplicing_requireOneAnnotatedSJ <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_threads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  lowMemoryMode <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>Experiment</dt>
<dd><p>(Required) A 2 or 3 column data frame, ideally generated by
<a href="findSamples.html">findSpliceWizOutput</a> or <a href="findSamples.html">findSamples</a>.
The first column designate the sample names, and the 2nd column
contains the path to the <a href="processBAM.html">processBAM</a> output file (of type
<code>sample.txt.gz</code>).
(Optionally) a 3rd column contains the coverage files (of type
<code>sample.cov</code>) of the corresponding samples.
NB: all other columns are ignored.</p></dd>


<dt>reference_path</dt>
<dd><p>(Required) The path to the reference generated by
<a href="Build-Reference-methods.html">Build-Reference-methods</a></p></dd>


<dt>output_path</dt>
<dd><p>(Required)  The path to contain the output files for the
collated dataset</p></dd>


<dt>IRMode</dt>
<dd><p>(default <code>SpliceOver</code>) The algorithm to calculate
'splice abundance' in IR quantification.
Valid options are <code>SpliceOver</code> and <code>SpliceMax</code>.
See details</p></dd>


<dt>novelSplicing</dt>
<dd><p>(default FALSE) Whether collateData will use
novel junction reads detected in samples to infer novel splice variants.
All tandem split reads (those bridging two consecutive splice junctions)
are used, as well as novel split reads that satisfy abundance criteria
(see <code>novelSplicing_minSamples</code>, <code>novelSplicing_minSamplesAboveThreshold</code>, and
<code>novelSplicing_countThreshold</code>) are used to synthesise a dataset-specific
SpliceWiz reference. See details.</p></dd>


<dt>forceStrandAgnostic</dt>
<dd><p>(default <code>FALSE</code>) In poorly-prepared stranded
libraries, it may be better to quantify in unstranded mode. Set this to
<code>TRUE</code> if your stranded libraries may be contaminated with unstranded reads</p></dd>


<dt>novelSplicing_minSamples</dt>
<dd><p>(default 3) Novel junctions are included in
building of novel reference  if number samples with non-zero counts
exceeds this number.</p></dd>


<dt>novelSplicing_countThreshold</dt>
<dd><p>(default 10) Threshold of split-reads across
novel junctions; used in conjunction with
<code>novelSplicing_minSamplesAboveThreshold</code></p></dd>


<dt>novelSplicing_minSamplesAboveThreshold</dt>
<dd><p>(default 1) Novel junctions are
included in building of novel reference if novel junction reads are
above a pre-defined threshold exceeds this number</p></dd>


<dt>novelSplicing_requireOneAnnotatedSJ</dt>
<dd><p>(default <code>TRUE</code>) The default requires
novel junctions to have one annotated splice site. If this is disabled,
collateData will include novel junctions where neither splice site is
annotated.</p></dd>


<dt>overwrite</dt>
<dd><p>(default <code>FALSE</code>) If <code>collateData()</code> has previously been run
using the same set of samples, it will not be overwritten unless this is
set to <code>TRUE</code>.</p></dd>


<dt>n_threads</dt>
<dd><p>(default <code>1</code>) The number of threads to use. If you run out
of memory, try lowering the number of threads</p></dd>


<dt>lowMemoryMode</dt>
<dd><p>(default <code>TRUE</code>) <code>collateData()</code> will perform
optimizations to conserve memory if this is set to <code>TRUE</code>. Otherwise,
will prioritise performance.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code>collateData()</code> writes to the directory given by <code>output_path</code>.
This output directory is portable (i.e. it can be moved to a different
location after running <code>collateData()</code> before running <a href="makeSE.html">makeSE</a>), but
individual files within the output folder should not be moved.<br><br></p>


<p>Also, the <a href="processBAM.html">processBAM</a> and collateData output folders should be copied to
the same destination and their relative paths preserved. Otherwise, the
locations of the "COV" files will not be recorded in the collated data and
will have to be re-assigned using <code>covfile(se)&lt;-</code>. See <a href="makeSE.html">makeSE</a></p>


    </div>
    <div id="details">
    <h2>Details</h2>
    <p>All sample <a href="processBAM.html">processBAM</a> outputs must be generated using the same
reference.</p>
<p>The combination of junction counts and IR quantification from
<a href="processBAM.html">processBAM</a> is used to calculate percentage spliced in (PSI) of alternative
splice events, and intron retention ratios (IR-ratio) of retained introns.
Also, QC information is collated. Data is organised in a H5file and FST files
for memory and processor efficient downstream access using <a href="makeSE.html">makeSE</a>.</p>
<p>The original IRFinder algorithm, see the following
<a href="https://github.com/williamritchie/IRFinder/wiki/IRFinder-Output" class="external-link">wiki</a>,
uses <code>SpliceMax</code> to estimate abundance of spliced transcripts.
This calculates the number of mapped splice events
that share the boundary coordinate of either the left or right flanking
exon <code>SpliceLeft,SpliceRight</code>, estimating splice abundance as the larger
of the two values.</p>
<p>SpliceWiz proposes a new algorithm, <code>SpliceOver</code>,
to account for the possibility that the major isoform shares neither
boundary, but arises from either of the flanking exon clusters. Exon
clusters are contiguous regions covered by exons from any transcript
(except those designated as <code>retained_intron</code> or
<code>sense_intronic</code>), and are separated by
obligate intronic regions (genomic regions that are introns for all
transcripts). For introns that are internal to a single exon cluster
(i.e. akin to "known-exon" introns from IRFinder), <code>SpliceOver</code>
uses <a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html" class="external-link">GenomicRanges::findOverlaps</a> to sum all splice reads that overlap
the same genomic region as the intron of interest.</p>
<p>Detection of novel ASEs:  When <code>novelSplicing</code> is set to <code>TRUE</code>,
novel junctions (split reads across unannotated junctions from samples
of the dataset being collated) are used in conjunction with the reference
to compile a list of novel ASEs. To avoid being overwhelmed by a large
number of false positive novel junctions (often due to mis-alignments),
a simple filtering strategy is used. This involves including novel
junctions only if it occurs in a minimum number of samples (default 3),
or if the number of split reads of a novel junction is above a pre-defined
threshold (default 10) in a certain number of samples (default 1). These
parameters can be set using <code>novelSplicing_minSamples</code>,
<code>novelSplicing_countThreshold</code> and <code>novelSplicing_minSamplesAboveThreshold</code>
respectively.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="processBAM.html">processBAM</a>, <a href="makeSE.html">makeSE</a></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="Build-Reference-methods.html">buildRef</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"Reference"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    fasta <span class="op">=</span> <span class="fu">chrZ_genome</span><span class="op">(</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    gtf <span class="op">=</span> <span class="fu">chrZ_gtf</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:54 SpliceWiz reference already exists in given directory</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">bams</span> <span class="op">&lt;-</span> <span class="fu"><a href="example-SpliceWiz-data.html">SpliceWiz_example_bams</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="processBAM.html">processBAM</a></span><span class="op">(</span><span class="va">bams</span><span class="op">$</span><span class="va">path</span>, <span class="va">bams</span><span class="op">$</span><span class="va">sample</span>,</span></span>
<span class="r-in"><span>  reference_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"Reference"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  output_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"SpliceWiz_Output"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:54 processBAM has already been run on given BAM files</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="findSamples.html">findSpliceWizOutput</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"SpliceWiz_Output"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">collateData</span><span class="op">(</span><span class="va">expr</span>,</span></span>
<span class="r-in"><span>  reference_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"Reference"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  output_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"Collated_output"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:54 Using MulticoreParam 1 threads</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:54 Validating Experiment; checking COV files...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:54 SpliceWiz already collated this experiment in given directory</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alex Chit Hei Wong.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

