<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calls SpliceWiz's C++ function to retrieve coverage from a COV file — Coverage • SpliceWiz</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Calls SpliceWiz's C++ function to retrieve coverage from a COV file — Coverage"><meta property="og:description" content="This function returns an RLE / RLEList or data.frame
containing coverage data from the given COV file
COV files are generated by SpliceWiz's processBAM and BAM2COV functions.
It records alignment coverage for each nucleotide in the given BAM file.
It stores this data in &quot;COV&quot; format, which is an indexed BGZF-compressed
format specialised for the storage of unstranded and stranded alignment
coverage in RNA sequencing.
Unlike BigWig files, COV files store coverage for both positive and negative
strands.
These functions retrieves coverage data from the specified COV file. They
are computationally efficient as they utilise random-access to rapidly
search for the requested data from the COV file."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SpliceWiz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/SW_Cookbook.html">SpliceWiz: the cookbook</a>
    </li>
    <li>
      <a href="../articles/SW_QuickStart.html">SpliceWiz: Quick Start</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/alexchwong/SpliceWiz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calls SpliceWiz's C++ function to retrieve coverage from a COV file</h1>
    <small class="dont-index">Source: <a href="https://github.com/alexchwong/SpliceWiz/blob/HEAD/R/Coverage.R" class="external-link"><code>R/Coverage.R</code></a></small>
    <div class="hidden name"><code>Coverage.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function returns an RLE / RLEList or data.frame
containing coverage data from the given COV file<br><br>
COV files are generated by SpliceWiz's <a href="processBAM.html">processBAM</a> and <a href="processBAM.html">BAM2COV</a> functions.
It records alignment coverage for each nucleotide in the given BAM file.
It stores this data in "COV" format, which is an indexed BGZF-compressed
format specialised for the storage of unstranded and stranded alignment
coverage in RNA sequencing.<br><br>
Unlike BigWig files, COV files store coverage for both positive and negative
strands.<br><br>
These functions retrieves coverage data from the specified COV file. They
are computationally efficient as they utilise random-access to rapidly
search for the requested data from the COV file.<br><br></p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getCoverage</span><span class="op">(</span><span class="va">file</span>, seqname <span class="op">=</span> <span class="st">""</span>, start <span class="op">=</span> <span class="fl">0</span>, end <span class="op">=</span> <span class="fl">0</span>, strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"*"</span>, <span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">getCoverage_DF</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  seqname <span class="op">=</span> <span class="st">""</span>,</span>
<span>  start <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  end <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"*"</span>, <span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">getCoverageRegions</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  strandMode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"unstranded"</span>, <span class="st">"forward"</span>, <span class="st">"reverse"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">getCoverageBins</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  <span class="va">region</span>,</span>
<span>  bins <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  strandMode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"unstranded"</span>, <span class="st">"forward"</span>, <span class="st">"reverse"</span><span class="op">)</span>,</span>
<span>  <span class="va">bin_size</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>file</dt>
<dd><p>(Required) The file name of the COV file</p></dd>


<dt>seqname</dt>
<dd><p>(Required for <code>getCoverage_DF</code>) A string denoting the
chromosome name. If left blank in <code>getCoverage</code>, retrieves RLEList
containing coverage of the entire file.</p></dd>


<dt>start, end</dt>
<dd><p>1-based genomic coordinates. If <code>start = 0</code> and
<code>end = 0</code>, will retrieve RLE of specified chromosome.</p></dd>


<dt>strand</dt>
<dd><p>Either "*", "+", or "-"</p></dd>


<dt>regions</dt>
<dd><p>A GRanges object for a set of regions to obtain mean / total
coverage from the given COV file.</p></dd>


<dt>strandMode</dt>
<dd><p>The stranded-ness of the RNA-seq experiment. "unstranded"
means that an unstranded protocol was used. Stranded protocols can be
either "forward", where the first read is the same strand as the
expressed transcript, or "reverse" where the second strand is the same
strand as the expressed transcript.</p></dd>


<dt>region</dt>
<dd><p>In <code>getCoverageBins</code>, a single query region as a GRanges object</p></dd>


<dt>bins</dt>
<dd><p>In <code>getCoverageBins</code>, the number of bins to divide the given
<code>region</code>. If <code>bin_size</code> is given, overrides this parameter</p></dd>


<dt>bin_size</dt>
<dd><p>In <code>getCoverageBins</code>, the number of nucleotides per bin</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>For <code>getCoverage</code>: If seqname is left as "", returns an RLEList of the
whole BAM file, with each RLE in the list containing coverage data for
one chromosome. Otherwise, returns an RLE containing coverage data for
the requested genomic region</p>


<p>For <code>getCoverage_DF</code>: Returns a two-column data frame, with the first column
<code>coordinate</code> denoting genomic coordinate, and the second column <code>value</code></p>


<p>containing the coverage depth for each coordinate nucleotide.</p>


<p>For <code>getCoverageRegions</code>: Returns a GRanges object with an extra metacolumn:
<code>cov_mean</code>, which gives the mean coverage of each of the given ranges.</p>


<p>For <code>getCoverageBins</code>: Returns a GRanges object which spans the given
<code>region</code>, divided by the number of <code>bins</code> or by width as given by
<code>bin_size</code>. Mean coverage in each bin is calculated (returned by the
<code>cov_mean</code> metadata column). This function is useful for retrieving
coverage of a large region for visualisation, especially when the
size of the region vastly exceeds the width of the figure.</p>
    </div>
    <div id="functions">
    <h2>Functions</h2>
    
<ul><li><p><code>getCoverage</code>: Retrieves alignment coverage as an RLE or RLElist</p></li>
<li><p><code>getCoverage_DF</code>: Retrieves alignment coverage as a data.frame</p></li>
<li><p><code>getCoverageRegions</code>: Retrieves total and mean coverage of a GRanges object
from a COV file</p></li>
<li><p><code>getCoverageBins</code>: Retrieves coverage of a single region from a COV file,
binned by the given number of bins or bin_size</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="example-SpliceWiz-data.html">SpliceWiz_example_NxtSE</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">cov_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="NxtSE-class.html">covfile</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Retrieve Coverage as RLE</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">cov</span> <span class="op">&lt;-</span> <span class="fu">getCoverage</span><span class="op">(</span><span class="va">cov_file</span>, seqname <span class="op">=</span> <span class="st">"chrZ"</span>,</span></span>
<span class="r-in"><span>  start <span class="op">=</span> <span class="fl">10000</span>, end <span class="op">=</span> <span class="fl">20000</span>,</span></span>
<span class="r-in"><span>  strand <span class="op">=</span> <span class="st">"*"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Retrieve Coverage as data.frame</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">cov.df</span> <span class="op">&lt;-</span> <span class="fu">getCoverage_DF</span><span class="op">(</span><span class="va">cov_file</span>, seqname <span class="op">=</span> <span class="st">"chrZ"</span>,</span></span>
<span class="r-in"><span>  start <span class="op">=</span> <span class="fl">10000</span>, end <span class="op">=</span> <span class="fl">20000</span>,</span></span>
<span class="r-in"><span>  strand <span class="op">=</span> <span class="st">"*"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Retrieve mean coverage of 100-nt window regions as defined</span></span></span>
<span class="r-in"><span><span class="co"># in a GRanges object:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    seqnames <span class="op">=</span> <span class="st">"chrZ"</span>,</span></span>
<span class="r-in"><span>    ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>        start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">99901</span>, by <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>        end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">100000</span>, by <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">)</span>, strand <span class="op">=</span> <span class="st">"-"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gr.unstranded</span> <span class="op">&lt;-</span> <span class="fu">getCoverageRegions</span><span class="op">(</span><span class="va">cov_file</span>,</span></span>
<span class="r-in"><span>    regions <span class="op">=</span> <span class="va">gr</span>,</span></span>
<span class="r-in"><span>    strandMode <span class="op">=</span> <span class="st">"unstranded"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gr.stranded</span> <span class="op">&lt;-</span> <span class="fu">getCoverageRegions</span><span class="op">(</span><span class="va">cov_file</span>,</span></span>
<span class="r-in"><span>    regions <span class="op">=</span> <span class="va">gr</span>,</span></span>
<span class="r-in"><span>    strandMode <span class="op">=</span> <span class="st">"reverse"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Retrieve binned coverage of a large region</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gr.fetch</span> <span class="op">&lt;-</span> <span class="fu">getCoverageBins</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">cov_file</span>,</span></span>
<span class="r-in"><span>    region <span class="op">=</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="st">"chrZ"</span>,</span></span>
<span class="r-in"><span>        ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fl">100</span>, end <span class="op">=</span> <span class="fl">100000</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>        strand <span class="op">=</span> <span class="st">"*"</span></span></span>
<span class="r-in"><span>    <span class="op">)</span>,</span></span>
<span class="r-in"><span>    bins <span class="op">=</span> <span class="fl">2000</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Plot coverage using ggplot:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: ggplot2</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">cov.df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">coordinate</span>, y <span class="op">=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="va">theme_white</span></span></span>
<span class="r-plt img"><img src="Coverage-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">gr.unstranded</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="op">(</span><span class="va">start</span> <span class="op">+</span> <span class="va">end</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span>, y <span class="op">=</span> <span class="va">cov_mean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="va">theme_white</span></span></span>
<span class="r-plt img"><img src="Coverage-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">gr.fetch</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="op">(</span><span class="va">start</span> <span class="op">+</span> <span class="va">end</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, y <span class="op">=</span> <span class="va">cov_mean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="va">theme_white</span></span></span>
<span class="r-plt img"><img src="Coverage-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Export COV data as BigWig</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">cov_whole</span> <span class="op">&lt;-</span> <span class="fu">getCoverage</span><span class="op">(</span><span class="va">cov_file</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bw_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"sample.bw"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">export</span><span class="op">(</span><span class="va">cov_whole</span>, <span class="va">bw_file</span>, <span class="st">"bw"</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alex Chit Hei Wong.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

