<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate low mappability genomic regions — Mappability-methods • SpliceWiz</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate low mappability genomic regions — Mappability-methods"><meta property="og:description" content="These functions empirically calculate low-mappability (Mappability Exclusion)
regions using the given reference. A splice-aware alignment software
capable of aligning reads to the genome is required.
See details and examples below."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SpliceWiz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/SW_Cookbook.html">SpliceWiz: the cookbook</a>
    </li>
    <li>
      <a href="../articles/SW_QuickStart.html">SpliceWiz: Quick Start</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/alexchwong/SpliceWiz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate low mappability genomic regions</h1>
    <small class="dont-index">Source: <a href="https://github.com/alexchwong/SpliceWiz/blob/HEAD/R/Mappability.R" class="external-link"><code>R/Mappability.R</code></a></small>
    <div class="hidden name"><code>Mappability-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>These functions empirically calculate low-mappability (Mappability Exclusion)
regions using the given reference. A splice-aware alignment software
capable of aligning reads to the genome is required.
See details and examples below.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">generateSyntheticReads</span><span class="op">(</span></span>
<span>  <span class="va">reference_path</span>,</span>
<span>  read_len <span class="op">=</span> <span class="fl">70</span>,</span>
<span>  read_stride <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  error_pos <span class="op">=</span> <span class="fl">35</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">alt_fasta_file</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">calculateMappability</span><span class="op">(</span></span>
<span>  <span class="va">reference_path</span>,</span>
<span>  aligned_bam <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">reference_path</span>, <span class="st">"Mappability"</span>, <span class="st">"Aligned.out.bam"</span><span class="op">)</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  n_threads <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>reference_path</dt>
<dd><p>The directory of the reference prepared by
<a href="Build-Reference-methods.html">getResources</a></p></dd>


<dt>read_len</dt>
<dd><p>The nucleotide length of the synthetic reads</p></dd>


<dt>read_stride</dt>
<dd><p>The nucleotide distance between adjacent synthetic reads</p></dd>


<dt>error_pos</dt>
<dd><p>The position of the procedurally-generated nucleotide error
from the start of each synthetic reads</p></dd>


<dt>verbose</dt>
<dd><p>Whether additional status messages are shown</p></dd>


<dt>alt_fasta_file</dt>
<dd><p>(Optional) The path to the user-supplied genome fasta
file, if different to that found inside the <code>resource</code> subdirectory of the
<code>reference_path</code>. If <a href="Build-Reference-methods.html">getResources</a> has already been run,
this parameter should be omitted.</p></dd>


<dt>aligned_bam</dt>
<dd><p>The BAM file of alignment of the synthetic reads generated
by <code>generateSyntheticReads()</code>. Users should use a genome splice-aware
aligner, preferably the same aligner used to align the samples in their
experiment.</p></dd>


<dt>threshold</dt>
<dd><p>Genomic regions with this alignment read depth (or below)
in the aligned synthetic read BAM are defined as low
mappability regions.</p></dd>


<dt>n_threads</dt>
<dd><p>The number of threads used to calculate mappability
exclusion regions from aligned bam file of synthetic reads.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<ul><li><p>For <code>generateSyntheticReads</code>: writes <code>Reads.fa</code> to the <code>Mappability</code>
subdirectory inside the given <code>reference_path</code>.</p></li>
<li><p>For <code>calculateMappability</code>: writes a gzipped BED file
named <code>MappabilityExclusion.bed.gz</code> to the <code>Mappability</code> subdirectory
inside <code>reference_path</code>.
This BED file is automatically used by <a href="Build-Reference-methods.html">buildRef</a> if its
<code>MappabilityRef</code> parameter is not specified.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>Creating a Mappability Exclusion BED file is a three-step process.
<br></p><ul><li><p>First, using <code>generateSyntheticReads()</code>, synthetic reads are systematically
generated using the given genome contained within <code>reference_path</code>, prepared
via <a href="Build-Reference-methods.html">getResources</a>. Alternatively, use <code>alt_fasta_file</code> to set the genome
sequence if this is different to that prepared by <code>getResources</code> or if
<code>getResources</code> is not yet run.</p></li>
<li><p>Second, an aligner such as STAR (preferably the same aligner used for the
subsequent RNA-seq experiment) is required to align these reads to the source
genome. Poorly mapped regions of the genome will be reflected by regions of
low coverage depth.</p></li>
<li><p>Finally, the BAM file containing the aligned reads is analysed
using <code>calculateMappability()</code>, to identify
low-mappability regions to compile the Mappability Exclusion BED file.</p></li>
</ul><p>It is recommended to leave all parameters to their default settings. Regular
users should only specify <code>reference_path</code>, <code>aligned_bam</code> and <code>n_threads</code>,
as required.</p>
<p>NB: <a href="STAR-methods.html">STAR_mappability</a> runs all 3 steps required, using the <code>STAR</code> aligner.
This only works in systems where <code>STAR</code> is installed.</p>
<p>NB2: <a href="Build-Reference-methods.html">buildFullRef</a> builds the STAR reference, then calculates mappability.
It then uses the calculated mappability regions to build the SpliceWiz
reference.</p>
<p>NB3: In systems where <code>STAR</code> is not available, consider using HISAT2 or
Rsubread. A working example using Rsubread is shown below.</p>
    </div>
    <div id="functions">
    <h2>Functions</h2>
    
<ul><li><p><code>generateSyntheticReads</code>: Generates synthetic reads from a
genome FASTA file, for mappability calculations.</p></li>
<li><p><code>calculateMappability</code>: Generate a BED file defining
low mappability regions, using reads generated by
<code>generateSyntheticReads()</code>, aligned to the genome.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="Build-Reference-methods.html">Build-Reference-methods</a></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># (1a) Creates genome resource files</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">ref_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"refWithMapExcl"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="Build-Reference-methods.html">getResources</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span></span>
<span class="r-in"><span>    fasta <span class="op">=</span> <span class="fu">chrZ_genome</span><span class="op">(</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    gtf <span class="op">=</span> <span class="fu">chrZ_gtf</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:16 Converting FASTA to local TwoBitFile...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:17 Connecting to genome TwoBitFile...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:17 Making local copy of GTF file...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># (1b) Systematically generate reads based on the example genome:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">generateSyntheticReads</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="va">ref_path</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:17 Extracting temp genome FASTA from TwoBit file</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:17 Generating synthetic reads, saving to /private/var/folders/xy/6ybn2drx76lblc_fqhfwt8qw0000gr/T/Rtmpd5xBOp/refWithMapExcl/Mappability/Reads.fa</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10145 synthetic reads generated</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:17 Cleaning temp genome / gene annotation files</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># (2) Align the generated reads using Rsubread:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># (2a) Build the Rsubread genome index:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">ref_path</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">Rsubread</span><span class="fu">::</span><span class="fu">buildindex</span><span class="op">(</span>basename <span class="op">=</span> <span class="st">"./reference_index"</span>,</span></span>
<span class="r-in"><span>    reference <span class="op">=</span> <span class="fu">chrZ_genome</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># (2b) Align the synthetic reads using Rsubread::subjunc()</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">Rsubread</span><span class="fu">::</span><span class="fu">subjunc</span><span class="op">(</span></span></span>
<span class="r-in"><span>    index <span class="op">=</span> <span class="st">"./reference_index"</span>,</span></span>
<span class="r-in"><span>    readfile1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ref_path</span>, <span class="st">"Mappability"</span>, <span class="st">"Reads.fa"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    output_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ref_path</span>, <span class="st">"Mappability"</span>, <span class="st">"AlignedReads.bam"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    useAnnotation <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    annot.ext <span class="op">=</span> <span class="fu">chrZ_gtf</span><span class="op">(</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    isGTF <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># (3) Analyse the aligned reads in the BAM file for low-mappability regions:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">calculateMappability</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="va">ref_path</span>,</span></span>
<span class="r-in"><span>    aligned_bam <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ref_path</span>, <span class="st">"Mappability"</span>, <span class="st">"AlignedReads.bam"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># (4) Build the example reference using the calculated Mappability Exclusions</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="Build-Reference-methods.html">buildRef</a></span><span class="op">(</span><span class="va">ref_path</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># NB the default is to search for the BED file generated by</span></span></span>
<span class="r-in"><span><span class="co"># `calculateMappability()` in the given reference_path</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alex Chit Hei Wong.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

