<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>SpliceWiz: efficient and precise alternative splicing analysis in R — SpliceWiz-package • SpliceWiz</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="SpliceWiz: efficient and precise alternative splicing analysis in R — SpliceWiz-package"><meta property="og:description" content="SpliceWiz is a computationally efficient and user friendly workflow that
analyses aligned short-read RNA sequencing for differential
intron retention and alternative splicing."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SpliceWiz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/SW_Cookbook.html">SpliceWiz: the cookbook</a>
    </li>
    <li>
      <a href="../articles/SW_QuickStart.html">SpliceWiz: Quick Start</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/alexchwong/SpliceWiz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>SpliceWiz: efficient and precise alternative splicing analysis in R</h1>
    <small class="dont-index">Source: <a href="https://github.com/alexchwong/SpliceWiz/blob/HEAD/R/SpliceWiz-package.R" class="external-link"><code>R/SpliceWiz-package.R</code></a></small>
    <div class="hidden name"><code>SpliceWiz-package.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>SpliceWiz is a computationally efficient and user friendly workflow that
analyses aligned short-read RNA sequencing for differential
intron retention and alternative splicing.</p>
    </div>


    <div id="details">
    <h2>Details</h2>
    <p>SpliceWiz uses isoform-specific alignments to quantify percent-spliced-in
ratios (i.e. ratio of the "included" isoform, as a proportion of "included"
and "excluded" isoforms). For intron retention (IR), the abundance of the
intron-retaining transcript (included isoform) is quantified using the
trimmed-mean depth of intron coverage with reads, whereas the spliced
transcript (excluded isoform) is measured as the splicing of the intron as
well as that of overlapping introns (since splicing of any overlapping intron
implies the intron of interest is not retained). For other forms of
alternative splicing, junction reads (reads aligned across splice junctions)
are used to quantify included and excluded isoforms.</p>
<p>SpliceWiz processes BAM files (aligned RNA sequencing) using
<a href="https://rdrr.io/pkg/ompBAM/man/ompBAM-package.html" class="external-link">ompBAM::ompBAM-package</a>. ompBAM
is a C++ library that allows R packages (via the Rcpp framework) to
efficiently read BAM files using OpenMP-based multi-threading. SpliceWiz
processes BAM files via the <a href="processBAM.html">processBAM</a> function, using a splicing and
intron reference built from any given genome / gene annotation resource
using the <a href="Build-Reference-methods.html">buildRef</a> function. <a href="processBAM.html">processBAM</a> generates two outputs per
BAM file: a <code>txt.gz</code> file which is a gzip-compressed text file with multiple
tables, containing information including junction read counts and intron
retention metrics. This output is very similar to that of
<a href="https://github.com/williamritchie/IRFinder" class="external-link">IRFinder</a>, as the analysis
steps of SpliceWiz's BAM processing was built on an improved version of
IRFinder's source code (version 1.3.1). Additionally, <a href="processBAM.html">processBAM</a> outputs
a COV file, which is a binary bgzf-compressed file that contains
strand-specific coverage data.</p>
<p>Once individual files have been analysed, SpliceWiz compiles a dataset using
these individual outputs, using <a href="collateData.html">collateData</a>. This function unifies
junctions detected across the dataset, and generates included / excluded
counts of all putative IR events and annotated alternative splicing events
(ASEs). This dataset is exported as a collection of files including an
H5 database. The data is later imported into the R session using the
<a href="makeSE.html">makeSE</a> function, as a <a href="NxtSE-class.html">NxtSE</a> object.</p>
<p>The <a href="NxtSE-class.html">NxtSE</a> object is a specialized
<a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a> object tailored for use in SpliceWiz.
Annotation of rows provide information about ASEs via <a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a>, while
columns allows users to provide annotations via <a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a>.</p>
<p>SpliceWiz offers several novel filters via the <a href="ASEFilter-class.html">ASEFilter</a>
class. See <a href="ASEFilter-class.html">ASEFilter</a> for details.</p>
<p>Once the <a href="NxtSE-class.html">NxtSE</a> is annotated and filtered, differential
analysis is performed, using limma, DESeq2 or DoubleExpSeq wrappers.
These wrappers model isoform counts as log-normal, negative-binomial,
or beta-binomial distributions, respectively. See <a href="ASE-methods.html">ASE-methods</a> for details.</p>
<p>Finally, SpliceWiz provides visualisation tools to illustrate alternative
splicing using coverage plots, including a novel method to normalise RNA-seq
coverage grouped by experimental condition. This approach accounts for
variations introduced by sequenced library size and gene expression.
SpliceWiz efficiently computes and visualises means and variations in
per-nucleotide coverage depth across alternate exons in genomic loci.</p>
<p>The main functions are:</p><ul><li><p><a href="Build-Reference-methods.html">Build-Reference-methods</a> - Prepares genome and gene annotation
references from FASTA and GTF files and synthesizes the SpliceWiz reference
for processing BAM files, collating the <a href="NxtSE-class.html">NxtSE</a> object.</p></li>
<li><p><a href="STAR-methods.html">STAR-methods</a> - (Optional) Provides wrapper functions to build the STAR
genome reference and alignment of short-read FASTQ raw sequencing files.
This functionality is only available on systems with STAR installed.</p></li>
<li><p><a href="processBAM.html">processBAM</a> - OpenMP/C++ based algorithm to analyse
single or multiple BAM files.</p></li>
<li><p><a href="collateData.html">collateData</a> - Collates an experiment based on multiple IRFinder outputs
for individual samples, into one unified H5-based data structure.</p></li>
<li><p><a href="makeSE.html">makeSE</a> - Constructs a <a href="NxtSE-class.html">NxtSE</a> (H5-based
SummarizedExperiment) object, specialised to house measurements of retained
introns and junction counts of alternative splice events.</p></li>
<li><p><a href="Run_SpliceWiz_Filters.html">applyFilters</a> - Use default or custom filters to remove alternative
splicing or IR events pertaining to low-abundance genes and transcripts.</p></li>
<li><p><a href="ASE-methods.html">ASE-methods</a> - one-step method to perform differential alternate splice
event (ASE) analysis on a NxtSE object using limma or DESeq2.</p></li>
<li><p><a href="make_plot_data.html">make_plot_data</a>: Functions that compile individual and group-mean percent
spliced in (PSI) values of IR and alternative splice events; useful to
produce scatter plots or heatmaps.</p></li>
<li><p><a href="plotCoverage.html">plotCoverage</a>: Generate RNA-seq coverage plots of
individual samples or across samples grouped by user-specified conditions</p></li>
</ul><p>See the
<a href="../doc/SW_QuickStart.html">SpliceWiz Quick-Start</a>
for worked examples on how to use SpliceWiz
<a href="../doc/SW_Cookbook.html">SpliceWiz Cookbook</a>
for real-life usage examples</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Middleton R, Gao D, Thomas A, Singh B, Au A, Wong JJ, Bomane A, Cosson B,
Eyras E, Rasko JE, Ritchie W.
IRFinder: assessing the impact of intron retention on mammalian gene
expression. Genome Biol. 2017 Mar 15;18(1):51.
<a href="https://doi.org/10.1186/s13059-017-1184-4" class="external-link">https://doi.org/10.1186/s13059-017-1184-4</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Alex Wong</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alex Chit Hei Wong.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

