<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>STAR wrappers for building reference for STAR, and aligning RNA-sequencing — STAR-methods • SpliceWiz</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="STAR wrappers for building reference for STAR, and aligning RNA-sequencing — STAR-methods"><meta property="og:description" content="These functions run the STAR aligner to build a STAR genome reference,
calculate mappability exclusion regions using STAR, and align one or more
FASTQ files (single or paired) to the generated genome. These functions only
work on Linux-based systems with STAR installed. STAR must be
accessible via $PATH. See details and examples"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SpliceWiz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/SW_Cookbook.html">SpliceWiz: the cookbook</a>
    </li>
    <li>
      <a href="../articles/SW_QuickStart.html">SpliceWiz: Quick Start</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/alexchwong/SpliceWiz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>STAR wrappers for building reference for STAR, and aligning RNA-sequencing</h1>
    <small class="dont-index">Source: <a href="https://github.com/alexchwong/SpliceWiz/blob/HEAD/R/STAR_utils.R" class="external-link"><code>R/STAR_utils.R</code></a></small>
    <div class="hidden name"><code>STAR-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>These functions run the STAR aligner to build a STAR genome reference,
calculate mappability exclusion regions using STAR, and align one or more
FASTQ files (single or paired) to the generated genome. These functions only
work on Linux-based systems with STAR installed. STAR must be
accessible via <code>$PATH</code>. See details and examples</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">STAR_version</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">STAR_buildRef</span><span class="op">(</span></span>
<span>  <span class="va">reference_path</span>,</span>
<span>  STAR_ref_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">reference_path</span>, <span class="st">"STAR"</span><span class="op">)</span>,</span>
<span>  also_generate_mappability <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  map_depth_threshold <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  sjdbOverhang <span class="op">=</span> <span class="fl">149</span>,</span>
<span>  n_threads <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  additional_args <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">STAR_mappability</span><span class="op">(</span></span>
<span>  <span class="va">reference_path</span>,</span>
<span>  STAR_ref_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">reference_path</span>, <span class="st">"STAR"</span><span class="op">)</span>,</span>
<span>  map_depth_threshold <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  n_threads <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">STAR_alignExperiment</span><span class="op">(</span></span>
<span>  <span class="va">Experiment</span>,</span>
<span>  <span class="va">STAR_ref_path</span>,</span>
<span>  <span class="va">BAM_output_path</span>,</span>
<span>  trim_adaptor <span class="op">=</span> <span class="st">"AGATCGGAAG"</span>,</span>
<span>  two_pass <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_threads <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">STAR_alignReads</span><span class="op">(</span></span>
<span>  fastq_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"./sample_1.fastq"</span><span class="op">)</span>,</span>
<span>  fastq_2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">STAR_ref_path</span>,</span>
<span>  <span class="va">BAM_output_path</span>,</span>
<span>  two_pass <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  trim_adaptor <span class="op">=</span> <span class="st">"AGATCGGAAG"</span>,</span>
<span>  memory_mode <span class="op">=</span> <span class="st">"NoSharedMemory"</span>,</span>
<span>  additional_args <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_threads <span class="op">=</span> <span class="fl">4</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>reference_path</dt>
<dd><p>The path to the reference.
<a href="Build-Reference-methods.html">getResources</a> must first be run using this path
as its <code>reference_path</code></p></dd>


<dt>STAR_ref_path</dt>
<dd><p>(Default - the "STAR" subdirectory under
<code>reference_path</code>) The directory containing the STAR reference to be
used or to contain the newly-generated STAR reference</p></dd>


<dt>also_generate_mappability</dt>
<dd><p>Whether <code>STAR_buildRef()</code> also
calculates Mappability Exclusion regions.</p></dd>


<dt>map_depth_threshold</dt>
<dd><p>(Default 4) The depth of mapped reads
threshold at or below which Mappability exclusion regions are defined. See
<a href="Mappability-methods.html">Mappability-methods</a>.
Ignored if <code>also_generate_mappability = FALSE</code></p></dd>


<dt>sjdbOverhang</dt>
<dd><p>(Default = 149) A STAR setting indicating the length of
the donor / acceptor sequence on each side of the junctions. Ideally equal
to (mate_length - 1). As the most common read length is 150, the default
of this function is 149. See the STAR aligner manual for details.</p></dd>


<dt>n_threads</dt>
<dd><p>The number of threads to run the STAR aligner.</p></dd>


<dt>additional_args</dt>
<dd><p>A character vector of additional arguments to be
parsed into STAR. See examples below.</p></dd>


<dt>...</dt>
<dd><p>Additional arguments to be parsed into
<code><a href="Mappability-methods.html">generateSyntheticReads()</a></code>. See <a href="Mappability-methods.html">Mappability-methods</a>.</p></dd>


<dt>Experiment</dt>
<dd><p>A two or three-column data frame with the columns denoting
sample names, forward-FASTQ and reverse-FASTQ files. This can be
conveniently generated using <a href="findSamples.html">findFASTQ</a></p></dd>


<dt>BAM_output_path</dt>
<dd><p>The path under which STAR outputs the aligned BAM
files. In <code>STAR_alignExperiment()</code>, STAR will output aligned
BAMS inside subdirectories of this folder, named by sample names. In
<code>STAR_alignReads()</code>, STAR will output directly into this path.</p></dd>


<dt>trim_adaptor</dt>
<dd><p>The sequence of the Illumina adaptor to trim via STAR's
<code>--clip3pAdapterSeq</code> option</p></dd>


<dt>two_pass</dt>
<dd><p>Whether to use two-pass mapping. In
<code>STAR_alignExperiment()</code>, STAR will first align every sample
and generate a list of splice junctions but not BAM files. The junctions
are then given to STAR to generate a temporary genome (contained within
<code>_STARgenome</code>) subdirectory within that of the first sample), using
these junctions to improve novel junction detection. In
<code>STAR_alignReads()</code>, STAR will run <code>--twopassMode Basic</code></p></dd>


<dt>fastq_1, fastq_2</dt>
<dd><p>In STAR_alignReads: character vectors giving the
path(s) of one or more FASTQ (or FASTA) files to be aligned.
If single reads are to be aligned, omit <code>fastq_2</code></p></dd>


<dt>memory_mode</dt>
<dd><p>The parameter to be parsed to <code>--genomeLoad</code>; either
<code>NoSharedMemory</code> or <code>LoadAndKeep</code> are used.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>None. STAR will output files into the given output directories.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>Pre-requisites</strong></p>
<p><code>STAR_buildRef</code> requires <a href="Build-Reference-methods.html">getResources</a> to be run to fetch the
required genome and gene annotation files.</p>
<p><code>STAR_mappability</code>, <code>STAR_alignExperiment</code> and <code>STAR_alignReads</code> requires a
<code>STAR</code> genome, which can be built using <code>STAR_buildRef</code></p>
<p><strong>Function Description</strong></p>
<p>For <code>STAR_buildRef</code>: this function
will create a <code>STAR</code> genome reference in the <code>STAR</code> subdirectory in the
path given by <code>reference_path</code>. Optionally, it will run <code>STAR_mappability</code>
if <code>also_generate_mappability</code> is set to <code>TRUE</code></p>
<p>For <code>STAR_mappability</code>: this function will first
will run <a href="Mappability-methods.html">generateSyntheticReads</a>, then use the given <code>STAR</code> genome to
align the synthetic reads using <code>STAR</code>. The aligned BAM file will then be
processed using <a href="Mappability-methods.html">calculateMappability</a> to calculate the
lowly-mappable genomic regions,
producing the <code>MappabilityExclusion.bed.gz</code> output file.</p>
<p>For <code>STAR_alignReads</code>: aligns a single or pair of FASTQ files to the given
<code>STAR</code> genome using the <code>STAR</code> aligner.</p>
<p>For <code>STAR_alignExperiment</code>: aligns a set of FASTQ or paired FASTQ files
using the given
<code>STAR</code> genome using the <code>STAR</code> aligner.
A data.frame specifying sample names and corresponding FASTQ files are
required</p>
    </div>
    <div id="functions">
    <h2>Functions</h2>
    
<ul><li><p><code>STAR_version</code>: Checks whether STAR is installed, and its version</p></li>
<li><p><code>STAR_buildRef</code>: Creates a STAR genome reference.</p></li>
<li><p><code>STAR_mappability</code>: Calculates lowly-mappable genomic regions using STAR</p></li>
<li><p><code>STAR_alignExperiment</code>: Aligns multiple sets of FASTQ files, belonging to
multiple samples</p></li>
<li><p><code>STAR_alignReads</code>: Aligns a single sample (with single or paired FASTQ
or FASTA files)</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="Build-Reference-methods.html">Build-Reference-methods</a> <a href="findSamples.html">findSamples</a> <a href="Mappability-methods.html">Mappability-methods</a><br><br><a href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf" class="external-link">The latest STAR documentation</a></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># 0) Check that STAR is installed and compatible with SpliceWiz</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">STAR_version</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:52 STAR wrappers are only supported on Linux</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Nov 15 16:16:52 STAR is not installed</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># The below workflow illustrates</span></span></span>
<span class="r-in"><span><span class="co"># 1) Getting the reference resource</span></span></span>
<span class="r-in"><span><span class="co"># 2) Building the STAR Reference, including Mappability Exclusion calculation</span></span></span>
<span class="r-in"><span><span class="co"># 3) Building the SpliceWiz Reference, using the Mappability Exclusion file</span></span></span>
<span class="r-in"><span><span class="co"># 4) Aligning (a) one or (b) multiple raw sequencing samples.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 1) Reference generation from Ensembl's FTP links</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">FTP</span> <span class="op">&lt;-</span> <span class="st">"ftp://ftp.ensembl.org/pub/release-94/"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="Build-Reference-methods.html">getResources</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="st">"Reference_FTP"</span>,</span></span>
<span class="r-in"><span>    fasta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">FTP</span>, <span class="st">"fasta/homo_sapiens/dna/"</span>,</span></span>
<span class="r-in"><span>        <span class="st">"Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    gtf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">FTP</span>, <span class="st">"gtf/homo_sapiens/"</span>,</span></span>
<span class="r-in"><span>        <span class="st">"Homo_sapiens.GRCh38.94.chr.gtf.gz"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 2) Generates STAR genome within the SpliceWiz reference. Also generates</span></span></span>
<span class="r-in"><span><span class="co"># mappability exclusion gzipped BED file inside the "Mappability/" sub-folder</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">STAR_buildRef</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="st">"Reference_FTP"</span>,</span></span>
<span class="r-in"><span>    n_threads <span class="op">=</span> <span class="fl">8</span>,</span></span>
<span class="r-in"><span>    also_generate_mappability <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 2 alt) Generates STAR genome of the example SpliceWiz genome.</span></span></span>
<span class="r-in"><span><span class="co">#     This demonstrates using custom STAR parameters, as the example </span></span></span>
<span class="r-in"><span><span class="co">#     SpliceWiz genome is ~100k in length, </span></span></span>
<span class="r-in"><span><span class="co">#     so --genomeSAindexNbases needs to be</span></span></span>
<span class="r-in"><span><span class="co">#     adjusted to be min(14, log2(GenomeLength)/2 - 1)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="Build-Reference-methods.html">getResources</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="st">"Reference_chrZ"</span>,</span></span>
<span class="r-in"><span>    fasta <span class="op">=</span> <span class="fu">chrZ_genome</span><span class="op">(</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    gtf <span class="op">=</span> <span class="fu">chrZ_gtf</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">STAR_buildRef</span><span class="op">(</span></span></span>
<span class="r-in"><span>    reference_path <span class="op">=</span> <span class="st">"Reference_chrZ"</span>,</span></span>
<span class="r-in"><span>    n_threads <span class="op">=</span> <span class="fl">8</span>,</span></span>
<span class="r-in"><span>    additional_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"--genomeSAindexNbases"</span>, <span class="st">"7"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    also_generate_mappability <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 3) Build SpliceWiz reference using the newly-generated </span></span></span>
<span class="r-in"><span><span class="co">#    Mappability exclusions</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#' NB: also specifies to use the hg38 nonPolyA resource</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="Build-Reference-methods.html">buildRef</a></span><span class="op">(</span>reference_path <span class="op">=</span> <span class="st">"Reference_FTP"</span>, genome_type <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 4a) Align a single sample using the STAR reference</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">STAR_alignReads</span><span class="op">(</span></span></span>
<span class="r-in"><span>    STAR_ref_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"Reference_FTP"</span>, <span class="st">"STAR"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    BAM_output_path <span class="op">=</span> <span class="st">"./bams/sample1"</span>,</span></span>
<span class="r-in"><span>    fastq_1 <span class="op">=</span> <span class="st">"sample1_1.fastq"</span>, fastq_2 <span class="op">=</span> <span class="st">"sample1_2.fastq"</span>,</span></span>
<span class="r-in"><span>    n_threads <span class="op">=</span> <span class="fl">8</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 4b) Align multiple samples, using two-pass alignment</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">Experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_A"</span>, <span class="st">"sample_B"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    forward <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"raw_data"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_A"</span>, <span class="st">"sample_B"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_A_1.fastq"</span>, <span class="st">"sample_B_1.fastq"</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    reverse <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"raw_data"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_A"</span>, <span class="st">"sample_B"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_A_2.fastq"</span>, <span class="st">"sample_B_2.fastq"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">STAR_alignExperiment</span><span class="op">(</span></span></span>
<span class="r-in"><span>    Experiment <span class="op">=</span> <span class="va">Experiment</span>,</span></span>
<span class="r-in"><span>    STAR_ref_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"Reference_FTP"</span>, <span class="st">"STAR"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    BAM_output_path <span class="op">=</span> <span class="st">"./bams"</span>,</span></span>
<span class="r-in"><span>    two_pass <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    n_threads <span class="op">=</span> <span class="fl">8</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alex Chit Hei Wong.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

